---
title: "scRNA_Seq_workflow"
author: "Carlos Eduardo Madureira Trufen"
date: "2022-10-14"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown


# Load libraries
```{r}
# library(DropletUtils)

library(AUCell)
library(aplot)
library(batchelor)
# library(beachmat)
# library(biomaRt)
library(bluster)
library(cluster)
library(celldex)
library(corral)
library(cowplot)
# library(clusterProfiler)
library(clustree)
library(DESeq2)
library(edgeR) 
library(factoextra)
library(ggeasy)
library(ggpubr)
library(GSEABase)
library(HDF5Array)
library(kBET)
library(limma)
library(Matrix)
library(mmtable2)
library(org.Mm.eg.db)
library(msigdbr)
library(patchwork)
library(PCAtools)
library(pheatmap)
library(qs)
library(Rhdf5lib)
library(rhdf5)
library(Seurat)
library(slingshot)
library(scran)
library(scater)
library(scDblFinder)
library(scRNAseq)
library(SingleCellExperiment)
library(tictoc)
library(tidyverse)
library(TSCAN)
library(uwot)
library(zellkonverter)
```

# Define colors as in scater package
```{r}
# form https://rdrr.io/bioc/scater/src/R/plot_colours.R
tableau_10_medium <- c(
  "#729ECE", "#FF9E4A", "#67BF5C", "#ED665D",
  "#AD8BC9", "#A8786E", "#ED97CA", "#A2A2A2",
  "#CDCC5D", "#6DCCDA"
)
```

```{r}
# ora_analysis <- function(genes_list, database = KEGG_2019_Mouse, subset = c(), pvalue_cutoff = 1, qvalue_cutoff = 1, p.adjust_cutoff = 0.1, count_at_least = 3){
ora_analysis <- function(genes_list, database = KEGG_2019_Mouse, pvalue_cutoff = 1, qvalue_cutoff = 1, p.adjust_cutoff = 0.1, count_at_least = 3){
  library(clusterProfiler)
  library(tidyverse)
  gmt_df <- database %>% clusterProfiler::read.gmt()
  # gmt_df <- gmt_df %>% dplyr::filter(term %in% subset)
  terms_list <- gmt_df %>% split(f = .$term) %>% purrr::map(pull, gene)
  genes_set <- gmt_df[, "gene"] %>% unique()
  enrichment_result <- genes_list %>% 
    purrr::map(clusterProfiler::enricher, pvalueCutoff = pvalue_cutoff, qvalueCutoff = qvalue_cutoff, universe = genes_set, TERM2GENE = gmt_df)
  clProf.df <- enrichment_result %>% purrr::map(as.data.frame) %>% bind_rows(.id = "Cluster")
  clProf.df <- clProf.df %>% dplyr::filter(p.adjust < p.adjust_cutoff & Count >= count_at_least)
  geneClusters <- genes_list
  compareCluster_output <- new("compareClusterResult", 
                               compareClusterResult = clProf.df, 
                               geneClusters = list(geneClusters), 
                               fun = "enricher",
                               .call = match.call(expand.dots = TRUE))
  return(compareCluster_output)
}

shape_data <- function(ora_results_output) {
  ora_results_output@compareClusterResult <- ora_results_output@compareClusterResult %>% 
    mutate_at(.vars = c("pvalue", "p.adjust", "qvalue"), .funs = formatC, digits = 2, format = "e") 
  return(ora_results_output)
}

get_genes_from_pathway <- function(df, pathway){
  df %>% 
  dplyr::filter(Description == pathway) %>% 
  pull("geneID") %>% 
  str_split(pattern = "/") %>% 
  unlist() %>% 
  sort() %>% 
  stringr::str_to_title()
}

save_pheatmap_tiff <- function(x, filename, width = 12, height = 7, res = 600) {
   stopifnot(!missing(x))
   stopifnot(!missing(filename))
   tiff(filename, width = width, height = height, units = "cm", res = res)
   grid::grid.newpage()
   grid::grid.draw(x$gtable)
   dev.off()
}
# RIG_I_genes <- ora_results_Kegg_list$`2.1` %>% get_genes_from_pathway(pathway = "RIG-I-like receptor signaling pathway")

# creating sparse data https://stackoverflow.com/questions/49758083/how-to-create-sparse-matrix-in-r
library(Matrix)

# https://stackoverflow.com/questions/43117608/r-binding-sparse-matrices-of-different-sizes-on-rows
merge_sparse <- function(...) {
  
  cnnew <- character()
  rnnew <- character()
  x <- vector()
  i <- numeric()
  j <- numeric()
  
  for (M in list(...)) {
  
  cnold <- colnames(M)
  rnold <- rownames(M)
  
  cnnew <- union(cnnew,cnold)
  rnnew <- union(rnnew,rnold)
  
  cindnew <- match(cnold,cnnew)
  rindnew <- match(rnold,rnnew)
  ind <- unname(which(M != 0,arr.ind=T))
  i <- c(i,rindnew[ind[,1]])
  j <- c(j,cindnew[ind[,2]])
  x <- c(x,M@x)
  }
  
  sparseMatrix(i=i,j=j,x=x,dims=c(length(rnnew),length(cnnew)),dimnames=list(rnnew,cnnew))
}
```

# Enrichment analysis (ORA)
# custom ORA
# Databases
```{r}
path_to_gmt <- "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/GMT/"
# path_to_gmt <- "/run/user/1002/gvfs/smb-share:server=primus.img.cas.cz,share=home/trufenc/GMT/"

WikiPathways_2019_Human <- path_to_gmt %>% paste0("WikiPathways_2019_Human.txt") # WikiPathways_2019_Human gene sets
WikiPathways_2019_Mouse <- path_to_gmt %>% paste0("WikiPathways_2019_Mouse.txt") # WikiPathways_2019_Mouse gene sets
KEGG_2019_Human <- path_to_gmt %>% paste0("KEGG_2019_Human.txt") # KEGG_2019_Human gene sets
KEGG_2019_Mouse <- path_to_gmt %>% paste0("KEGG_2019_Mouse.txt") # KEGG_2019_Mouse gene sets
LINCS_L1000_Chem_Pert_down <- path_to_gmt %>% paste0("LINCS_L1000_Chem_Pert_down.txt") # LINCS_L1000_Chem_Pert_down gene sets
LINCS_L1000_Chem_Pert_up <- path_to_gmt %>% paste0("LINCS_L1000_Chem_Pert_up.txt") # LINCS_L1000_Chem_Pert_up gene sets
Reactome_2016 <- path_to_gmt %>% paste0("Reactome_2016.txt") # Reactome_2016 gene sets
GO_Molecular_Function_2018 <- path_to_gmt %>% paste0("GO_Molecular_Function_2018.txt") # GO_Molecular_Function_2018 gene sets
GO_Cellular_Component_2018 <- path_to_gmt %>% paste0("GO_Cellular_Component_2018.txt") # GO_Cellular_Component_2018 gene sets
GO_Biological_Process_2018 <- path_to_gmt %>% paste0("GO_Biological_Process_2018.txt") # GO_Biological_Process_2018 gene sets
GO_Molecular_Function_2021 <- path_to_gmt %>% paste0("GO_Molecular_Function_2021.txt") # GO_Molecular_Function_2021 gene sets
GO_Cellular_Component_2021 <- path_to_gmt %>% paste0("GO_Cellular_Component_2021.txt") # GO_Cellular_Component_2021 gene sets
GO_Biological_Process_2021 <- path_to_gmt %>% paste0("GO_Biological_Process_2021.txt") # GO_Biological_Process_2021 gene sets
BioPlanet_2019 <- path_to_gmt %>% paste0("BioPlanet_2019.txt") # BioPlanet_2019 gene sets

C8_cell_type_signatures <- path_to_gmt %>% paste0("c8.all.v7.4.symbols.gmt") # C8 MSigDB
```

# Path to files
```{r}
# source_path <- "/media/carlos/HD_10_TB/scRNAseq_Immunology/raw_data/"
# source_path <- "/run/user/1002/gvfs/smb-share:server=primus.img.cas.cz,share=home/trufenc/scRNAseq_Immunology/raw_data/"
source_path <- "~/PRIMUS/home/trufenc/scRNAseq_Immunology/raw_data/"
# source_path <- "~/PRIMUS/transfer/Kadu/scRNAseq_Immunology/raw_data/"
# source_path <- "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/raw_data/"
```

```{r}
# path_to_files_E7 <- "2021_Splichalova_E7_aggr_filtered"
path_to_files_E7 <- "2021_Splichalova_E7_single_filtered/outs/filtered_feature_bc_matrix.h5"
path_to_files_E8 <- "2021_Splichalova_E8_merged_filtered/outs/filtered_feature_bc_matrix.h5"
path_to_files_E9 <- "2021_Splichalova_E9_merged_filtered/outs/filtered_feature_bc_matrix.h5"  
```

# Load data, convert to matrix
```{r}
# path_to_file <- path_to_files_E7
# path_to_file <- path_to_files_E8
# path_to_file <- path_to_files_E9
# 
# file_to_count_data <- function(path_to_file){
#   
#   sc_data <- source_path %>%
#     paste0(path_to_file) %>%
#     h5read(name = "matrix")
#   
#   features_df <- sc_data$features %>% bind_cols() %>% rownames_to_column() %>% mutate(rowname = rowname %>% as.numeric)
#   features_df <- features_df %>% mutate(name = name %>% make.unique)
#   features_df <- qread(file = paste0(source_path, "feature_data.qs"))
#   
#   mart <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", dataset = "mmusculus_gene_ensembl")
#   genes_info <- biomaRt::getBM(values = features_df$id, attributes = c("entrezgene_id"), mart = mart)
#   counts_matrix <- sparseMatrix(
#     dims = sc_data$shape,
#     i = as.numeric(sc_data$indices),
#     p = as.numeric(sc_data$indptr),
#     x = as.numeric(sc_data$data),
#     index1 = FALSE
#     )
#   
#   spread_df <- counts_matrix %>% summary() %>% as.data.frame()
#   df <- spread_df %>% pivot_wider(names_from = j, values_from = x) %>% purrr::set_names("rowname", sc_data$barcodes)
#   df <- df %>% inner_join(features_df) %>% dplyr::select(names(features_df), everything())
#   df <- df %>% dplyr::select(-c("rowname", "_all_tag_keys", "feature_type", "genome", "id"))
#   df <- df %>% remove_rownames() %>% column_to_rownames("name") 
#   df <- df %>% replace(is.na(.), 0)
#   counts_matrix <- df %>% as.matrix()
#   }
```

# Load data, convert to sce file
```{r}
load_h5_into_sce <- function(path_to_file) {
  library(Matrix)
  library(rhdf5)
  library(SingleCellExperiment)
  library(tidyverse)
  
  sc_data <- path_to_file %>%
    h5read(name = "matrix")

  counts_matrix <- sparseMatrix(
    dims = sc_data$shape,
    i = as.numeric(sc_data$indices),
    p = as.numeric(sc_data$indptr),
    x = as.numeric(sc_data$data),
    index1 = FALSE
  )

  spread_df <- counts_matrix %>%
    summary() %>%
    as.data.frame()

  df <- spread_df %>%
    pivot_wider(names_from = j, values_from = x) %>%
    purrr::set_names("rowname", sc_data$barcodes)

  df <- df %>%
    inner_join(features_df) %>%
    dplyr::select(names(features_df), everything())

  df <- df %>% dplyr::select(-c("rowname", "_all_tag_keys", "feature_type", "genome", "id"))

  df <- df %>%
    remove_rownames() %>%
    column_to_rownames("name")

  df <- df %>% replace(is.na(.), 0)

  counts_matrix <- df %>% as.matrix()

  sce <- SingleCellExperiment(assays = list(counts = counts_matrix), colData = sc_data$barcodes)
  
  rowData(sce) <- rownames(df)
  
  return(sce)
}

sce_E7 <- source_path %>%
  paste0(path_to_files_E7) %>% 
  load_h5_into_sce()
  
```

# Quality control - Mitochondrial genes
```{r}
# I will consider it was already done

sce_7.5 <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_7.5_raw_reboot_2022_06_10.qs", nthreads = 50)
sce_8.5 <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_8.5_raw_reboot_2022_06_10.qs", nthreads = 50)
sce_9.5  <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_9.5_raw_reboot_2022_06_10.qs", nthreads = 50)

sce <- sce_7.5
sce <- sce_8.5
sce <- sce_9.5

qc_mito <- function(sce){
  rowData(sce) <- rowData(sce) %>% 
    as.data.frame() %>% 
    mutate(featureType = dplyr::if_else(condition = value %>% str_detect(pattern = "mt-"), true = "mito", false = "cellular")) %>% as("DFrame")
  
  colData(sce) <- sce %>% perCellQCMetrics(subsets = list(Mt = rowData(sce)$featureType=="mito"))
  qc <- data.frame(
    low_lib_size = isOutlier(sce[["sum"]], log = TRUE, type = "lower"),
    low_n_features = isOutlier(sce[["detected"]], log = TRUE, type = "lower"),
    Mt_percent = isOutlier(sce[["subsets_Mt_percent"]], type = "higher")
    ) %>% 
    mutate(discard = purrr::reduce(., `|`)) %>% 
    as("DFrame")
  colData(sce) <- as.data.frame(colData(sce)) %>% cbind(as.data.frame(qc)) %>% as("DFrame")
  sce$qc_mito_discard <- qc$discard
  sce$Mt_percent <- qc$Mt_percent
  return(sce)
  }

sce <- sce %>% qc_mito()

sce %>% plotColData(x = "sum", y = "subsets_Mt_percent", colour_by = "discard")
sce %>% plotColData(x = "sum", y = "subsets_Mt_percent", colour_by = "Mt_percent")
```

```{r}
process_1_sce <- function(sce){
  # Size factors
  sizeFactors(sce) <- scater::librarySizeFactors(sce)
  # Normalize and log-transform data
  sce  <- sce %>% scater::logNormCounts()
  
  ### Feature selection
  # Remove ribosomal genes
  c2_sets <- msigdbr(species = "Mus musculus", category = "C2")
  ribo_set <- c2_sets %>% dplyr::filter(gs_name == "KEGG_RIBOSOME") %>% pull(gene_symbol)
  ribo_discard <- sce %>% rownames() %in% ribo_set
  
  sce <- sce[!ribo_discard, ] # remove ribosomal genes
  sce <- sce[!rownames(sce) %>% str_detect(pattern = "Hist"), ]# remove histone genes
  
  # Variance of the log-counts
  dec_default <- sce %>% modelGeneVar()
  # dec_default %>% qs::qsave(file = paste0(source_path, "dec_default.qs"))
  fit_default <- metadata(dec_default)
  
  # Coefficient of variation
  dec_cv2 <- sce %>% modelGeneCV2()
  fit_cv2 <- dec_cv2 %>% metadata()
  plot(fit_cv2$mean, fit_cv2$cv2, log = "xy")
  curve(fit_cv2$trend(x), col = "dodgerblue", add = TRUE, lwd = 2)
  
  # Based on the largest metrics
  hvg_var <- dec_default %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames()
  
  gene_subset <- hvg_var %>% unique()
  
  return(list(sce = sce,
              gene_subset = gene_subset))
} 

processed_1_sce <- sce %>% process_1_sce()

```

# Dimensionality reduction results
```{r}
seed <- 123
gene_subset <- processed_1_sce$gene_subset
sce <- processed_1_sce$sce
  
sce_dimentionality_reduction <- function(sce, gene_subset, seed = 3){
  ## Choosing the number of PCs
  set.seed(seed)
  # bpparam <- BiocParallel::MulticoreParam(workers = n_threads)
  sce <- sce %>% scater::runPCA(ncomponents = 50, subset_row = gene_subset)
  
  # Based on population structure
  # This strategy is the most pragmatic as it directly addresses the role of the bias-variance trade-off in downstream analyses, specificallyclustering.
  pcs <- sce %>% reducedDim()
  choices <- pcs %>% getClusteredPCs()
  val <- metadata(choices)$chosen
  reducedDim(sce, "PCA_clust") <- pcs[, 1:val]
  
  plot(choices$n.pcs, choices$n.clusters,
  xlab = "Number of PCs", ylab = "Number of clusters"
)
  abline(a = 1, b = 1, col = "red")
  abline(v = val, col = "grey80", lty = 2)

  # tsne
  # set.seed(123)
  set.seed(seed)
  sce <- sce %>% scater::runTSNE(dimred = "PCA_clust", perplexity = 20)

  # umap - Uniform manifold approximation and projection
  # set.seed(123)
  set.seed(seed)
  sce <- sce %>% scater::runUMAP(dimred = 'PCA_clust', n_neighbors = 20)
  return(sce)
}

sce <- processed_1_sce$sce %>% sce_dimentionality_reduction(gene_subset = processed_1_sce$gene_subset)

## Dimensionality reduction for visualization
plotReducedDim(sce, dimred = "PCA_clust")
plotReducedDim(sce, dimred = "PCA_clust", ncomponents = 4)
sce %>% plotReducedDim(dimred = "TSNE")
sce %>% plotReducedDim(dimred = "UMAP")

cowplot::plot_grid(
  sce %>% plotReducedDim(dimred = "TSNE"), 
  sce %>% plotReducedDim(dimred = "UMAP")
  )
```

# Doublet detection with clusters
```{r}
# findDoubletClusters() identifies clusters with expression profiles lying between two other clusters
set.seed(123)
# bpparam <- BiocParallel::MulticoreParam(workers = 51)
sce <- sce %>% scDblFinder(clusters = colLabels(sce))

sce_dbl_finder_results_df <- data.frame(
  cxds_score = sce$scDblFinder.cxds_score,
  weighted = sce$scDblFinder.weighted,
  score = sce$scDblFinder.score,
  class = sce$scDblFinder.class
  )

cowplot::plot_grid(sce %>% plotTSNE(colour_by = "scDblFinder.score"), 
                   sce %>% plotUMAP(colour_by = "scDblFinder.score")
                   )

# removing the doublets
sce_no_doublet <- sce[, which(sce$scDblFinder.class != "doublet")]
# sce_no_doublet <- sce[, which(sce$scDblFinder.score < 0.9)]

# TSNE with different seeds
# for(i in 1:100){
#   set.seed(i)
#   print(i)
#   sce_no_doublet <- sce_no_doublet %>% runTSNE(dimred = "PCA_clust")
#   sce_no_doublet %>%
#     plotTSNE() %>%
#     ggsave(filename = paste0( "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/seed_number_tsne/sce_E7.5_no_doublets_tsne_seed_number_", i, ".pdf"), width = 10, height = 7)
# }

set.seed(3)
sce_no_doublet <- sce_no_doublet %>% runTSNE(dimred = "PCA_clust")
sce_no_doublet %>% plotTSNE(colour_by = "scDblFinder.score")

sce_no_doublet <- sce_no_doublet %>% runUMAP(dimred = "PCA_clust")
sce_no_doublet %>% plotUMAP(colour_by = "scDblFinder.score")

cowplot::plot_grid(sce_no_doublet %>% plotTSNE(colour_by = "scDblFinder.score"),
                   sce_no_doublet %>% plotUMAP(colour_by = "scDblFinder.score")
                   )

cowplot::plot_grid(sce_no_doublet %>% plotTSNE(colour_by = I(as.character(sce_no_doublet$scDblFinder.cluster))),
                   sce_no_doublet %>% plotUMAP(colour_by = I(as.character(sce_no_doublet$scDblFinder.cluster)))
                   )

# sce_no_doublet %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_7.5_reboot_no_histone_no_doublet_2022_10_17.qs", nthreads = 50)
# sce_no_doublet %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_8.5_reboot_no_histone_no_doublet_2022_10_17.qs", nthreads = 50)
sce_no_doublet %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_9.5_reboot_no_histone_no_doublet_2022_10_17.qs", nthreads = 50)
```

```{r}
sce_no_doublet <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_7.5_reboot_no_histone_no_doublet_2022_10_17.qs", nthreads = 50)
# sce_no_doublet <- qs::qread(file = "/run/user/1002/gvfs/smb-share:server=primus.img.cas.cz,share=data/83_BIOINFORMATICS/Kadu/sce_7.5_reboot_no_histone_no_doublet_2022_10_17.qs", nthreads = 50)

sce_no_doublet <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_8.5_reboot_no_histone_no_doublet_2022_10_17.qs", nthreads = 50)
# sce_no_doublet <- qs::qread(file = "/run/user/1002/gvfs/smb-share:server=primus.img.cas.cz,share=data/83_BIOINFORMATICS/Kadu/sce_8.5_reboot_no_histone_no_doublet_2022_10_17.qs", nthreads = 50)

sce_no_doublet <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_9.5_reboot_no_histone_no_doublet_2022_10_17.qs", nthreads = 50)
# sce_no_doublet <- qs::qread(file = "/run/user/1002/gvfs/smb-share:server=primus.img.cas.cz,share=data/83_BIOINFORMATICS/Kadu/sce_9.5_reboot_no_histone_no_doublet_2022_10_17.qs", nthreads = 50)

# gene_tsne_umap_plots <- function(gene, sce){
#   
#   tsne_df <- sce@int_colData@listData$reducedDims$TSNE %>% as.data.frame()
#   umap_df <- sce@int_colData@listData$reducedDims$UMAP %>% as.data.frame()
#   gene_expression <- sce@assays@data$logcounts[gene,] %>% as.numeric()
# 
#   gene_plots <- list(
#     tsne_plot = tsne_df %>% ggplot(aes(x = V1, y = V2, color = gene_expression)) +
#       geom_point() +
#       scale_color_continuous(low = "grey", high = "red", guide = "colourbar", aesthetics = "colour") +
#       theme_bw() + 
#       labs(x = "TSNE 1", y = "TSNE 2", title = gene),
#   
#   umap_plot = umap_df %>% ggplot(aes(x = V1, y = V2, color = gene_expression)) +
#     geom_point() +
#     scale_color_continuous(low = "grey", high = "red", guide = "colourbar", aesthetics = "colour") +
#     theme_bw() + 
#     labs(x = "TSNE 1", y = "TSNE 2", title = gene)) %>%
#   purrr::set_names("tsne", "umap")
# 
#   return(gene_plots)
# }

gene_tsne_umap_plots <- function(gene, sce){
  
  tsne_df <- sce@int_colData@listData$reducedDims$TSNE %>% 
    as.data.frame() %>% 
    mutate(gene_expression = sce@assays@data$logcounts[gene,] %>% as.numeric()) %>% 
    arrange(gene_expression)
  
  umap_df <- sce@int_colData@listData$reducedDims$UMAP %>% 
    as.data.frame() %>% 
    mutate(gene_expression = sce@assays@data$logcounts[gene,] %>% as.numeric()) %>% 
    arrange(gene_expression)

gene_plots <- list(
  # tsne_plot = tsne_df %>% ggplot(aes(x = V1, y = V2, color = gene_expression, shape = batch)) +
  tsne_plot = tsne_df %>% ggplot(aes(x = V1, y = V2, color = gene_expression)) +
    geom_point() +
    scale_color_continuous(low = "grey", high = "red", guide = "colourbar", aesthetics = "colour") +
    theme_bw() + 
    labs(x = "TSNE 1", y = "TSNE 2", title = gene),
  
  umap_plot = umap_df %>% ggplot(aes(x = V1, y = V2, color = gene_expression)) +
    geom_point() +
    scale_color_continuous(low = "grey", high = "red", guide = "colourbar", aesthetics = "colour") +
    theme_bw() + 
    labs(x = "UMAP 1", y = "UMAP 2", title = gene) +
    easy_plot_title_size(size = 20)
) %>% 
  purrr::set_names("tsne", "umap")
return(gene_plots)
}

genes <- c("Pou5f1", "Pf4", "Runx1", "Myb","Lyve1", "Gata1", "Epor", "Hbb-y", "Pecam1", "Cdh5", "Cldn5", "Tek", "Tlr2", "Irf8", "Csf1r", "Fcgr3", "Mme", "Fst", "Cited1", "Kdelr3", "Hand1", "Irx3", "Pax6", "Gata4")

genes <- c("Runx1", "Myb", "Gata1", "Hbb-y", "Lyve1", "Cdh5", "Cldn5", "Fst", "Kdelr3", "Epor", "Hbb-bh1")

genes <- c("Runx1", "Myb", "Gata1", "Hbb-bh1", "Mesp1", "Pim2", "Slc7a3", "Hand1", "Mixl1", "Fst", "Kdelr3", "Gata4")

genes <- c("Ddi2")

genes <- c("Kdr", "Aplnr")

genes <- genes %>% intersect(rownames(sce_no_doublet)) %>% sort()

plots_list <- purrr::map(seq_along(genes), function(i){
  print(i)
 gene_tsne_umap_plots(gene = genes[i], sce = sce_no_doublet)
 }) %>% purrr::set_names(genes)

plots_list$Aplnr$tsne %>% ggsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/tsne_plots/Aplnr_E7.5_tsne_plot.png", device = "png", width = 10, height = 7, dpi = 900)

plots_list$Kdr$tsne %>% ggsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/tsne_plots/Kdr_E7.5_tsne_plot.png", device = "png", width = 10, height = 7, dpi = 900)

plots_list$Ddi2$tsne %>% ggsave(file = "/run/user/1002/gvfs/smb-share:server=primus.img.cas.cz,share=data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/tsne_plots/DDi2_E7.5_tsne_plot.png", device = "png", width = 10, height = 7, dpi = 900)

plots_list$Ddi2$tsne %>% ggsave(file = "/run/user/1002/gvfs/smb-share:server=primus.img.cas.cz,share=data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/tsne_plots/DDi2_E8.5_tsne_plot.png", device = "png", width = 10, height = 7, dpi = 900)

plots_list$Ddi2$tsne %>% ggsave(file = "/run/user/1002/gvfs/smb-share:server=primus.img.cas.cz,share=data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/tsne_plots/DDi2_E9.5_tsne_plot.png", device = "png", width = 10, height = 7, dpi = 900)
```

# Obs: For now, we skip annotation
# singleR cluster annotation - E7.5
# ```{r}
# library(MouseGastrulationData)
# library(SingleR)
# 
# sce_no_doublet <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_7.5_reboot_no_histone_no_doublet_2022_10_17.qs", nthreads = 50)
# # sce_no_doublet <- qs::qread(file = "/run/user/1002/gvfs/smb-share:server=primus.img.cas.cz,share=data/83_BIOINFORMATICS/Kadu/sce_7.5_reboot_no_histone_no_doublet_2022_10_17.qs", nthreads = 5)
# 
# 
# as_metadata <- AtlasSampleMetadata %>% dplyr::filter(stage == "E7.5")
# sce_gas <- EmbryoAtlasData(as_metadata$sample, type = "processed")
# sce_gas <- sce_gas %>% scater::logNormCounts()
# rownames(sce_gas) <- rowData(sce_gas)[, 2] %>% make.unique()
# 
# singlets <- which(!(colData(sce_gas)$doublet | colData(sce_gas)$stripped))
# plot(
#     x = reducedDim(sce_gas, "umap")[singlets, 1],
#     y = reducedDim(sce_gas, "umap")[singlets, 2],
#     col = EmbryoCelltypeColours[colData(sce_gas)$celltype[singlets]],
#     pch = 19,
#     xaxt = "n", yaxt = "n",
#     xlab = "UMAP1", ylab = "UMAP2"
# )
# 
# library(BiocParallel)
# # from https://github.com/LTLA/SingleR/issues/62
# setAutoBPPARAM(BiocParallel::SerialParam())
# BiocParallel::register(BiocParallel::SerialParam())
# # bpparam <- BiocParallel::MulticoreParam(workers = 50)
# pred <- SingleR(test = sce_no_doublet,
#                      ref = sce_gas[, singlets],
#                      labels = colData(sce_gas)$celltype[singlets]
#                      # assay.type.test = 1
#                      # BPPARAM = BiocParallel::SerialParam()
#                      )
# # pred %>% qs::qsave(file = "~/PRIMUS/home/trufenc/singleR_predictions_sce_reboot_no_histone_no_doublets_2022_10_18.qs", nthreads = 20)
# # pred <- qs::qread(file = "~/PRIMUS/home/trufenc/singleR_predictions_sce_reboot_no_histone_no_doublets_2022_10_18.qs", nthreads = 20)
# 
# pred %>% plotDeltaDistribution(ncol = 4)
# 
# 
# 
# 
# # from https://www.nature.com/articles/s41467-022-28803-w
# # from https://github.com/IanevskiAleksandr/sc-type
# library(HGNChelper)
# source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/gene_sets_prepare.R"); source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/sctype_score_.R")
# 
# # get cell-type-specific gene sets from our in-built database (DB)
# gs_list = gene_sets_prepare(path_to_db_file = "https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/ScTypeDB_short.xlsx", cell_type =  "Immune system")
# scRNAseqData <- readRDS(gzcon(url('https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/exampleData.RDS')))
# es.max <- sctype_score(scRNAseqData = scRNAseqData, scaled = TRUE, gs = gs_list$gs_positive, gs2 = gs_list$gs_negative)
# ```

# Clustering
```{r}
# sce_no_doublet <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_7.5_reboot_no_histone_no_doublet_2022_10_17.qs", nthreads = 50)
sce_no_doublet <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_8.5_reboot_no_histone_no_doublet_2022_10_17.qs", nthreads = 50)
# sce_no_doublet <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_9.5_reboot_no_histone_no_doublet_2022_10_17.qs", nthreads = 50)

kgraph_centers <- seq(from = 100, to = 1500, by = 100)

kgraph_clusters_list <- purrr::map(kgraph_centers, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sce_no_doublet, "PCA_clust"),
    TwoStepParam(
      first = KmeansParam(centers = i),
      second = NNGraphParam(k = 5)
    )
  )
})

combined_kgraph_1 <- kgraph_clusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_centers))

clustree(combined_kgraph_1, prefix = "kgraph_", edge_arrow = FALSE)

kgraph_k <- seq(from = 5, to = 100, by = 5)

kgraph_clusters_list <- purrr::map(kgraph_k, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sce_no_doublet, "PCA_clust"),
    TwoStepParam(
      first = KmeansParam(centers = 300),
      second = NNGraphParam(k = i)
    )
  )
})

combined_kgraph_2 <- kgraph_clusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_k))

clustree(combined_kgraph_2, prefix = "kgraph_", edge_arrow = FALSE)

cowplot::plot_grid(
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[1]])),
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[2]])),
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[3]])),
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[4]])),
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[5]])),
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[6]])),
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[7]])),
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[8]])),
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[9]])),
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[10]])),
  ncol = 5
)

cowplot::plot_grid(
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[11]])),
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[12]])),
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[13]])),
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[14]])),
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[15]])),
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[16]])),
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[17]])),
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[18]])),
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[19]])),
  sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[20]])),
  ncol = 5
)


```




# Clustering for E7.5
```{r}
table(combined_kgraph_2[6])

sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[6]]))
sce_no_doublet %>% plotUMAP(colour_by = I(combined_kgraph_2[[6]]))

cowplot::plot_grid(sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[6]])), 
                   sce_no_doublet %>% plotUMAP(colour_by = I(combined_kgraph_2[[6]])))

cluster_E7.5 <- combined_kgraph_2[[6]] %>% 
  forcats::fct_recode("E7.5_1" = "1", 
                      "E7.5_2" = "2",
                      "E7.5_3" = "3"
                      # "E7.5_4" = "4",
                      # "E7.5_5" = "5",
                      # "E7.5_6" = "6",
                      # "E7.5_7" = "7",
                      # "E7.5_8" = "8",
                      # "E7.5_9" = "9"
                      # "E7.5_10" = "10"
                      )

colLabels(sce_no_doublet) <- cluster_E7.5


# Silhouete
distance_matrix <- dist(reducedDim(sce_no_doublet, "PCA_clust"))
sil <- silhouette(as.numeric(colLabels(sce_no_doublet)), dist = distance_matrix)
sil %>% 
  fviz_silhouette() +
  scale_colour_manual(values = tableau_10_medium, labels = c("E7.5_1", "E7.5_2", "E7.5_3")) +
  #, "E7.5_4" , "E7.5_5", "E7.5_6", "E7.5_7", "E7.5_8", "E7.5_9"
  scale_fill_manual(values = tableau_10_medium, labels = c("E7.5_1", "E7.5_2", "E7.5_3"))

# Subclustering E7.5_1
### Sub-clustering cluster 8
sce_cluster_1 <- sce_no_doublet[, which(colLabels(sce_no_doublet) == "E7.5_1")]

dec_cluster_1 <- sce_cluster_1 %>% modelGeneVar()
hvg_var_cluster_1 <- dec_cluster_1 %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames()
sce_cluster_1 <- sce_cluster_1 %>% scater::runPCA(ncomponents = 50, subset_row = hvg_var_cluster_1)

pcs_cluster_1 <- sce_cluster_1 %>% reducedDim()
choices_cluster_1 <- pcs_cluster_1 %>% getClusteredPCs()
val_cluster_1 <- metadata(choices_cluster_1)$chosen
reducedDim(sce_cluster_1, "PCA_subclust_1") <- pcs_cluster_1[, 1:val_cluster_1]
plotReducedDim(sce_cluster_1, dimred = "PCA_subclust_1", ncomponents = 4)

kgraph_centers <- seq(from = 10, to = 100, by = 10)
# kgraph_centers <- seq(from = 5, to = 100, by = 1)

kgraph_subclusters_list <- purrr::map(kgraph_centers, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sce_cluster_1, "PCA_subclust_1"),
    TwoStepParam(
      first = KmeansParam(centers = i),
      second = NNGraphParam(k = 5)
    )
  )
})

combined_kgraph_sub1 <- kgraph_subclusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_centers))

clustree(combined_kgraph_sub1, prefix = "kgraph_", edge_arrow = FALSE)

kgraph_k <- seq(from = 5, to = 100, by = 5)

kgraph_subclusters_list <- purrr::map(kgraph_k, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sce_cluster_1, "PCA_subclust_1"),
    TwoStepParam(
      first = KmeansParam(centers = 50),
      second = NNGraphParam(k = i)
    )
  )
})

combined_kgraph_sub2 <- kgraph_subclusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_k))

clustree(combined_kgraph_sub2, prefix = "kgraph_", edge_arrow = FALSE)

cowplot::plot_grid(
  sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[1]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[2]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[3]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[4]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[5]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[6]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[7]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[8]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[9]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[10]])),
  ncol = 5
)

table(kgraph_subclusters_list[[]])

sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[3]]))
sce_cluster_1 %>% plotUMAP(colour_by = I(kgraph_subclusters_list[[3]]))

cowplot::plot_grid(sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[3]])), 
                   sce_cluster_1 %>% plotUMAP(colour_by = I(kgraph_subclusters_list[[3]])))


kgraph_subclusters_list[[3]] <- kgraph_subclusters_list[[3]] %>% forcats::fct_recode("E7.5_1.1" = "1",
                                                                                     "E7.5_1.2" = "2"
                                                                                     # "E7.5_1.3" = "3"
                                                                                     )


# update original clusters with new subclustering
new_levels <- colLabels(sce_no_doublet) %>% as.character()
# new_levels[new_levels == 1] <- kgraph_subclusters_list[[4]] %>% as.character()
new_levels[new_levels == "E7.5_1"] <- kgraph_subclusters_list[[3]] %>% as.character()

colLabels(sce_no_doublet) <- factor(new_levels)

cowplot::plot_grid(
  sce_no_doublet %>% plotTSNE(colour_by = I(colLabels(sce_no_doublet))), 
  sce_no_doublet %>% plotUMAP(colour_by = I(colLabels(sce_no_doublet)))
  )



# Subclustering E7.5_3
sce_cluster_3 <- sce_no_doublet[, which(colLabels(sce_no_doublet) == "E7.5_3")]

dec_cluster_3 <- sce_cluster_3 %>% modelGeneVar()
hvg_var_cluster_3 <- dec_cluster_3 %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames()
sce_cluster_3 <- sce_cluster_3 %>% scater::runPCA(ncomponents = 50, subset_row = hvg_var_cluster_3)

pcs_cluster_3 <- sce_cluster_3 %>% reducedDim()
choices_cluster_3 <- pcs_cluster_3 %>% getClusteredPCs()
val_cluster_3 <- metadata(choices_cluster_3)$chosen
reducedDim(sce_cluster_3, "PCA_subclust_3") <- pcs_cluster_3[, 1:val_cluster_3]
plotReducedDim(sce_cluster_3, dimred = "PCA_subclust_3", ncomponents = 4)

kgraph_centers <- seq(from = 10, to = 100, by = 10)
# kgraph_centers <- seq(from = 5, to = 100, by = 1)

kgraph_subclusters_list <- purrr::map(kgraph_centers, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sce_cluster_3, "PCA_subclust_3"),
    TwoStepParam(
      first = KmeansParam(centers = i),
      second = NNGraphParam(k = 5)
    )
  )
})

combined_kgraph_sub1 <- kgraph_subclusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_centers))

clustree(combined_kgraph_sub1, prefix = "kgraph_", edge_arrow = FALSE)

kgraph_k <- seq(from = 5, to = 100, by = 5)

kgraph_subclusters_list <- purrr::map(kgraph_k, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sce_cluster_3, "PCA_subclust_3"),
    TwoStepParam(
      first = KmeansParam(centers = 40),
      second = NNGraphParam(k = i)
    )
  )
})

combined_kgraph_sub2 <- kgraph_subclusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_k))

clustree(combined_kgraph_sub2, prefix = "kgraph_", edge_arrow = FALSE)

cowplot::plot_grid(
  sce_cluster_3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[1]])),
  sce_cluster_3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[2]])),
  sce_cluster_3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[3]])),
  sce_cluster_3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[4]])),
  sce_cluster_3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[5]])),
  sce_cluster_3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[6]])),
  sce_cluster_3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[7]])),
  sce_cluster_3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[8]])),
  sce_cluster_3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[9]])),
  sce_cluster_3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[10]])),
  ncol = 5
)

table(kgraph_subclusters_list[[1]])

sce_cluster_3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[1]]))
sce_cluster_3 %>% plotUMAP(colour_by = I(kgraph_subclusters_list[[1]]))

cowplot::plot_grid(sce_cluster_3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[1]])), 
                   sce_cluster_3 %>% plotUMAP(colour_by = I(kgraph_subclusters_list[[1]])))


kgraph_subclusters_list[[1]] <- kgraph_subclusters_list[[1]] %>% forcats::fct_recode("E7.5_3.1" = "1",
                                                                                     "E7.5_3.2" = "2",
                                                                                     "E7.5_3.3" = "3",
                                                                                     "E7.5_3.4" = "4"
                                                                                     )


# update original clusters with new subclustering
new_levels <- colLabels(sce_no_doublet) %>% as.character()
# new_levels[new_levels == 1] <- kgraph_subclusters_list[[4]] %>% as.character()
new_levels[new_levels == "E7.5_3"] <- kgraph_subclusters_list[[1]] %>% as.character()

colLabels(sce_no_doublet) <- factor(new_levels)

cowplot::plot_grid(
  sce_no_doublet %>% plotTSNE(colour_by = I(colLabels(sce_no_doublet))), 
  sce_no_doublet %>% plotUMAP(colour_by = I(colLabels(sce_no_doublet)))
  )





# Subclustering E7.5_3.3
sce_cluster_3.3 <- sce_no_doublet[, which(colLabels(sce_no_doublet) == "E7.5_3.3")]

dec_cluster_3.3 <- sce_cluster_3.3 %>% modelGeneVar()
hvg_var_cluster_3.3 <- dec_cluster_3.3 %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames()
sce_cluster_3.3 <- sce_cluster_3.3 %>% scater::runPCA(ncomponents = 50, subset_row = hvg_var_cluster_3.3)

pcs_cluster_3.3 <- sce_cluster_3.3 %>% reducedDim()
choices_cluster_3.3 <- pcs_cluster_3.3 %>% getClusteredPCs()
val_cluster_3.3 <- metadata(choices_cluster_3.3)$chosen
reducedDim(sce_cluster_3.3, "PCA_subclust_3.3") <- pcs_cluster_3.3[, 1:val_cluster_3.3]
plotReducedDim(sce_cluster_3.3, dimred = "PCA_subclust_3.3", ncomponents = 4)

kgraph_centers <- seq(from = 10, to = 100, by = 10)
# kgraph_centers <- seq(from = 5, to = 100, by = 1)

kgraph_subclusters_list <- purrr::map(kgraph_centers, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sce_cluster_3.3, "PCA_subclust_3.3"),
    TwoStepParam(
      first = KmeansParam(centers = i),
      second = NNGraphParam(k = 5)
    )
  )
})

combined_kgraph_sub1 <- kgraph_subclusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_centers))

clustree(combined_kgraph_sub1, prefix = "kgraph_", edge_arrow = FALSE)

kgraph_k <- seq(from = 5, to = 100, by = 5)

kgraph_subclusters_list <- purrr::map(kgraph_k, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sce_cluster_3.3, "PCA_subclust_3.3"),
    TwoStepParam(
      first = KmeansParam(centers = 20),
      second = NNGraphParam(k = i)
    )
  )
})

combined_kgraph_sub2 <- kgraph_subclusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_k))

clustree(combined_kgraph_sub2, prefix = "kgraph_", edge_arrow = FALSE)

cowplot::plot_grid(
  sce_cluster_3.3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[1]])),
  sce_cluster_3.3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[2]])),
  sce_cluster_3.3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[3]])),
  sce_cluster_3.3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[4]])),
  sce_cluster_3.3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[5]])),
  sce_cluster_3.3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[6]])),
  sce_cluster_3.3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[7]])),
  sce_cluster_3.3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[8]])),
  sce_cluster_3.3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[9]])),
  sce_cluster_3.3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[10]])),
  ncol = 5
)

table(kgraph_subclusters_list[[2]])

sce_cluster_3.3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[2]]))
sce_cluster_3.3 %>% plotUMAP(colour_by = I(kgraph_subclusters_list[[2]]))

cowplot::plot_grid(sce_cluster_3.3 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[2]])), 
                   sce_cluster_3.3 %>% plotUMAP(colour_by = I(kgraph_subclusters_list[[2]])))


kgraph_subclusters_list[[2]] <- kgraph_subclusters_list[[2]] %>% forcats::fct_recode("E7.5_3.3.1" = "1",
                                                                                     "E7.5_3.3.2" = "2"
                                                                                     # "E7.5_3.3.3" = "3",
                                                                                     # "E7.5_3.3.4" = "4"
                                                                                     )


# update original clusters with new subclustering
new_levels <- colLabels(sce_no_doublet) %>% as.character()
# new_levels[new_levels == 1] <- kgraph_subclusters_list[[4]] %>% as.character()
new_levels[new_levels == "E7.5_3.3"] <- kgraph_subclusters_list[[2]] %>% as.character()

colLabels(sce_no_doublet) <- factor(new_levels)

cowplot::plot_grid(
  sce_no_doublet %>% plotTSNE(colour_by = I(colLabels(sce_no_doublet))), 
  sce_no_doublet %>% plotUMAP(colour_by = I(colLabels(sce_no_doublet)))
  )



sce_no_doublet %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_7.5_reboot_no_histone_no_doublet_clustered_2022_10_20.qs", nthreads = 50)
```

# Clustering for E8.5
```{r}
table(combined_kgraph_2[1])

sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[1]]))
sce_no_doublet %>% plotUMAP(colour_by = I(combined_kgraph_2[[1]]))

cowplot::plot_grid(sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[1]])), 
                   sce_no_doublet %>% plotUMAP(colour_by = I(combined_kgraph_2[[1]])))

cluster_E8.5 <- combined_kgraph_2[[1]] %>% 
  forcats::fct_recode("E8.5_1" = "1", 
                      "E8.5_2" = "2",
                      "E8.5_3" = "3",
                      "E8.5_4" = "4",
                      "E8.5_5" = "5",
                      "E8.5_6" = "6",
                      "E8.5_7" = "7",
                      "E8.5_8" = "8",
                      "E8.5_9" = "9",
                      "E8.5_10" = "10",
                      "E8.5_11" = "11",
                      "E8.5_12" = "12"
                      )

colLabels(sce_no_doublet) <- cluster_E8.5


# Silhouete
distance_matrix <- dist(reducedDim(sce_no_doublet, "PCA_clust"))
sil <- silhouette(as.numeric(colLabels(sce_no_doublet)), dist = distance_matrix)
sil %>% 
  fviz_silhouette() +
  scale_colour_manual(values = tableau_10_medium, labels = c("E8.5_1", "E8.5_2", "E8.5_3", "E8.5_4", "E8.5_5", "E8.5_6", "E8.5_7", "E8.5_8", "E8.5_9", "E8.5_10", "E8.5_11", "E8.5_12")) +
  # 
  scale_fill_manual(values = tableau_10_medium, labels = c("E8.5_1", "E8.5_2", "E8.5_3", "E8.5_4", "E8.5_5", "E8.5_6", "E8.5_7", "E8.5_8", "E8.5_9", "E8.5_10", "E8.5_11", "E8.5_12"))

# # Subclustering E8.5_1
# ### Sub-clustering cluster 8
# sce_cluster_1 <- sce_no_doublet[, which(colLabels(sce_no_doublet) == "E8.5_1")]
# 
# dec_cluster_1 <- sce_cluster_1 %>% modelGeneVar()
# hvg_var_cluster_1 <- dec_cluster_1 %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames()
# sce_cluster_1 <- sce_cluster_1 %>% scater::runPCA(ncomponents = 50, subset_row = hvg_var_cluster_1)
# 
# pcs_cluster_1 <- sce_cluster_1 %>% reducedDim()
# choices_cluster_1 <- pcs_cluster_1 %>% getClusteredPCs()
# val_cluster_1 <- metadata(choices_cluster_1)$chosen
# reducedDim(sce_cluster_1, "PCA_subclust_1") <- pcs_cluster_1[, 1:val_cluster_1]
# plotReducedDim(sce_cluster_1, dimred = "PCA_subclust_1", ncomponents = 4)
# 
# kgraph_centers <- seq(from = 10, to = 100, by = 10)
# # kgraph_centers <- seq(from = 5, to = 100, by = 1)
# 
# kgraph_subclusters_list <- purrr::map(kgraph_centers, function(i) {
#   set.seed(123)
#   clusterRows(
#     reducedDim(sce_cluster_1, "PCA_subclust_1"),
#     TwoStepParam(
#       first = KmeansParam(centers = i),
#       second = NNGraphParam(k = 5)
#     )
#   )
# })
# 
# combined_kgraph_sub1 <- kgraph_subclusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_centers))
# 
# clustree(combined_kgraph_sub1, prefix = "kgraph_", edge_arrow = FALSE)
# 
# kgraph_k <- seq(from = 5, to = 100, by = 5)
# 
# kgraph_subclusters_list <- purrr::map(kgraph_k, function(i) {
#   set.seed(123)
#   clusterRows(
#     reducedDim(sce_cluster_1, "PCA_subclust_1"),
#     TwoStepParam(
#       first = KmeansParam(centers = 50),
#       second = NNGraphParam(k = i)
#     )
#   )
# })
# 
# combined_kgraph_sub2 <- kgraph_subclusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_k))
# 
# clustree(combined_kgraph_sub2, prefix = "kgraph_", edge_arrow = FALSE)
# 
# cowplot::plot_grid(
#   sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[1]])),
#   sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[2]])),
#   sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[3]])),
#   sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[4]])),
#   sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[5]])),
#   sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[6]])),
#   sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[7]])),
#   sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[8]])),
#   sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[9]])),
#   sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[10]])),
#   ncol = 5
# )
# 
# table(kgraph_subclusters_list[[2]])
# 
# sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[1]]))
# sce_cluster_1 %>% plotUMAP(colour_by = I(kgraph_subclusters_list[[1]]))
# 
# cowplot::plot_grid(sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[1]])), 
#                    sce_cluster_1 %>% plotUMAP(colour_by = I(kgraph_subclusters_list[[1]])))
# 
# 
# kgraph_subclusters_list[[1]] <- kgraph_subclusters_list[[1]] %>% forcats::fct_recode("E8.5_1.1" = "1",
#                                                                                      "E8.5_1.2" = "2"
#                                                                                      # "E8.5_1.3" = "3"
#                                                                                      )
# 
# 
# # update original clusters with new subclustering
# new_levels <- colLabels(sce_no_doublet) %>% as.character()
# # new_levels[new_levels == 1] <- kgraph_subclusters_list[[4]] %>% as.character()
# new_levels[new_levels == "E8.5_1"] <- kgraph_subclusters_list[[1]] %>% as.character()
# 
# colLabels(sce_no_doublet) <- factor(new_levels)
# 
# cowplot::plot_grid(sce_no_doublet %>% plotTSNE(colour_by = I(colLabels(sce_no_doublet))), sce_no_doublet %>% plotUMAP(colour_by = I(colLabels(sce_no_doublet))))

sce_no_doublet %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_8.5_reboot_no_histone_no_doublet_clustered_2022_10_20.qs", nthreads = 50)
```

# Clustering for E9.5
```{r}
table(combined_kgraph_2[2])

sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[2]]))
sce_no_doublet %>% plotUMAP(colour_by = I(combined_kgraph_2[[2]]))

cowplot::plot_grid(sce_no_doublet %>% plotTSNE(colour_by = I(combined_kgraph_2[[2]])), 
                   sce_no_doublet %>% plotUMAP(colour_by = I(combined_kgraph_2[[2]])))

cluster_E9.5 <- combined_kgraph_2[[2]] %>% 
  forcats::fct_recode("E9.5_1" = "1", 
                      "E9.5_2" = "2",
                      "E9.5_3" = "3",
                      "E9.5_4" = "4",
                      "E9.5_5" = "5",
                      "E9.5_6" = "6",
                      "E9.5_7" = "7",
                      "E9.5_8" = "8",
                      "E9.5_9" = "9"
                      # "E9.5_10" = "10"
                      )

colLabels(sce_no_doublet) <- cluster_E9.5


# Silhouete
distance_matrix <- dist(reducedDim(sce_no_doublet, "PCA_clust"))
sil <- silhouette(as.numeric(colLabels(sce_no_doublet)), dist = distance_matrix)
sil %>% 
  fviz_silhouette() +
  scale_colour_manual(values = tableau_10_medium, labels = c("E9.5_1", "E9.5_2", "E9.5_3", "E9.5_4", "E9.5_5", "E9.5_6", "E9.5_7", "E9.5_8", "E9.5_9")) +
  # 
  scale_fill_manual(values = tableau_10_medium, labels = c("E9.5_1", "E9.5_2", "E9.5_3", "E9.5_4", "E9.5_5", "E9.5_6", "E9.5_7", "E9.5_8", "E9.5_9"))

# Subclustering E9.5_8
### Sub-clustering cluster 8
sce_cluster_8 <- sce_no_doublet[, which(colLabels(sce_no_doublet) == "E9.5_8")]

dec_cluster_8 <- sce_cluster_8 %>% modelGeneVar()
hvg_var_cluster_8 <- dec_cluster_8 %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames()
sce_cluster_8 <- sce_cluster_8 %>% scater::runPCA(ncomponents = 50, subset_row = hvg_var_cluster_8)

pcs_cluster_8 <- sce_cluster_8 %>% reducedDim()
choices_cluster_8 <- pcs_cluster_8 %>% getClusteredPCs()
val_cluster_8 <- metadata(choices_cluster_8)$chosen
reducedDim(sce_cluster_8, "PCA_subclust_8") <- pcs_cluster_8[, 1:val_cluster_8]
plotReducedDim(sce_cluster_8, dimred = "PCA_subclust_8", ncomponents = 4)

kgraph_centers <- seq(from = 10, to = 100, by = 10)
# kgraph_centers <- seq(from = 5, to = 100, by = 1)

kgraph_subclusters_list <- purrr::map(kgraph_centers, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sce_cluster_8, "PCA_subclust_8"),
    TwoStepParam(
      first = KmeansParam(centers = i),
      second = NNGraphParam(k = 5)
    )
  )
})

combined_kgraph_sub1 <- kgraph_subclusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_centers))

clustree(combined_kgraph_sub1, prefix = "kgraph_", edge_arrow = FALSE)

kgraph_k <- seq(from = 5, to = 100, by = 5)

kgraph_subclusters_list <- purrr::map(kgraph_k, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sce_cluster_8, "PCA_subclust_8"),
    TwoStepParam(
      first = KmeansParam(centers = 20),
      second = NNGraphParam(k = i)
    )
  )
})

combined_kgraph_sub2 <- kgraph_subclusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_k))

clustree(combined_kgraph_sub2, prefix = "kgraph_", edge_arrow = FALSE)

table(kgraph_subclusters_list[[1]])

sce_cluster_8 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[1]]))
sce_cluster_8 %>% plotUMAP(colour_by = I(kgraph_subclusters_list[[1]]))

cowplot::plot_grid(sce_cluster_8 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[1]])), 
                   sce_cluster_8 %>% plotUMAP(colour_by = I(kgraph_subclusters_list[[1]])))


kgraph_subclusters_list[[1]] <- kgraph_subclusters_list[[1]] %>% forcats::fct_recode("E9.5_8.1" = "1",
                                                                                     "E9.5_8.2" = "2"
                                                                                     # "E9.5_8.3" = "3"
                                                                                     )


# update original clusters with new subclustering
new_levels <- colLabels(sce_no_doublet) %>% as.character()
# new_levels[new_levels == 1] <- kgraph_subclusters_list[[4]] %>% as.character()
new_levels[new_levels == "E9.5_8"] <- kgraph_subclusters_list[[1]] %>% as.character()

colLabels(sce_no_doublet) <- factor(new_levels)

cowplot::plot_grid(sce_no_doublet %>% plotTSNE(colour_by = I(colLabels(sce_no_doublet))), sce_no_doublet %>% plotUMAP(colour_by = I(colLabels(sce_no_doublet))))

sce_no_doublet %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_9.5_reboot_no_histone_no_doublet_clustered_2022_10_20.qs", nthreads = 50)
```


# scmap cluster annotation - E7.5
```{r}
library(MouseGastrulationData)
library(scmap)

# sce_no_doublet <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_7.5_reboot_no_histone_no_doublet_clustered_2022_10_20.qs", nthreads = 50)
# sce_no_doublet <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_8.5_reboot_no_histone_no_doublet_clustered_2022_10_20.qs", nthreads = 50)
sce_no_doublet <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_9.5_reboot_no_histone_no_doublet_clustered_2022_10_20.qs", nthreads = 50)

rowData(sce_no_doublet) <- rowData(sce_no_doublet) %>% 
  as.data.frame() %>% 
  dplyr::rename(feature_symbol = value) %>% 
  as("DFrame")

# as_metadata <- AtlasSampleMetadata %>% dplyr::filter(stage == "E7.5")
as_metadata <- AtlasSampleMetadata %>% dplyr::filter(stage == "E8.5")
# as_metadata <- AtlasSampleMetadata
sce_gas <- EmbryoAtlasData(as_metadata$sample, type = "processed")
sce_gas <- sce_gas %>% scater::logNormCounts()
rownames(sce_gas) <- rowData(sce_gas)[, 2] %>% make.unique()

singlets <- which(!(colData(sce_gas)$doublet | colData(sce_gas)$stripped))
plot(
    x = reducedDim(sce_gas, "umap")[singlets, 1],
    y = reducedDim(sce_gas, "umap")[singlets, 2],
    col = EmbryoCelltypeColours[colData(sce_gas)$celltype[singlets]],
    pch = 19,
    xaxt = "n", yaxt = "n",
    xlab = "UMAP1", ylab = "UMAP2"
)

set.seed(123)
# colLabels(sce_gas) <- paste(sce_gas$cell, sce_gas$celltype, sep = "_")
rowData(sce_gas) <- rowData(sce_gas) %>% as.data.frame() %>% dplyr::rename(feature_symbol = SYMBOL) %>% as("DFrame")
sce_gas <- sce_gas %>% selectFeatures(suppress_plot = FALSE)
# sce_gas <- sce_gas %>% indexCell()
sce_gas <- sce_gas %>% indexCluster(cluster_col = "celltype")

heatmap(as.matrix(metadata(sce_gas)$scmap_cluster_index))

scmapCluster_results <- scmapCluster(
  projection = sce_no_doublet, 
  index_list = list(sce_gas = metadata(sce_gas)$scmap_cluster_index)
)

sce_no_doublet$annotation <- scmapCluster_results$scmap_cluster_labs

sce_no_doublet %>%
  plotTSNE(colour_by = "annotation") +
  ggeasy::easy_all_text_size(size = 20)

# sce_no_doublet %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_7.5_reboot_no_histone_no_doublet_clustered_annotated_2022_10_20.qs", nthreads = 50)
# sce_no_doublet %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_8.5_reboot_no_histone_no_doublet_clustered_annotated_2022_10_20.qs", nthreads = 50)
# sce_no_doublet %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_9.5_reboot_no_histone_no_doublet_clustered_annotated_2022_10_20.qs", nthreads = 50)
```


# DE analysis on the embryonic datasets
```{r}
# sce_E7.5 <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_7.5_reboot_no_histone_no_doublet_clustered_annotated_2022_10_20.qs", nthreads = 50)
# sce_E8.5 <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_8.5_reboot_no_histone_no_doublet_clustered_annotated_2022_10_20.qs", nthreads = 50)
sce_E9.5 <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_9.5_reboot_no_histone_no_doublet_clustered_annotated_2022_10_20.qs", nthreads = 50)

# markers <- sce_E7.5 %>% findMarkers(test = "binom", direction = "up", lfc = 1, pval.type = "any")
# markers <- sce_E8.5 %>% findMarkers(test = "binom", direction = "up", lfc = 1, pval.type = "any")
markers <- sce_E9.5 %>% findMarkers(test = "binom", direction = "up", lfc = 1, pval.type = "any")
# markers <- sub_mnn_out %>% findMarkers(test = "binom", direction = "up", lfc = 1, pval.type = "any")

markers_list <- purrr::map(seq_along(markers), function(i) markers[[i]])
# markers_list <- markers_list %>% purrr::set_names(sort(unique(sce_E7.5$label)))
# markers_list <- markers_list %>% purrr::set_names(sort(unique(sce_E8.5$label)))
markers_list <- markers_list %>% purrr::set_names(sort(unique(sce_E9.5$label)))
# markers_list <- markers_list %>% purrr::set_names(sort(unique(sub_mnn_out$label)))
DE_list <- purrr::map(seq_along(markers), function(i) markers_list[[i]] %>% as.data.frame() %>% dplyr::filter(FDR <= 0.05 & summary.logFC >= 1.0 & Top <= 10)) %>% 
  # purrr::set_names(sce_E7.5$label %>% unique() %>% sort())
  # purrr::set_names(sce_E8.5$label %>% unique() %>% sort())
  purrr::set_names(sce_E9.5$label %>% unique() %>% sort())

DE_list <- DE_list %>%
  purrr::map(mutate_at, .vars = c("p.value", "FDR"), .funs = formatC, digits = 2, format = "e") %>% 
  purrr::map(mutate_if, .predicate = is.numeric, .funs = round, digits = 2) 

DE_list <- DE_list %>%
  purrr::map(rownames_to_column, var = "gene")

# DE_df <- DE_list %>%
#   bind_rows(.id = "reference_cluster") %>% 
#   vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/DE_results_sce_E9.5.tsv", delim = "\t")

DE_df <- DE_list %>%
  bind_rows(.id = "reference_cluster") %>%
  vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/DE_results_sce_E8.5.tsv", delim = "\t")

# DE_df <- DE_list %>%
#   bind_rows(.id = "reference_cluster") %>% 
#   vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/DE_results_sce_E7.5.tsv", delim = "\t")


DEG_list <- DE_list %>% 
  # purrr::map(rownames_to_column, var = "gene") %>% 
  purrr::map(pull, gene) %>% 
  purrr::map(str_to_upper) %>% 
  purrr::map(sort) %>% 
  # purrr::set_names(sub_mnn_out$label %>% unique() %>% sort())
  purrr::set_names(sce_E9.5$label %>% unique() %>% sort())


# Based on the largest metrics
# hvg_E7.5 <- sce_E7.5 %>% modelGeneVar() %>% metadata() %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames() %>% unique() %>% sort()



for(i in 1:length(DE_list)) {
  for(j in 1:length(rownames(DE_list[[i]]))) {
    violin_plot <- sce_E7.5 %>% 
      plotExpression(x = "label", features = rownames(DE_list[[i]])[j],  colour_by = "label") +
      theme_bw() +
      theme(axis.text.x = element_blank()) +
      ggeasy::easy_all_text_size(size = 20) +
      guides(color = guide_legend("Cluster")) +
      labs(x = "", y = "Normalized gene expression") +
      ggeasy::easy_legend_at(to = "bottom") 
    
    violin_plot %>% ggsave(filename = paste0( "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/violin_plots/", names(DE_list)[i], "_", rownames(DE_list[[i]])[j],"_ violin_plot.png"), width = 10, height = 5)  
  }
}

for(i in 1:length(DE_list)) {
  for(j in 1:length(rownames(DE_list[[i]]))) {
    violin_plot <- sce_E8.5 %>% 
      plotExpression(x = "label", features = rownames(DE_list[[i]])[j],  colour_by = "label") +
      theme_bw() +
      theme(axis.text.x = element_blank()) +
      ggeasy::easy_all_text_size(size = 20) +
      guides(color = guide_legend("Cluster")) +
      labs(x = "", y = "Normalized gene expression") +
      ggeasy::easy_legend_at(to = "bottom") 
    
    violin_plot %>% ggsave(filename = paste0( "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/violin_plots/", names(DE_list)[i], "_", rownames(DE_list[[i]])[j],"_ violin_plot.png"), width = 10, height = 5)  
  }
}

for(i in 1:length(DE_list)) {
  for(j in 1:length(rownames(DE_list[[i]]))) {
    violin_plot <- sce_E9.5 %>% 
      plotExpression(x = "label", features = rownames(DE_list[[i]])[j],  colour_by = "label") +
      theme_bw() +
      theme(axis.text.x = element_blank()) +
      ggeasy::easy_all_text_size(size = 20) +
      guides(color = guide_legend("Cluster")) +
      labs(x = "", y = "Normalized gene expression") +
      ggeasy::easy_legend_at(to = "bottom") 
    
    violin_plot %>% ggsave(filename = paste0( "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/violin_plots/", names(DE_list)[i], "_", rownames(DE_list[[i]])[j],"_ violin_plot.png"), width = 10, height = 5)  
  }
}
```

# Merge datasets
```{r}
# sce_E7.5 <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_7.5_reboot_no_histone_no_doublet_clustered_2022_10_20.qs", nthreads = 50)
# sce_E8.5 <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_8.5_reboot_no_histone_no_doublet_clustered_2022_10_20.qs", nthreads = 50)
# sce_E9.5 <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_9.5_reboot_no_histone_no_doublet_clustered_2022_10_20.qs", nthreads = 50)

sce_E7.5 <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_7.5_reboot_no_histone_no_doublet_clustered_annotated_2022_10_20.qs", nthreads = 50)
sce_E8.5 <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_8.5_reboot_no_histone_no_doublet_clustered_annotated_2022_10_20.qs", nthreads = 50)
sce_E9.5 <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_9.5_reboot_no_histone_no_doublet_clustered_annotated_2022_10_20.qs", nthreads = 50)


sce_E7.5$batch <- "7.5"
sce_E8.5$batch <- "8.5"
sce_E9.5$batch <- "9.5"

colData(sce_E8.5) <- colData(sce_E8.5) %>% 
  as.data.frame() %>% 
  dplyr::rename(annotation = sce_gas) %>% 
  # dplyr::select(-c(sum.1, detected.1, total.1)) %>% 
  dplyr::select(colnames(colData(sce_E7.5))) %>% 
  as(Class = "DFrame")

colData(sce_E9.5) <- colData(sce_E9.5) %>% 
  as.data.frame() %>% 
  dplyr::rename(annotation = sce_gas) %>% 
  dplyr::select(colnames(colData(sce_E7.5))) %>% 
  as(Class = "DFrame")

dec_E7.5_default <- sce_E7.5 %>% modelGeneVar()
dec_E8.5_default <- sce_E8.5 %>% modelGeneVar()
dec_E9.5_default <- sce_E9.5 %>% modelGeneVar()

fit_E7.5_default <- metadata(dec_E7.5_default)
fit_E8.5_default <- metadata(dec_E8.5_default)
fit_E9.5_default <- metadata(dec_E9.5_default)

# Based on the largest metrics
hvg_var_E7.5 <- dec_E7.5_default %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames()
hvg_var_E8.5 <- dec_E8.5_default %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames()
hvg_var_E9.5 <- dec_E9.5_default %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames()

hv_genes_E7.5 <- hvg_var_E7.5 %>% unique() %>% sort()
hv_genes_E8.5 <- hvg_var_E8.5 %>% unique() %>% sort()
hv_genes_E9.5 <- hvg_var_E9.5 %>% unique() %>% sort()

## Quick start
#  The first and most obvious is to subset all batches to the common “universe” of features.
universe <- union(hv_genes_E7.5, union(hv_genes_E8.5, hv_genes_E9.5))

set_diff_E7.5 <- setdiff(universe, rownames(sce_E7.5))
set_diff_E8.5 <- setdiff(universe, rownames(sce_E8.5))
set_diff_E9.5 <- setdiff(universe, rownames(sce_E9.5))

merged_E7.5_E8.5 <- Seurat::RowMergeSparseMatrices(
  mat1 = sce_E7.5@assays@data$counts,
  mat2 = sce_E8.5@assays@data$counts
  )

merged_all <- Seurat::RowMergeSparseMatrices(
  mat1 = merged_E7.5_E8.5,
  mat2 = sce_E9.5@assays@data$counts
  )


## Diagnosing batch effects
#  it is worth examining whether there is any batch effect in this dataset.
# Synchronizing the metadata for cbind()ing.

# build matrix with empty values 

# creating sparse data https://stackoverflow.com/questions/49758083/how-to-create-sparse-matrix-in-r
library(Matrix)

# https://stackoverflow.com/questions/43117608/r-binding-sparse-matrices-of-different-sizes-on-rows
merge_sparse <- function(...) {
  
  cnnew <- character()
  rnnew <- character()
  x <- vector()
  i <- numeric()
  j <- numeric()
  
  for (M in list(...)) {
  
  cnold <- colnames(M)
  rnold <- rownames(M)
  
  cnnew <- union(cnnew,cnold)
  rnnew <- union(rnnew,rnold)
  
  cindnew <- match(cnold,cnnew)
  rindnew <- match(rnold,rnnew)
  ind <- unname(which(M != 0,arr.ind=T))
  i <- c(i,rindnew[ind[,1]])
  j <- c(j,cindnew[ind[,2]])
  x <- c(x,M@x)
  }
  
  sparseMatrix(i=i,j=j,x=x,dims=c(length(rnnew),length(cnnew)),dimnames=list(rnnew,cnnew))
}

empty_matrix_E7.5 <- matrix(0, 
  nrow = length(setdiff(union(rownames(sce_E8.5), rownames(sce_E9.5)), rownames(sce_E7.5))), 
  ncol = ncol(sce_E7.5)
  ) %>% 
  as.data.frame() %>% 
  mutate(genes = setdiff(union(rownames(sce_E8.5), rownames(sce_E9.5)), rownames(sce_E7.5))) %>% 
  column_to_rownames(var = "genes") %>%
  purrr::set_names(colnames(sce_E7.5)) %>%
  as.matrix(.) %>%
  Matrix(., sparse = T)

merged_E7.5 <- merge_sparse(sce_E7.5@assays@data$counts, empty_matrix_E7.5)

sce_E7.5_merged <- SingleCellExperiment(assays = list(counts = merged_E7.5), colData = sce_E7.5@colData) 

sce_E7.5_merged <- sce_E7.5_merged[order(row.names(sce_E7.5_merged)), ]

sce_E7.5_merged <- sce_E7.5_merged %>% logNormCounts()

empty_matrix_E8.5 <- matrix(0, 
  nrow = length(setdiff(union(rownames(sce_E7.5), rownames(sce_E9.5)), rownames(sce_E8.5))), 
  ncol = ncol(sce_E8.5)
  ) %>% 
  as.data.frame() %>% 
  mutate(genes = setdiff(union(rownames(sce_E7.5), rownames(sce_E9.5)), rownames(sce_E8.5))) %>% 
  column_to_rownames(var = "genes") %>%
  purrr::set_names(colnames(sce_E8.5)) %>%
  as.matrix(.) %>%
  Matrix(., sparse = T)

merged_E8.5 <- merge_sparse(sce_E8.5@assays@data$counts, empty_matrix_E8.5)

sce_E8.5_merged <- SingleCellExperiment(assays = list(counts = merged_E8.5), colData = sce_E8.5@colData) 

sce_E8.5_merged <- sce_E8.5_merged[order(row.names(sce_E8.5_merged)), ]

sce_E8.5_merged <- sce_E8.5_merged %>% logNormCounts()

empty_matrix_E9.5 <- matrix(0, 
  nrow = length(setdiff(union(rownames(sce_E7.5), rownames(sce_E8.5)), rownames(sce_E9.5))), 
  ncol = ncol(sce_E9.5)
  ) %>% 
  as.data.frame() %>% 
  mutate(genes = setdiff(union(rownames(sce_E7.5), rownames(sce_E8.5)), rownames(sce_E9.5))) %>% 
  column_to_rownames(var = "genes") %>%
  purrr::set_names(colnames(sce_E9.5)) %>%
  as.matrix(.) %>%
  Matrix(., sparse = T)

merged_E9.5 <- merge_sparse(sce_E9.5@assays@data$counts, empty_matrix_E9.5)

sce_E9.5_merged <- SingleCellExperiment(assays = list(counts = merged_E9.5), colData = sce_E9.5@colData) 

sce_E9.5_merged <- sce_E9.5_merged[order(row.names(sce_E9.5_merged)), ]

sce_E9.5_merged <- sce_E9.5_merged %>% logNormCounts()

merged_E7.5_E8.5 <- Seurat::RowMergeSparseMatrices(
  mat1 = sce_E7.5_merged@assays@data$counts,
  mat2 = sce_E8.5_merged@assays@data$counts
  )

merged_all <- Seurat::RowMergeSparseMatrices(
  mat1 = merged_E7.5_E8.5,
  mat2 = sce_E9.5_merged@assays@data$counts
  )

colData(sce_E7.5_merged) <- colData(sce_E7.5_merged) %>% as.data.frame() %>% purrr::set_names(names(sce_E7.5_merged@colData)) %>% as("DFrame")
colData(sce_E8.5_merged) <- colData(sce_E8.5_merged) %>% as.data.frame() %>% purrr::set_names(names(sce_E7.5_merged@colData)) %>% as("DFrame")
colData(sce_E9.5_merged) <- colData(sce_E9.5_merged) %>% as.data.frame() %>% purrr::set_names(names(sce_E7.5_merged@colData)) %>% as("DFrame")

colnames(merged_all) <- colnames(merged_all) %>% make.unique()
merge_all_col_data <- rbind(rbind(sce_E7.5_merged@colData, sce_E8.5_merged@colData), sce_E9.5_merged@colData) %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "cell") %>% 
  mutate(cell = cell %>% make.unique()) %>% 
  column_to_rownames(var = "cell") %>% 
  as("DFrame")


# uncorrected <- cbind(sce_E7.5, cbind(sce_E8.5, sce_E9.5))
uncorrected <- SingleCellExperiment(assays = list(counts = merged_all), colData = merge_all_col_data) 
uncorrected$batch <- c(sce_E7.5$batch, sce_E8.5$batch, sce_E9.5$batch)
uncorrected$annotation <- c(sce_E7.5$annotation, sce_E8.5$annotation, sce_E9.5$annotation)
colLabels(uncorrected) <- colLabels(uncorrected) %>% forcats::fct_drop()
uncorrected <- uncorrected %>% logNormCounts()

# # Using RandomParam() as it is more efficient for file-backed matrices.
bpparam <- BiocParallel::MulticoreParam(workers = 51)
set.seed(123)

tic()
uncorrected <- uncorrected %>% 
  scater::runPCA(subset_row = universe,
                 BSPARAM = BiocSingular::RandomParam(),
                 # BPPARAM = bpparam,
                 exprs_values = "logcounts"
                 )
toc()

pcs <- uncorrected %>% reducedDim()
choices <- pcs %>% getClusteredPCs()
val <- metadata(choices)$chosen
reducedDim(uncorrected, "PCA_clust") <- pcs[, 1:val]

# uncorrected %>% qs::qsave(file = "~/PRIMUS/home/trufenc/uncorrected_2nd_reboot_2022_10_20.qs", nthreads = 50)
# uncorrected <- qs::qread(file = "~/PRIMUS/home/trufenc/uncorrected_2nd_reboot_2022_10_20.qs", nthreads = 50)

uncorrected %>% qs::qsave(file = "~/PRIMUS/home/trufenc/uncorrected_2nd_reboot_annotated_2022_10_20.qs", nthreads = 50)
# uncorrected <- qs::qread(file = "~/PRIMUS/home/trufenc/uncorrected_2nd_reboot_annotated_2022_10_20.qs", nthreads = 50)

# We use graph-based clustering on the components to obtain a summary of the population structure.
snn_gr <- uncorrected %>% buildSNNGraph(use.dimred = "PCA_clust")
clusters <- igraph::cluster_walktrap(snn_gr)$membership
tab <- table(Cluster = clusters, Batch = uncorrected$batch)
tab

# We can also visualize the corrected coordinates using a t-SNE plot
set.seed(123)
uncorrected <- uncorrected %>% runTSNE(dimred = "PCA_clust")
tsne_uncorrected <- uncorrected %>% plotTSNE(colour_by = "batch")
uncorrected <- uncorrected %>% runUMAP(dimred = "PCA_clust")
set.seed(123)
umap_uncorrected <- uncorrected %>% plotUMAP(colour_by = "batch")

# uncorrected %>% qs::qsave(file = "~/PRIMUS/home/trufenc/uncorrected_2nd_reboot_2022_10_20.qs", nthreads = 50)
# uncorrected <- qs::qread(file = "~/PRIMUS/home/trufenc/uncorrected_2nd_reboot_2022_10_20.qs", nthreads = 50)

uncorrected %>% qs::qsave(file = "~/PRIMUS/home/trufenc/uncorrected_2nd_reboot_annotated_2022_10_20.qs", nthreads = 50)
# uncorrected <- qs::qread(file = "~/PRIMUS/home/trufenc/uncorrected_2nd_reboot_annotated_2022_10_20.qs", nthreads = 50)

library(RColorBrewer)
cowplot::plot_grid(tsne_uncorrected, umap_uncorrected)

cowplot::plot_grid(uncorrected %>% 
                     plotTSNE(colour_by = "label") + 
                     scale_color_discrete(colorRampPalette(RColorBrewer::brewer.pal(name = "Paired", n = 12))(25)) +
                     ggeasy::easy_all_text_size(size = 20), 
                   uncorrected %>% 
                     plotUMAP(colour_by = "label") + 
                     scale_color_discrete(colorRampPalette(RColorBrewer::brewer.pal(name = "Paired", n = 12))(25)) +
                     ggeasy::easy_all_text_size(size = 20)
                   )

cowplot::plot_grid(uncorrected %>% 
                     plotTSNE(colour_by = "batch") + 
                     scale_color_discrete(colorRampPalette(RColorBrewer::brewer.pal(name = "Paired", n = 12))(25)) +
                     ggeasy::easy_all_text_size(size = 20), 
                   uncorrected %>% 
                     plotUMAP(colour_by = "batch") + 
                     scale_color_discrete(colorRampPalette(RColorBrewer::brewer.pal(name = "Paired", n = 12))(25)) +
                     ggeasy::easy_all_text_size(size = 20)
                   )

cowplot::plot_grid(uncorrected %>% 
                     plotTSNE(colour_by = "annotation") + 
                     scale_color_discrete(colorRampPalette(RColorBrewer::brewer.pal(name = "Paired", n = 12))(37)), 
                   uncorrected %>% 
                     plotUMAP(colour_by = "annotation") + 
                     scale_color_discrete(colorRampPalette(RColorBrewer::brewer.pal(name = "Paired", n = 12))(37))
                   )

```

# Batch effect removal
```{r}
# uncorrected <- qs::qread(file = "~/PRIMUS/home/trufenc/uncorrected_2nd_reboot_2022_10_20.qs", nthreads = 50)
uncorrected <- qs::qread(file = "~/PRIMUS/home/trufenc/uncorrected_2nd_reboot_annotated_2022_10_20.qs", nthreads = 50)

# Fast MNN
## Performing MNN correction
set.seed(123)
mnn_out <- fastMNN(sce_E7.5_merged, sce_E8.5_merged, sce_E9.5_merged, d = 50, k = 20, BSPARAM = BiocSingular::RandomParam(deferred = TRUE))
mnn_out$batch <- mnn_out$batch %>% as.factor() %>% forcats::fct_recode(`E7.5` = "1", `E8.5` = "2", `E9.5` = "3") 

# cluster the PCs
pcs <- mnn_out %>% reducedDim()
choices <- pcs %>% getClusteredPCs()
val <- metadata(choices)$chosen
reducedDim(mnn_out, "PCA_clust") <- pcs[, 1:val]

# mnn_out %>% qs::qsave(file = "~/PRIMUS/home/trufenc/mnn_2nd_reboot_2022_10_21.qs", nthreads = 50)
# mnn_out <- qs::qread(file = "~/PRIMUS/home/trufenc/mnn_2nd_reboot_2022_10_21.qs", nthreads = 50)

mnn_out %>% qs::qsave(file = "~/PRIMUS/home/trufenc/mnn_2nd_reboot_annotated_2022_10_21.qs", nthreads = 50)
# mnn_out <- qs::qread(file = "~/PRIMUS/home/trufenc/mnn_2nd_reboot_annotated_2022_10_21.qs", nthreads = 50)

# We cluster on the low-dimensional corrected coordinates to obtain a partitioning of the cells that serves as a proxy for the population structure.
snn_gr <- mnn_out %>% buildSNNGraph(use.dimred = "corrected")
clusters_mnn <- igraph::cluster_walktrap(snn_gr)$membership
tab_mnn <- table(Cluster = colLabels(uncorrected), Batch = mnn_out$batch)
tab_mnn <- table(Cluster = clusters_mnn, Batch = mnn_out$batch)
tab_mnn

set.seed(123)
mnn_out <- mnn_out %>% runTSNE(dimred = "corrected")
mnn_out$batch <- factor(mnn_out$batch)
tsne_corrected <- mnn_out %>% plotTSNE(colour_by = "batch")
# mnn_out %>% plotTSNE(colour_by = I(factor(clusters_mnn)))
mnn_out %>% plotTSNE(colour_by = I(colLabels(uncorrected)))
uncorrected %>% plotTSNE(colour_by = I(colLabels(uncorrected)))

set.seed(123)
mnn_out <- mnn_out %>% runUMAP(dimred = 'corrected', n_neighbors = 20)
umap_corrected <- mnn_out %>% plotUMAP(colour_by = "batch")
# mnn_out %>% plotUMAP(colour_by = I(factor(clusters_mnn)))
mnn_out %>% plotUMAP(colour_by = I(colLabels(uncorrected)))
uncorrected %>% plotUMAP(colour_by = I(colLabels(uncorrected)))

colLabels(mnn_out) <- colLabels(uncorrected)
mnn_out$annotation <- uncorrected$annotation

plot_grid(tsne_uncorrected, tsne_corrected, umap_uncorrected, umap_corrected, ncol = 2)

# mnn_out %>% qs::qsave(file = "~/PRIMUS/home/trufenc/mnn_2nd_reboot_2022_10_21.qs", nthreads = 50)
# mnn_out <- qs::qread(file = "~/PRIMUS/home/trufenc/mnn_2nd_reboot_2022_10_21.qs", nthreads = 50)

mnn_out %>% qs::qsave(file = "~/PRIMUS/home/trufenc/mnn_2nd_reboot_annotated_2022_10_21.qs", nthreads = 50)
# mnn_out <- qs::qread(file = "~/PRIMUS/home/trufenc/mnn_2nd_reboot_annotated_2022_10_21.qs", nthreads = 50)

pca_data_mnn <- prcomp(x = mnn_out@assays@data$reconstructed %>% t(), center = TRUE) 
batch_silhouette_mnn <- batch_sil(
  pca.data = pca_data_mnn, 
  batch = as.numeric(mnn_out$batch %>% dplyr::recode("E7.5" = "1", "E8.5" = "2", "E9.5" = "3")))

cowplot::plot_grid(mnn_out %>% 
                     plotTSNE(colour_by = "label") + 
                     scale_color_discrete(colorRampPalette(RColorBrewer::brewer.pal(name = "Paired", n = 12))(25)) +
                     ggeasy::easy_all_text_size(size = 20),
                   mnn_out %>% 
                     plotUMAP(colour_by = "label") + 
                     scale_color_discrete(colorRampPalette(RColorBrewer::brewer.pal(name = "Paired", n = 12))(25)) +
                     ggeasy::easy_all_text_size(size = 20)
                   )

cowplot::plot_grid(mnn_out %>% 
                     plotTSNE(colour_by = "batch") + 
                     scale_color_discrete(colorRampPalette(RColorBrewer::brewer.pal(name = "Paired", n = 12))(25)) +
                     ggeasy::easy_all_text_size(size = 20),
                   mnn_out %>% 
                     plotUMAP(colour_by = "batch") + 
                     scale_color_discrete(colorRampPalette(RColorBrewer::brewer.pal(name = "Paired", n = 12))(25)) +
                     ggeasy::easy_all_text_size(size = 20)
                   )

cowplot::plot_grid(mnn_out %>%
                     plotTSNE(colour_by = "annotation") +
                     scale_color_discrete(colorRampPalette(RColorBrewer::brewer.pal(name = "Paired", n = 12))(37)),
                   mnn_out %>%
                     plotUMAP(colour_by = "annotation") +
                     scale_color_discrete(colorRampPalette(RColorBrewer::brewer.pal(name = "Paired", n = 12))(37))
                   )
```

# Subset mnn output
```{r}
# mnn_out <- qs::qread(file = "~/PRIMUS/home/trufenc/mnn_2nd_reboot_2022_10_21.qs", nthreads = 50)
# uncorrected <- qs::qread(file = "~/PRIMUS/home/trufenc/uncorrected_2nd_reboot_2022_10_20.qs", nthreads = 50)

mnn_out <- qs::qread(file = "~/PRIMUS/home/trufenc/mnn_2nd_reboot_annotated_2022_10_21.qs", nthreads = 50)
uncorrected <- qs::qread(file = "~/PRIMUS/home/trufenc/uncorrected_2nd_reboot_annotated_2022_10_20.qs", nthreads = 50)

colnames(mnn_out) <- colnames(mnn_out) %>% make.unique()
colnames(uncorrected) <- colnames(uncorrected) %>% make.unique()


mnn_out@assays@data$counts <- uncorrected@assays@data$counts
mnn_out@assays@data$logcounts <- uncorrected@assays@data$logcounts

of_interest_df <- mnn_out@int_colData$reducedDims$TSNE %>% 
  as.data.frame() %>% 
  mutate(cell = mnn_out %>% colData() %>% rownames()) %>%  
  mutate(of_interest = dplyr::if_else(condition = V2 < 30 & between(V1, left = -37, right = 5), true = TRUE, false = FALSE)) 

of_interest_df %>% 
  ggplot(aes(x = V1, y = V2)) +
  geom_point(aes(colour = of_interest)) +
  theme_bw()
 
cells_of_interest <- of_interest_df %>% dplyr::filter(of_interest == TRUE) %>% pull(cell)

mnn_out_of_interest <- mnn_out[, cells_of_interest]

uncorrected_of_interest <- uncorrected[, cells_of_interest]

uncorrected_of_interest$batch <- uncorrected$batch[of_interest_df$of_interest %>% which()]
uncorrected_of_interest$label <- uncorrected$label[of_interest_df$of_interest %>% which()]
uncorrected_of_interest$annotation <- uncorrected$annotation[of_interest_df$of_interest %>% which()]

identical(as.character(mnn_out_of_interest$batch), as.character(mnn_out$batch[of_interest_df$of_interest %>% which()]))
identical(mnn_out_of_interest$batch, mnn_out$batch[of_interest_df$of_interest %>% which()])

df <- data.frame(sub_batch = mnn_out_of_interest$batch,
           batch = mnn_out$batch[of_interest_df$of_interest %>% which()]) %>% 
  mutate(identical = dplyr::if_else(condition = sub_batch == batch, true = TRUE, false = FALSE))

df <- data.frame(sub_label = mnn_out_of_interest$label,
           label = mnn_out$label[of_interest_df$of_interest %>% which()]) %>% 
  mutate(identical = dplyr::if_else(condition = sub_label == label, true = TRUE, false = FALSE))

# uncorrected_of_interest %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/uncorrected_of_interest_2nd_reboot_2022_10_21.qs", nthreads = 50)
# uncorrected_of_interest <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/uncorrected_of_interest_2nd_reboot_2022_10_21.qs", nthreads = 50)

uncorrected_of_interest %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/uncorrected_of_interest_2nd_reboot_annotated_2022_10_21.qs", nthreads = 50)
uncorrected_of_interest <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/uncorrected_of_interest_2nd_reboot_annotated_2022_10_21.qs", nthreads = 50)

uncorrected_of_interest$batch

sub_sce_7.5_of_interest <- uncorrected_of_interest[, which(uncorrected_of_interest$batch == "7.5")]
sub_sce_8.5_of_interest <- uncorrected_of_interest[, which(uncorrected_of_interest$batch == "8.5")]
sub_sce_9.5_of_interest <- uncorrected_of_interest[, which(uncorrected_of_interest$batch == "9.5")]

# dec_of_interest_mnn_default <- uncorrected_of_interest %>% modelGeneVar()
# dec_of_interest_mnn_default <- uncorrected_of_interest %>% modelGeneVar(block = uncorrected_of_interest$label, equiweight = TRUE)
dec_of_interest_mnn_default <- uncorrected_of_interest %>% modelGeneVar(block = uncorrected_of_interest$batch, equiweight = TRUE)
fit_of_interest_mnn_default <- metadata(dec_of_interest_mnn_default)
# Based on the largest metrics
hvg_var_of_interest_mnn <- dec_of_interest_mnn_default %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames()
hv_genes_of_interest_mnn <- hvg_var_of_interest_mnn %>% unique() %>% sort()

hvgs_no_block <- uncorrected_of_interest %>% modelGeneVar() %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames() %>% unique() %>% sort()
hvgs_cluster_block <- uncorrected_of_interest %>% modelGeneVar(block = uncorrected_of_interest$label, equiweight = TRUE) %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames() %>% unique() %>% sort()
hvgs_batch_block <- uncorrected_of_interest %>% modelGeneVar(block = uncorrected_of_interest$batch, equiweight = TRUE) %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames() %>% unique() %>% sort()

non_exclusive_hvgs <- union(union(intersect(hvgs_no_block, hvgs_batch_block), intersect(hvgs_no_block, hvgs_cluster_block)), intersect(hvgs_cluster_block, hvgs_batch_block))

library(VennDiagram)
draw.triple.venn(area1 = hvgs_no_block %>% length(),                          # Different color for each set
                 area2 = hvgs_batch_block %>% length(),
                 area3 = hvgs_cluster_block %>% length(),
                 n12 = intersect(hvgs_no_block, hvgs_batch_block) %>% length(),
                 n23 = intersect(hvgs_batch_block, hvgs_cluster_block) %>% length(),
                 n13 = intersect(hvgs_no_block, hvgs_cluster_block) %>% length(),
                 n123 = intersect(hvgs_no_block, intersect(hvgs_batch_block, hvgs_cluster_block)) %>% length(),
                 col = "red",
                 fill = c("blue", "green", "purple"),
                 category = c(" No Block", "By Batch", "By Cluster"))

data.frame(gene = union(hvgs_no_block, union(hvgs_batch_block, hvgs_cluster_block))) %>% 
  mutate(unique_to_no_block = dplyr::if_else(condition = gene %in% setdiff(hvgs_no_block, union(hvgs_batch_block, hvgs_cluster_block)), true = 1, false = 0)) %>% 
  mutate(unique_to_batch_block = dplyr::if_else(condition = gene %in% setdiff(hvgs_batch_block, union(hvgs_no_block, hvgs_cluster_block)), true = 1, false = 0)) %>% 
  mutate(unique_to_cluster_block = dplyr::if_else(condition = gene %in% setdiff(hvgs_cluster_block, union(hvgs_no_block, hvgs_batch_block)), true = 1, false = 0)) %>%          
  mutate(common_to_no_block_and_batch = dplyr::if_else(condition = gene %in% setdiff(intersect(hvgs_no_block, hvgs_batch_block), hvgs_cluster_block), true = 1, false = 0)) %>%          
  mutate(common_to_no_block_and_cluster = dplyr::if_else(condition = gene %in% setdiff(intersect(hvgs_no_block, hvgs_cluster_block), hvgs_batch_block), true = 1, false = 0)) %>%  
  mutate(common_to_batch_and_cluster = dplyr::if_else(condition = gene %in% setdiff(intersect(hvgs_batch_block, hvgs_cluster_block), hvgs_no_block), true = 1, false = 0)) %>%  
  mutate(common_to_all = dplyr::if_else(condition = gene %in% intersect(intersect(hvgs_no_block, hvgs_batch_block), hvgs_cluster_block), true = 1, false = 0)) %>%  
  pivot_longer(cols = unique_to_no_block:common_to_all, names_to = "Venn_diagram") %>% 
  dplyr::filter(value == 1) %>% 
  dplyr::select(-value) %>% 
  vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/hvgs_Venn_diagram_after_batch_effect_reboot.tsv", delim = "\t")
  
  
# dec_of_interest_mnn_default@listData %>% 
#   as.data.frame() %>% 
#   mutate(gene = dec_of_interest_mnn_default %>% as.data.frame() %>% rownames()) %>% 
#   dplyr::select(gene, everything()) %>% 
#   vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/hvgs_mnn_reboot_statistics_output.tsv", delim = "\t")

# data.frame(Gene = hvg_var_of_interest_mnn) %>% 
#   vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/hvgs_mnn_reboot.tsv", delim = "\t")


## Performing MNN correction
set.seed(123)
sub_mnn_out <- fastMNN(sub_sce_7.5_of_interest, sub_sce_8.5_of_interest, sub_sce_9.5_of_interest, d = 50, k = 20, subset.row = non_exclusive_hvgs, BSPARAM = BiocSingular::RandomParam(deferred = TRUE))
sub_mnn_out$batch <- sub_mnn_out$batch %>% as.factor() %>% forcats::fct_recode(`E7.5` = "1", `E8.5` = "2", `E9.5` = "3") 

sub_mnn_out@assays@data$counts <- mnn_out@assays@data$counts[non_exclusive_hvgs, of_interest_df$of_interest %>% which()]
sub_mnn_out@assays@data$logcounts <- mnn_out@assays@data$logcounts[non_exclusive_hvgs, of_interest_df$of_interest %>% which()]

sub_mnn_out$batch <- mnn_out$batch[of_interest_df$of_interest %>% which()]
sub_mnn_out$label <- mnn_out$label[of_interest_df$of_interest %>% which()]
sub_mnn_out$annotation <- mnn_out$annotation[of_interest_df$of_interest %>% which()]


# cluster the PCs
pcs <- sub_mnn_out %>% reducedDim()
choices <- pcs %>% getClusteredPCs()
val <- metadata(choices)$chosen
reducedDim(sub_mnn_out, "PCA_clust") <- pcs[, 1:val]

# We cluster on the low-dimensional corrected coordinates to obtain a partitioning of the cells that serves as a proxy for the population structure.
# snn_gr <- sub_mnn_out %>% buildSNNGraph(use.dimred = "corrected")
# clusters_mnn <- igraph::cluster_walktrap(snn_gr)$membership
# tab_mnn <- table(Cluster = colLabels(uncorrected), Batch = sub_mnn_out$batch)
# tab_mnn <- table(Cluster = clusters_mnn, Batch = sub_mnn_out$batch)
# tab_mnn

# set.seed(123)
set.seed(132)
sub_mnn_out <- sub_mnn_out %>% runTSNE(dimred = "corrected")
sub_mnn_out$batch <- factor(sub_mnn_out$batch)
sub_mnn_out %>% plotTSNE(colour_by = "batch")
# sub_mnn_out %>% plotTSNE(colour_by = I(factor(clusters_mnn)))
sub_mnn_out %>% plotTSNE(colour_by = I(colLabels(uncorrected_of_interest)))
uncorrected_of_interest %>% plotTSNE(colour_by = I(colLabels(uncorrected_of_interest)))

for(i in 1:100){
  print(i)
  set.seed(i)
  sub_mnn_out <- sub_mnn_out %>% runTSNE(dimred = "corrected")
  sub_mnn_out %>% 
    plotTSNE(colour_by = "label") %>% 
    ggsave(filename = paste0( "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/seed_number_tsne/sub_mnn_2nd_reboot_tsne_label_seed_number_", i, ".png"), width = 10, height = 7)
# sub_mnn_out %>% plotTSNE(colour_by = "annotation") %>% ggsave(filename = paste0( "/run/user/1002/gvfs/smb-share:server=primus.img.cas.cz,share=data/83_BIOINFORMATICS/Kadu/seed_number_tsne/tsne_annotation_seed_number_", i, "..pdf"), width = 10, height = 7)
# sub_mnn_out %>% plotTSNE(colour_by = "batch") %>% ggsave(filename = paste0( "/run/user/1002/gvfs/smb-share:server=primus.img.cas.cz,share=data/83_BIOINFORMATICS/Kadu/seed_number_tsne/tsne_batch_seed_number_", i, "..pdf"), width = 10, height = 7)
  }

# sub_mnn_out %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_20202_10_24.qs")
# sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_20202_10_24.qs")

sub_mnn_out %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_20202_10_24.qs")
sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_20202_10_24.qs")

set.seed(7)
# set.seed(25)
# set.seed(69)
# set.seed(75)
sub_mnn_out <- sub_mnn_out %>% runTSNE(dimred = "corrected")
sub_mnn_out$label <- sub_mnn_out$label %>% as.character()
sub_mnn_out %>% 
    plotTSNE(colour_by = I(dplyr::if_else(condition = sub_mnn_out$label %in% c("E8.5_3", "E9.5_8.1", "E9.5_8.2"), true = sub_mnn_out$label, false = NA_character_))) +
    scale_color_manual(
    values = c("#0000ff", # E8.5_3
               "#ff0e00", # E9.5_8.1
               "#209c41" # E9.5_8.2
               ), 
    labels = c("E8.5_3", 
               "E9.5_8.1",
               "E9.5_8.2"
               )) +
  ggeasy::easy_all_text_size(size = 20)
  

sub_mnn_out %>% 
  plotTSNE(colour_by = "label") + 
  scale_color_manual(
    values = c("#8a4201", # E7.5_3.1
               "#999999", # E7.5_3.2
               "#999999", # E7.5_3.3.1
               "#209c41", # E7.5_3.4
               "#c73b89", # E8.5_1"
               "#c73b89", # E8.5_11
               "#6fa8dc", # E8.5_2"
               "#0000ff", # E8.5_3
               "#9d4ad5", # E8.5_4
               "#9d4ad5", # E8.5_5
               "#ff9900", # E8.5_6
               "#9d4ad5", # E8.5_7
               "#6fa8dc", # E8.5_8
               "#9d4ad5", # E8.5_9
               "#ff0e00", # E9.5_1
               "#c6d844", # E9.5_2
               "#f056f0", # E9.5_3
               "#b44b7f", # E9.5_4
               "#46bdc6", # E9.5_5
               "#93c47d", # E9.5_6
               "#000000", # E9.5_7
               "#ff0e00", # E9.5_8.1
               "#f1c232" # E9.5_8.2
               ),
    labels = c("E7.5_3.1",
               "E7.5_3.2",
               "E7.5_3.3.1",
               "E7.5_3.4",
               "E8.5_1",
               "E8.5_11",
               "E8.5_2",
               "E8.5_3",
               "E8.5_4",
               "E8.5_5",
               "E8.5_6",
               "E8.5_7",
               "E8.5_8",
               "E8.5_9",
               "E9.5_1",
               "E9.5_2",
               "E9.5_3",
               "E9.5_4",
               "E9.5_5",
               "E9.5_6",
               "E9.5_7",
               "E9.5_8.1",
               "E9.5_8.2"
               )) +
  ggeasy::easy_all_text_size(size = 20)



set.seed(7)
sub_mnn_out <- sub_mnn_out %>% runUMAP(dimred = 'corrected', n_neighbors = 20)
umap_corrected <- sub_mnn_out %>% plotUMAP(colour_by = "batch")
# sub_mnn_out %>% plotUMAP(colour_by = I(factor(clusters_mnn)))
sub_mnn_out %>% plotUMAP(colour_by = I(colLabels(uncorrected_of_interest)))
uncorrected_of_interest %>% plotUMAP(colour_by = I(colLabels(uncorrected_of_interest)))

colLabels(sub_mnn_out) <- colLabels(uncorrected_of_interest)


sub_mnn_out$annotation <- uncorrected_of_interest$annotation
sub_mnn_out %>% plotTSNE(colour_by = "annotation")
sub_mnn_out %>% plotTSNE(colour_by = "label")
sub_mnn_out %>% plotTSNE(colour_by = "batch")

# sub_mnn_out %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_20202_10_24.qs", nthreads = 50)
# sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_20202_10_24.qs", nthreads = 50)

sub_mnn_out %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_20202_10_24.qs", nthreads = 50)
sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_20202_10_24.qs", nthreads = 50)
```

##### Clustering of sub_mnn_output
```{r}
kgraph_centers <- seq(from = 100, to = 1500, by = 100)

kgraph_clusters_list <- purrr::map(kgraph_centers, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sub_mnn_out, "PCA_clust"),
    TwoStepParam(
      first = KmeansParam(centers = i),
      second = NNGraphParam(k = 5)
    )
  )
})

combined_kgraph_1 <- kgraph_clusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_centers))

clustree(combined_kgraph_1, prefix = "kgraph_", edge_arrow = FALSE)

kgraph_k <- seq(from = 5, to = 100, by = 5)

kgraph_clusters_list <- purrr::map(kgraph_k, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sub_mnn_out, "PCA_clust"),
    TwoStepParam(
      first = KmeansParam(centers = 400),
      second = NNGraphParam(k = i)
    )
  )
})

combined_kgraph_2 <- kgraph_clusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_k))

clustree(combined_kgraph_2, prefix = "kgraph_", edge_arrow = FALSE)


cowplot::plot_grid(
  sub_mnn_out %>% plotTSNE(colour_by = I(combined_kgraph_2[[1]])),
  sub_mnn_out %>% plotTSNE(colour_by = I(combined_kgraph_2[[2]])),
  sub_mnn_out %>% plotTSNE(colour_by = I(combined_kgraph_2[[3]])),
  sub_mnn_out %>% plotTSNE(colour_by = I(combined_kgraph_2[[4]])),
  sub_mnn_out %>% plotTSNE(colour_by = I(combined_kgraph_2[[5]])),
  sub_mnn_out %>% plotTSNE(colour_by = I(combined_kgraph_2[[6]])),
  sub_mnn_out %>% plotTSNE(colour_by = I(combined_kgraph_2[[7]])),
  sub_mnn_out %>% plotTSNE(colour_by = I(combined_kgraph_2[[8]])),
  sub_mnn_out %>% plotTSNE(colour_by = I(combined_kgraph_2[[9]])),
  sub_mnn_out %>% plotTSNE(colour_by = I(combined_kgraph_2[[10]])),
  ncol = 5
)



table(combined_kgraph_2[7])

sub_mnn_out %>% plotTSNE(colour_by = I(combined_kgraph_2[[7]]))
sub_mnn_out %>% plotUMAP(colour_by = I(combined_kgraph_2[[7]]))

cowplot::plot_grid(sub_mnn_out %>% plotTSNE(colour_by = I(combined_kgraph_2[[7]])), 
                   sub_mnn_out %>% plotUMAP(colour_by = I(combined_kgraph_2[[7]]))
                   )

cluster_sub_mnn_out <- combined_kgraph_2[[7]] %>% 
  forcats::fct_recode("cluster_1" = "1", 
                      "cluster_2" = "2"
                      )

colLabels(sub_mnn_out) <- cluster_sub_mnn_out



# sub_mnn_out %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_20202_10_24.qs", nthreads = 50)
# sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_20202_10_24.qs", nthreads = 50)

sub_mnn_out %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_20202_10_24.qs", nthreads = 50)
sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_20202_10_24.qs", nthreads = 50)

sub_mnn_out %>% 
  plotTSNE(colour_by = "label") +
  ggeasy::easy_all_text_size(size = 20)

### Sub-clustering cluster 1
sce_cluster_1 <- sub_mnn_out[, which(colLabels(sub_mnn_out) == "cluster_1")]

dec_cluster_1 <- sce_cluster_1 %>% modelGeneVar()
hvg_var_cluster_1 <- dec_cluster_1 %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames()
sce_cluster_1 <- sce_cluster_1 %>% scater::runPCA(ncomponents = 50, subset_row = hvg_var_cluster_1)

pcs_cluster_1 <- sce_cluster_1 %>% reducedDim()
choices_cluster_1 <- pcs_cluster_1 %>% getClusteredPCs()
val_cluster_1 <- metadata(choices_cluster_1)$chosen
reducedDim(sce_cluster_1, "PCA_subclust_1") <- pcs_cluster_1[, 1:val_cluster_1]
plotReducedDim(sce_cluster_1, dimred = "PCA_subclust_1", ncomponents = 4)

kgraph_centers <- seq(from = 100, to = 1000, by = 100)
# kgraph_centers <- seq(from = 5, to = 100, by = 1)

kgraph_subclusters_list <- purrr::map(kgraph_centers, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sce_cluster_1, "PCA_subclust_1"),
    TwoStepParam(
      first = KmeansParam(centers = i),
      second = NNGraphParam(k = 5)
    )
  )
})

combined_kgraph_sub1 <- kgraph_subclusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_centers))

clustree(combined_kgraph_sub1, prefix = "kgraph_", edge_arrow = FALSE)

kgraph_k <- seq(from = 5, to = 100, by = 5)

kgraph_subclusters_list <- purrr::map(kgraph_k, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sce_cluster_1, "PCA_subclust_1"),
    TwoStepParam(
      first = KmeansParam(centers = 300),
      second = NNGraphParam(k = i)
    )
  )
})

combined_kgraph_sub2 <- kgraph_subclusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_k))

clustree(combined_kgraph_sub2, prefix = "kgraph_", edge_arrow = FALSE)

cowplot::plot_grid(
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[1]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[2]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[3]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[4]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[5]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[6]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[7]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[8]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[9]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[10]])),
  ncol = 5
)

cowplot::plot_grid(
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[11]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[12]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[13]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[14]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[15]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[16]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[17]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[18]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[19]])),
  sce_cluster_1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[20]])),
  ncol = 5
)

table(kgraph_subclusters_list[[1]])

sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[1]]))
sce_cluster_1 %>% plotUMAP(colour_by = I(kgraph_subclusters_list[[1]]))

cowplot::plot_grid(sce_cluster_1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[1]])),
                   sce_cluster_1 %>% plotUMAP(colour_by = I(kgraph_subclusters_list[[1]]))
                   )

# We merged clusters 1 and 7 on purpose. Aos, clusters 2 and 5
kgraph_subclusters_list[[1]] <- kgraph_subclusters_list[[1]] %>% forcats::fct_recode("cluster_1.1" = "1",
                                                                                     "cluster_1.2" = "2",
                                                                                     "cluster_1.3" = "3",
                                                                                     "cluster_1.4" = "4",
                                                                                     "cluster_1.2" = "5",
                                                                                     "cluster_1.5" = "6",
                                                                                     "cluster_1.1" = "7",
                                                                                     "cluster_1.6" = "8"
                                                                                     )

# update original clusters with new subclustering
new_levels <- colLabels(sub_mnn_out) %>% as.character()
# new_levels[new_levels == 1] <- kgraph_subclusters_list[[4]] %>% as.character()
new_levels[new_levels == "cluster_1"] <- kgraph_subclusters_list[[1]] %>% as.character()

colLabels(sub_mnn_out) <- factor(new_levels)

cowplot::plot_grid(sub_mnn_out %>% plotTSNE(colour_by = I(colLabels(sub_mnn_out))) + ggeasy::easy_all_text_size(size = 20), 
                   sub_mnn_out %>% plotUMAP(colour_by = I(colLabels(sub_mnn_out))) + ggeasy::easy_all_text_size(size = 20)
                   )






### Sub-clustering sub-cluster 1.2
sce_cluster_1.2 <- sub_mnn_out[, which(colLabels(sub_mnn_out) == "cluster_1.2")]

dec_cluster_1.2 <- sce_cluster_1.2 %>% modelGeneVar()
hvg_var_cluster_1.2 <- dec_cluster_1.2 %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames()
sce_cluster_1.2 <- sce_cluster_1.2 %>% scater::runPCA(ncomponents = 10, subset_row = hvg_var_cluster_1.2)

pcs_cluster_1.2 <- sce_cluster_1.2 %>% reducedDim()
choices_cluster_1.2 <- pcs_cluster_1.2 %>% getClusteredPCs()
val_cluster_1.2 <- metadata(choices_cluster_1.2)$chosen
reducedDim(sce_cluster_1.2, "PCA_subclust_1.2") <- pcs_cluster_1.2[, 1:val_cluster_1.2]
plotReducedDim(sce_cluster_1.2, dimred = "PCA_subclust_1.2", ncomponents = 4)

# kgraph_centers <- seq(from = 10, to = 100, by = 10)
kgraph_centers <- seq(from = 5, to = 100, by = 5)

kgraph_subclusters_list <- purrr::map(kgraph_centers, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sce_cluster_1.2, "PCA_subclust_1.2"),
    TwoStepParam(
      first = KmeansParam(centers = i),
      second = NNGraphParam(k = 5)
    )
  )
})

combined_kgraph_sub1 <- kgraph_subclusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_centers))

clustree(combined_kgraph_sub1, prefix = "kgraph_", edge_arrow = FALSE)

kgraph_k <- seq(from = 5, to = 100, by = 5)

kgraph_subclusters_list <- purrr::map(kgraph_k, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sce_cluster_1.2, "PCA_subclust_1.2"),
    TwoStepParam(
      first = KmeansParam(centers = 15),
      second = NNGraphParam(k = i)
    )
  )
})

combined_kgraph_sub2 <- kgraph_subclusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_k))

clustree(combined_kgraph_sub2, prefix = "kgraph_", edge_arrow = FALSE)

cowplot::plot_grid(
  sce_cluster_1.2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[1]])),
  sce_cluster_1.2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[2]])),
  sce_cluster_1.2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[3]])),
  sce_cluster_1.2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[4]])),
  sce_cluster_1.2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[5]])),
  sce_cluster_1.2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[6]])),
  sce_cluster_1.2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[7]])),
  sce_cluster_1.2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[8]])),
  sce_cluster_1.2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[9]])),
  sce_cluster_1.2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[10]])),
)



table(kgraph_subclusters_list[[2]])

sce_cluster_1.2 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[2]]))
sce_cluster_1.2 %>% plotUMAP(colour_by = I(kgraph_subclusters_list[[2]]))

cowplot::plot_grid(sce_cluster_1.2 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[2]])), 
                   sce_cluster_1.2 %>% plotUMAP(colour_by = I(kgraph_subclusters_list[[2]]))
                   )



kgraph_subclusters_list[[2]] <- kgraph_subclusters_list[[2]] %>% forcats::fct_recode("cluster_1.2.1" = "1",
                                                                                     "cluster_1.2.2" = "2"
                                                                                     # "cluster_1.2.3" = "3"
                                                                                     )

# update original clusters with new subclustering
new_levels <- colLabels(sub_mnn_out) %>% as.character()
# new_levels[new_levels == 1] <- kgraph_subclusters_list[[4]] %>% as.character()
new_levels[new_levels == "cluster_1.2"] <- kgraph_subclusters_list[[2]] %>% as.character()

colLabels(sub_mnn_out) <- factor(new_levels)

cowplot::plot_grid(sub_mnn_out %>% plotTSNE(colour_by = I(colLabels(sub_mnn_out))), 
                   sub_mnn_out %>% plotUMAP(colour_by = I(colLabels(sub_mnn_out))))









### Sub-clustering sub-cluster 1.2.1
sce_cluster_1.2.1 <- sub_mnn_out[, which(colLabels(sub_mnn_out) == "cluster_1.2.1")]

dec_cluster_1.2.1 <- sce_cluster_1.2.1 %>% modelGeneVar()
hvg_var_cluster_1.2.1 <- dec_cluster_1.2.1 %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames()
sce_cluster_1.2.1 <- sce_cluster_1.2.1 %>% scater::runPCA(ncomponents = 10, subset_row = hvg_var_cluster_1.2.1)

pcs_cluster_1.2.1 <- sce_cluster_1.2.1 %>% reducedDim()
choices_cluster_1.2.1 <- pcs_cluster_1.2.1 %>% getClusteredPCs()
val_cluster_1.2.1 <- metadata(choices_cluster_1.2.1)$chosen
reducedDim(sce_cluster_1.2.1, "PCA_subclust_1.2.1") <- pcs_cluster_1.2.1[, 1:val_cluster_1.2.1]
plotReducedDim(sce_cluster_1.2.1, dimred = "PCA_subclust_1.2.1", ncomponents = 4)

# kgraph_centers <- seq(from = 10, to = 100, by = 10)
kgraph_centers <- seq(from = 5, to = 100, by = 5)

kgraph_subclusters_list <- purrr::map(kgraph_centers, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sce_cluster_1.2.1, "PCA_subclust_1.2.1"),
    TwoStepParam(
      first = KmeansParam(centers = i),
      second = NNGraphParam(k = 5)
    )
  )
})

combined_kgraph_sub1 <- kgraph_subclusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_centers))

clustree(combined_kgraph_sub1, prefix = "kgraph_", edge_arrow = FALSE)

kgraph_k <- seq(from = 5, to = 100, by = 5)

kgraph_subclusters_list <- purrr::map(kgraph_k, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sce_cluster_1.2.1, "PCA_subclust_1.2.1"),
    TwoStepParam(
      first = KmeansParam(centers = 35),
      second = NNGraphParam(k = i)
    )
  )
})

combined_kgraph_sub2 <- kgraph_subclusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_k))

clustree(combined_kgraph_sub2, prefix = "kgraph_", edge_arrow = FALSE)

cowplot::plot_grid(
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[1]])),
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[2]])),
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[3]])),
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[4]])),
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[5]])),
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[6]])),
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[7]])),
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[8]])),
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[9]])),
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[10]])),
  ncol = 5
)

cowplot::plot_grid(
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[11]])),
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[12]])),
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[13]])),
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[14]])),
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[15]])),
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[16]])),
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[17]])),
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[18]])),
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[19]])),
  sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[20]])),
  ncol = 5
)



table(kgraph_subclusters_list[[1]])

sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[1]]))
sce_cluster_1.2.1 %>% plotUMAP(colour_by = I(kgraph_subclusters_list[[1]]))

cowplot::plot_grid(sce_cluster_1.2.1 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[1]])), 
                   sce_cluster_1.2.1 %>% plotUMAP(colour_by = I(kgraph_subclusters_list[[1]]))
                   )



kgraph_subclusters_list[[1]] <- kgraph_subclusters_list[[1]] %>% forcats::fct_recode("cluster_1.2.1.1" = "1",
                                                                                     "cluster_1.2.1.2" = "2",
                                                                                     "cluster_1.2.1.3" = "3"
                                                                                     )

# update original clusters with new subclustering
new_levels <- colLabels(sub_mnn_out) %>% as.character()
# new_levels[new_levels == 1] <- kgraph_subclusters_list[[4]] %>% as.character()
new_levels[new_levels == "cluster_1.2.1"] <- kgraph_subclusters_list[[1]] %>% as.character()

colLabels(sub_mnn_out) <- factor(new_levels)

cowplot::plot_grid(sub_mnn_out %>% plotTSNE(colour_by = I(colLabels(sub_mnn_out))),
                   sub_mnn_out %>% plotUMAP(colour_by = I(colLabels(sub_mnn_out)))
                   )




### Sub-clustering cluster 2
sce_cluster_2 <- sub_mnn_out[, which(colLabels(sub_mnn_out) == "cluster_2")]

dec_cluster_2 <- sce_cluster_2 %>% modelGeneVar()
hvg_var_cluster_2 <- dec_cluster_2 %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames()
sce_cluster_2 <- sce_cluster_2 %>% scater::runPCA(ncomponents = 50, subset_row = hvg_var_cluster_2)

pcs_cluster_2 <- sce_cluster_2 %>% reducedDim()
choices_cluster_2 <- pcs_cluster_2 %>% getClusteredPCs()
val_cluster_2 <- metadata(choices_cluster_2)$chosen
reducedDim(sce_cluster_2, "PCA_subclust_2") <- pcs_cluster_2[, 1:val_cluster_2]
plotReducedDim(sce_cluster_2, dimred = "PCA_subclust_2", ncomponents = 4)

kgraph_centers <- seq(from = 100, to = 1000, by = 100)
# kgraph_centers <- seq(from = 5, to = 100, by = 1)

kgraph_subclusters_list <- purrr::map(kgraph_centers, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sce_cluster_2, "PCA_subclust_2"),
    TwoStepParam(
      first = KmeansParam(centers = i),
      second = NNGraphParam(k = 5)
    )
  )
})

combined_kgraph_sub1 <- kgraph_subclusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_centers))

clustree(combined_kgraph_sub1, prefix = "kgraph_", edge_arrow = FALSE)

kgraph_k <- seq(from = 5, to = 100, by = 5)

kgraph_subclusters_list <- purrr::map(kgraph_k, function(i) {
  set.seed(123)
  clusterRows(
    reducedDim(sce_cluster_2, "PCA_subclust_2"),
    TwoStepParam(
      first = KmeansParam(centers = 300),
      second = NNGraphParam(k = i)
    )
  )
})

combined_kgraph_sub2 <- kgraph_subclusters_list %>% bind_cols() %>% purrr::set_names(paste0("kgraph_", kgraph_k))

clustree(combined_kgraph_sub2, prefix = "kgraph_", edge_arrow = FALSE)


cowplot::plot_grid(
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[1]])),
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[2]])),
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[3]])),
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[4]])),
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[5]])),
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[6]])),
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[7]])),
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[8]])),
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[9]])),
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[10]])),
  ncol = 5
)

cowplot::plot_grid(
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[11]])),
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[12]])),
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[13]])),
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[14]])),
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[15]])),
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[16]])),
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[17]])),
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[18]])),
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[19]])),
  sce_cluster_2 %>% plotTSNE(colour_by = I(combined_kgraph_sub2[[20]])),
  ncol = 5
)

table(kgraph_subclusters_list[[4]])

sce_cluster_2 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[4]]))
sce_cluster_2 %>% plotUMAP(colour_by = I(kgraph_subclusters_list[[4]]))

cowplot::plot_grid(sce_cluster_2 %>% plotTSNE(colour_by = I(kgraph_subclusters_list[[4]])), sce_cluster_2 %>% plotUMAP(colour_by = I(kgraph_subclusters_list[[4]])))

# Obs: we are on purposing merging clusters 2.2 and 2.4
kgraph_subclusters_list[[4]] <- kgraph_subclusters_list[[4]] %>% forcats::fct_recode("cluster_2.1" = "1",
                                                                                     "cluster_2.2" = "2",
                                                                                     "cluster_2.3" = "3",
                                                                                     "cluster_2.2" = "4"
                                                                                     )


# update original clusters with new subclustering
new_levels <- colLabels(sub_mnn_out) %>% as.character()
# new_levels[new_levels == 1] <- kgraph_subclusters_list[[4]] %>% as.character()
new_levels[new_levels == "cluster_2"] <- kgraph_subclusters_list[[4]] %>% as.character()

colLabels(sub_mnn_out) <- factor(new_levels)

cowplot::plot_grid(sub_mnn_out %>% plotTSNE(colour_by = I(colLabels(sub_mnn_out))) + ggeasy::easy_all_text_size(size = 20), 
                   sub_mnn_out %>% plotUMAP(colour_by = I(colLabels(sub_mnn_out))) + ggeasy::easy_all_text_size(size = 20)
                   )

sub_mnn_out %>% 
  plotTSNE(colour_by = I(colLabels(sub_mnn_out))) + 
  scale_color_manual(
    values = c("#209c41", # cluster_1.1
               "#6fa8dc", # cluster_1.2.1.1
               "#f67618", # cluster_1.2.1.2
               "#000000", # cluster_1.2.1.3
               "#0b5394", # cluster_1.2.2
               "#999999", # cluster_1.3
               "#e71c0b", # cluster_1.4
               "#f5c530", # cluster_1.5
               "#c6d844", # cluster_1.6
               "#42af9d", # cluster_2.1
               "#971958", # cluster_2.2
               "#f156a4" # cluster_2.3
               ),
    labels = c("cluster_1.1",
               "cluster_1.2.1.1",
               "cluster_1.2.1.2",
               "cluster_1.2.1.3",
               "cluster_1.2.2",
               "cluster_1.3",
               "cluster_1.4",
               "cluster_1.5",
               "cluster_1.6",
               "cluster_2.1",
               "cluster_2.2",
               "cluster_2.3"
               )) +
  ggeasy::easy_all_text_size(size = 20)
  

# sub_mnn_out %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_reboot_sub_clusters_final.qs", nthreads = 50)
# sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_reboot_sub_clusters_final.qs", nthreads = 50)

# sub_mnn_out %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_reboot_annotated_sub_clusters_final.qs", nthreads = 50)
sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_reboot_annotated_sub_clusters_final.qs", nthreads = 50)
```

########## trajectory analysis
```{r}
sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_reboot_annotated_sub_clusters_final.qs", nthreads = 50)

# Remove clusters 1.3, 2.1, 2.2, and 2.3
sub_mnn_out <- sub_mnn_out[, sub_mnn_out$label %>% str_which(pattern = "cluster_1.3|cluster_2.1|cluster_2.2|cluster_2.3", negate = TRUE)]
sub_mnn_out$label <- sub_mnn_out$label %>% forcats::fct_drop()

centroids <- sub_mnn_out %>% reducedDim("PCA_clust")

mst <- centroids %>% createClusterMST(clusters = sub_mnn_out$label)

line_data_tsne <- centroids %>% reportEdges(mst = mst, clusters = sub_mnn_out$label)

sub_mnn_out %>% plotTSNE(colour_by = "label") +
  geom_line(data = line_data_tsne, mapping = aes(x = dim1, y = dim2, group = edge))

# We obtain a pseudotime ordering by projecting the cells onto the MST with mapCellsToEdges()
# we move each cell onto the closest edge of the MST
map_tscan <- sub_mnn_out %>% mapCellsToEdges(mst = mst, use.dimred = "PCA_clust")
# The pseudotime is then calculated as the distance along the MST to this new position from a “root node” with orderCells()
tscan_pseudo <- map_tscan %>% orderCells(mst)
head(tscan_pseudo)

common_pseudo <- tscan_pseudo@assays@data@listData[[1]] %>% rowMeans(na.rm = TRUE)
sub_mnn_out %>% plotTSNE(
  # colour_by = I(common_pseudo),
  colour_by = "label",
  text_by = "label", text_colour = "black") +
  geom_line(data = line_data_tsne, mapping = aes(x = dim1, y = dim2, group = edge)) +
    scale_color_manual(
    values = c("#209c41", # cluster_1.1
               "#6fa8dc", # cluster_1.2.1.1
               "#f67618", # cluster_1.2.1.2
               "#000000", # cluster_1.2.1.3
               "#0b5394", # cluster_1.2.2
               # "#999999", # cluster_1.3
               "#e71c0b", # cluster_1.4
               "#f5c530", # cluster_1.5
               "#c6d844" # cluster_1.6
               # "#42af9d", # cluster_2.1
               # "#971958", # cluster_2.2
               # "#f156a4" # cluster_2.3
               ),
    labels = c("cluster_1.1",
               "cluster_1.2.1.1",
               "cluster_1.2.1.2",
               "cluster_1.2.1.3",
               "cluster_1.2.2",
               # "cluster_1.3",
               "cluster_1.4",
               "cluster_1.5",
               "cluster_1.6"
               # "cluster_2.1",
               # "cluster_2.2",
               # "cluster_2.3"
               )) +
  ggeasy::easy_all_text_size(size = 20)

line_data_umap <- centroids %>% reportEdges(mst = mst, clusters = sub_mnn_out$label)

sub_mnn_out %>% plotUMAP(colour_by = "label") +
  geom_line(data = line_data_umap, mapping = aes(x = dim1, y = dim2, group = edge))

sub_mnn_out %>% plotUMAP(
  colour_by = I(common_pseudo),
  # colour_by = "annotation",
  text_by = "label", text_colour = "black"
  # text_by = "annotation", text_colour = "red"
) +
  geom_line(data = line_data_umap, mapping = aes(x = dim1, y = dim2, group = edge)) +
  ggeasy::easy_all_text_size(size = 20)

pseudo_og <- sub_mnn_out %>% quickPseudotime(use.dimred = "PCA_clust", clusters = sub_mnn_out$label, outgroup = TRUE)
# pseudo_og <- sub_mnn_out %>% quickPseudotime(use.dimred = "PCA", outgroup = FALSE)
plot(pseudo_og$mst)

# Principal curves
# sub_mnn_out_sling <- sub_mnn_out %>% slingshot(clusterLabels = colLabels(sub_mnn_out), reducedDim = "PCA_clust")
sub_mnn_out_sling <- sub_mnn_out %>% slingshot(clusterLabels = colLabels(sub_mnn_out), reducedDim = "PCA_clust", start.clus = "cluster_1.5")
# sub_mnn_out_sling <- sub_mnn_out %>% slingshot(clusterLabels = sub_mnn_out$annotation, reducedDim = "PCA_clust", start.clus = "Haematoendothelial progenitors")

# sub_mnn_out_sling %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_sling.qs", nthreads = 50)
# sub_mnn_out_sling <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_sling.qs", nthreads = 50)

# sub_mnn_out_sling %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_sling.qs", nthreads = 50)
# sub_mnn_out_sling <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_sling.qs", nthreads = 50)

embedded <- sub_mnn_out_sling %>% embedCurves("TSNE")
embedded <- slingCurves(embedded)[[1]] # only 1 path.
embedded <- data.frame(embedded$s[embedded$ord,])

sub_mnn_out_sling %>% 
  plotTSNE(colour_by = "slingPseudotime_1") +
  geom_path(data = embedded, aes(x = Dim.1, y = Dim.2), size = 1.2) +
  ggeasy::easy_all_text_size(size = 20)

# sub_mnn_out_sling2 <- sub_mnn_out %>% slingshot(clusterLabels = colLabels(sub_mnn_out), reducedDim = 'PCA_clust')
sub_mnn_out_sling2 <- sub_mnn_out %>% slingshot(clusterLabels = colLabels(sub_mnn_out), reducedDim = 'PCA_clust', start.clus = "cluster_1.5")
# sub_mnn_out_sling2 <- sub_mnn_out %>% slingshot(clusterLabels = sub_mnn_out$annotation, reducedDim = 'PCA_clust', start.clus = "Haematoendothelial progenitors")
pseudo_paths <- sub_mnn_out_sling2 %>% slingPseudotime()

pseudo_paths %>% 
  as.data.frame() %>% 
  mutate(cluster = as.character(colLabels(sub_mnn_out))) %>%
  # mutate(cluster = as.character(sub_mnn_out$annotation)) %>% 
  gather(key = "key", value = "value", -cluster) %>% 
  ggplot(aes(x = value, y = factor(key), color = cluster)) + 
  geom_point() +
  theme_bw() +
  labs(x = "Pseudotime", y = "Lineage") +
  ggeasy::easy_all_text_size(size = 20)


reducedDim(sub_mnn_out_sling2, "UMAP") <- reducedDim(sub_mnn_out, "UMAP")
embedded_UMAP <- sub_mnn_out_sling %>% embedCurves("UMAP")
embedded_UMAP <- slingCurves(embedded_UMAP)[[1]] # only 1 path.
embedded_UMAP <- data.frame(embedded_UMAP$s[embedded_UMAP$ord,])
sub_mnn_out_sling %>% 
  plotUMAP(colour_by = "slingPseudotime_1") +
  geom_path(data = embedded_UMAP, aes(x = Dim.1, y = Dim.2), size = 1.2) +
  ggeasy::easy_all_text_size(size = 20)

# Taking the rowMeans just gives us a single pseudo-time for all cells. Cells
# in segments that are shared across paths have similar pseudo-time values in 
# all paths anyway, so taking the rowMeans is not particularly controversial.
shared_pseudo <- pseudo_paths %>% rowMeans(na.rm = TRUE)

# Need to loop over the paths and add each one separately.
# gg <- sub_mnn_out_sling2 %>% plotUMAP(colour_by = I(shared_pseudo))
sling_plots_tsne <- sub_mnn_out_sling2 %>% plotTSNE(colour_by = I(shared_pseudo))
embedded_tsne <- sub_mnn_out_sling2 %>% embedCurves("TSNE")
embedded_tsne <- slingCurves(embedded_tsne)
for (path in embedded_tsne) {
  embedded_tsne <- data.frame(path$s[path$ord, ])
  sling_plots_tsne <- sling_plots_tsne + geom_path(data = embedded_tsne, aes(x = Dim.1, y = Dim.2), size = 1.2)
}

sling_plots_tsne + ggeasy::easy_all_text_size(size = 20)

sling_plots_umap <- sub_mnn_out_sling2 %>% plotUMAP(colour_by = I(shared_pseudo))
embedded_umap <- sub_mnn_out_sling2 %>% embedCurves("UMAP")
embedded_umap <- slingCurves(embedded_umap)
for (path in embedded_umap) {
  embedded_umap <- data.frame(path$s[path$ord, ])
  sling_plots_umap <- sling_plots_umap + geom_path(data = embedded_umap, aes(x = Dim.1, y = Dim.2), size = 1.2)
}

sling_plots_umap + ggeasy::easy_all_text_size(size = 20)

# By using the MST as a scaffold for the global structure, slingshot() can accommodate branching events based on divergence in the principal curves
plot_grid(sling_plots_tsne, sling_plots_umap)

curve_assignments <- sub_mnn_out_sling2 %>% slingBranchID()
table(curve_assignments)

### Characterizing trajectories
# Changes along a trajectory
pseudo <- sub_mnn_out@assays@data@listData$reconstructed %>% 
  data.matrix() %>% 
  testPseudotime(pseudotime = tscan_pseudo[,1]) %>% .[["cluster_1.5"]]
pseudo <- pseudo %>% as.data.frame %>% mutate(SYMBOL = rownames(rowData(sub_mnn_out)))
pseudo <- pseudo %>% arrange(p.value)
# sorted <- pseudo[order(pseudo$p.value),]

sub_mnn_out$TSCAN_first <- tscan_pseudo[,1]
sub_mnn_out$TSCAN_second <- tscan_pseudo[,2]

downregulated <- pseudo %>% dplyr::filter(logFC < 0)
upregulated <- pseudo %>% dplyr::filter(logFC > 0)
best_down <- head(rownames(downregulated), 10)
best_up <- head(rownames(upregulated), 10)

sub_mnn_out %>% plotExpression(features = best_down, x = "TSCAN_first", colour_by = "label", exprs_values = "reconstructed")
sub_mnn_out %>% plotExpression(features = best_up, x = "TSCAN_first", colour_by = "label", exprs_values = "reconstructed")

sub_mnn_out$pseudo_paths <- pseudo_paths

sub_mnn_out %>% qs::qsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_trajectory.qs", nthreads = 50)
```


#  Heatmap showing the expression of the indicated genes (smoothed over 15 adjacent cells) with cells ordered along the pseudotime axis
# from https://www.nature.com/articles/s41422-020-0300-2
# monocle 3 from https://cole-trapnell-lab.github.io/monocle3/docs/installation/
```{r}
library(monocle3)
sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_trajectory.qs", nthreads = 50)

pseudo_paths <- sub_mnn_out$pseudo_paths

pseudo_time <- pseudo_paths %>% rowMeans(na.rm = TRUE)
lineage_1_pseudo_time <- pseudo_paths[, "Lineage1"]
lineage_2_pseudo_time <- pseudo_paths[, "Lineage2"]
lineage_3_pseudo_time <- pseudo_paths[, "Lineage3"]
lineage_4_pseudo_time <- pseudo_paths[, "Lineage4"]

genes <- c("Cldn5", "Stab2", "Sox17", "Esam", "Lyve1", "Gja5", "Runx1", "Cd109", "Pcdh17", "Hp", "Tyrobp", "Coro2a", "Ptprc", "Alox5ap", "Cd52", "Cx3cr1", "Irf8", "Coro1a", "Bcl11a", "Cyth4", "Mpo", "Gp5", "Thbs1", "Treml1", "F2rl2", "F10", "Plek", "Gypa", "Klf1", "Hbb-bs", "Gata1", "Hba-a1", "Hbb-y")

# normalize data as in https://www.nature.com/articles/s41422-020-0300-2
sub_mnn_out@assays@data$tpm <- (sub_mnn_out@assays@data$counts %>% cpm() %>% "/"(10)) %>% "+"(1) %>% log2()

tpm_data_to_heatmap <- sub_mnn_out@assays@data$tpm[genes, ] %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "cell") %>% 
  mutate(pseudo_time = pseudo_time) %>% 
  arrange(pseudo_time) %>% 
  # group_by(pseudo_time) %>% tally() %>% arrange(desc(n))
  pivot_longer(cols = genes[1]:genes[length(genes)], names_to = "gene", values_to = "tpm") 

max_tpm <- tpm_data_to_heatmap$tpm %>% max()

tpm_data_to_heatmap %>% 
  ggplot(aes(x = cell, y = gene, fill = tpm)) + 
  geom_tile() +
  # scale_fill_gradientn(colours = c("gray", "white", "yellow", "orange", "red", "darkred"), values = c(0, 1, 5, 9, 13, max_tpm), name = "log2(TPM/10 + 1)") +
  scale_fill_gradient2(low = "white", mid = "yellow", high = "red", midpoint = (0 + max_tpm)/2, name ="log2(TPM/10 + 1)") +
  theme(axis.text.x = element_blank()) +
  labs(x = "", y = "")

 # guide = guide_legend(title.position = "left", title.theme = element_text(angle = 90))


# reordering the cells by pseudotime
average_pseudo_df <- sub_mnn_out@assays@data$tpm[genes, ] %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "cell") %>% 
  mutate(pseudo_time = pseudo_time) %>% 
  arrange(pseudo_time) %>% 
  dplyr::select(-pseudo_time) %>% 
  column_to_rownames(var = "cell") %>% 
  t() %>% 
  as.data.frame()

lineage_1_pseudo_df <- sub_mnn_out@assays@data$tpm[genes, ] %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "cell") %>% 
  mutate(pseudo_time = lineage_1_pseudo_time) %>% 
  drop_na(pseudo_time) %>% 
  arrange(pseudo_time) %>% 
  dplyr::select(-pseudo_time) %>% 
  column_to_rownames(var = "cell") %>% 
  t() %>% 
  as.data.frame()

lineage_2_pseudo_df <- sub_mnn_out@assays@data$tpm[genes, ] %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "cell") %>% 
  mutate(pseudo_time = lineage_2_pseudo_time) %>% 
  drop_na(pseudo_time) %>% 
  arrange(pseudo_time) %>% 
  dplyr::select(-pseudo_time) %>% 
  column_to_rownames(var = "cell") %>% 
  t() %>% 
  as.data.frame()

lineage_3_pseudo_df <- sub_mnn_out@assays@data$tpm[genes, ] %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "cell") %>% 
  mutate(pseudo_time = lineage_3_pseudo_time) %>% 
  drop_na(pseudo_time) %>% 
  arrange(pseudo_time) %>% 
  dplyr::select(-pseudo_time) %>% 
  column_to_rownames(var = "cell") %>% 
  t() %>% 
  as.data.frame()

lineage_4_pseudo_df <- sub_mnn_out@assays@data$tpm[genes, ] %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "cell") %>% 
  mutate(pseudo_time = lineage_4_pseudo_time) %>% 
  drop_na(pseudo_time) %>% 
  arrange(pseudo_time) %>% 
  dplyr::select(-pseudo_time) %>% 
  column_to_rownames(var = "cell") %>% 
  t() %>% 
  as.data.frame()

plot_dendro_heatmap_for_paper <- function(df) {
  library(ggdendro)
  library(tidyverse)
  
  cell_names <- names(df)
  
  # Calculate the dendrogram
  dend <- df %>%
    # column_to_rownames(var = "Gene") %>%
    as.matrix() %>%
    dist() %>%
    hclust() %>%
    as.dendrogram()
  
   dend_data <- dend %>% dendro_data()

  # Setup the data, so that the layout is inverted (this is more
  # "clear" than simply using coord_flip())
  segment_data <- with(
    segment(dend_data),
    data.frame(x = y, y = x, xend = yend, yend = xend)
  )

  # Use the dendrogram label data to position the gene labels
  gene_pos_table <- with(
    dend_data$labels,
    data.frame(y_center = x, gene = as.character(label), height = 1)
  )

  # Table to position the samples
  cell_pos_table <- data.frame(cell = cell_names) %>%
    mutate(x_center = (1:nrow(.)), width = 1)
  
  # Neglecting the gap parameters
  heatmap_data <- df %>%
    t() %>% 
    as.data.frame() %>% 
    rownames_to_column(var = "cell") %>% 
    pivot_longer(cols = genes[1]:genes[length(genes)], names_to = "gene", values_to = "tpm") %>% 
    left_join(gene_pos_table) %>%
    left_join(cell_pos_table)

  # Limits for the vertical axes
  gene_axis_limits <- with(
    gene_pos_table,
    c(min(y_center - 0.5 * height), max(y_center + 0.5 * height))
  ) +
    0.1 * c(-1, 1) # extra spacing: 0.1
  
  max_expr <- ceiling(max(heatmap_data$tpm))
  
  # Heatmap plot
  plt_hmap <- heatmap_data %>% 
    ggplot(aes(x = x_center, y = y_center, fill = tpm, height = height, width = width)) +
    geom_tile() +
    # scale_fill_gradient2("Expr", high = "red", low = "blue") +
    scale_fill_gradientn(name = "log2(TPM/10 + 1)",
                         limits = c(0, max_expr),
                         # colours = colorRampPalette(colors = c("blue", "gray", "yellow", "orange", "red", "darkred"))(50)) +
                         # colours = colorRampPalette(colors = c("steelblue1", "yellow", "orangered"))(50)) +
                         colours = rev(colorRampPalette(colors = RColorBrewer::brewer.pal(n = 5, name = "RdYlBu"))(50))) +
    # scale_fill_viridis(name = "log2(TPM/10 + 1)", 
    #                    limits = c(0, max_expr),
    #                    option = "viridis") +
    scale_x_continuous(
      breaks = cell_pos_table$x_center,
      labels = cell_pos_table$cell,
      expand = c(0, 0)
    ) +
    # For the y axis, alternatively set the labels as: gene_position_table$gene
    scale_y_continuous(
      breaks = gene_pos_table[, "y_center"],
      # labels = rep("", nrow(gene_pos_table)),
      labels = heatmap_data %>% arrange(y_center) %>% pull(gene) %>% unique(),
      limits = gene_axis_limits,
      expand = c(0, 0),
      position = "right"
    ) +
    labs(x = "", y = "") +
    theme_bw() +
    theme(legend.position = "bottom") +
    theme(
      axis.text.x = element_blank(),
      axis.ticks.y = element_blank(),
      # margin: top, right, bottom, and left
      plot.margin = unit(c(1, 0.2, 0.2, -0.7), "cm")
      # panel.grid.minor = element_blank()
    )

  # Dendrogram plot
  plt_dendr <- ggplot(segment_data) +
    geom_segment(aes(x = x, y = y, xend = xend, yend = yend)) +
    scale_x_reverse(expand = c(0, 0.5)) +
    scale_y_continuous(
      breaks = gene_pos_table$y_center,
      # labels = gene_pos_table$Gene,
      labels = NULL,
      limits = gene_axis_limits,
      expand = c(0, 0)
    ) +
    labs(x = "Distance", y = "", colour = "", size = "") +
    theme(panel.grid.minor = element_blank()) +
    theme_void() 
  
  # results <- list(
    # heatmap = plot_grid(plt_dendr, plt_hmap, align = "h", rel_widths = c(1, 1)),
    # heatmap_data = 
  # )
  
  # plot_grid(plt_dendr, plt_hmap, align = "h", rel_widths = c(0.4, 1))
  fig_list <- list(plt_dendr = plt_dendr,
                   plt_hmap = plt_hmap)

}

dendro_heatmap_average_fig <- plot_dendro_heatmap_for_paper(average_pseudo_df)
dendro_heatmap_lineage_1_fig <- plot_dendro_heatmap_for_paper(lineage_1_pseudo_df)
dendro_heatmap_lineage_2_fig <- plot_dendro_heatmap_for_paper(lineage_2_pseudo_df)
dendro_heatmap_lineage_3_fig <- plot_dendro_heatmap_for_paper(lineage_3_pseudo_df)
dendro_heatmap_lineage_4_fig <- plot_dendro_heatmap_for_paper(lineage_4_pseudo_df)

dendro_heatmap_average_fig %>% ggsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/dendro_heatmap_average_fig.png", width = 10, height = 7, dpi = 900)
dendro_heatmap_lineage_1_fig %>% ggsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/dendro_heatmap_lineage_1_fig.png", width = 10, height = 7, dpi = 900)
dendro_heatmap_lineage_2_fig %>% ggsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/dendro_heatmap_lineage_2_fig.png", width = 10, height = 7, dpi = 900)
dendro_heatmap_lineage_3_fig %>% ggsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/dendro_heatmap_lineage_3_fig.png", width = 10, height = 7, dpi = 900)
dendro_heatmap_lineage_4_fig %>% ggsave(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/dendro_heatmap_lineage_4_fig.png", width = 10, height = 7, dpi = 900)
```

```{r}
sc_info_df <- data.frame(
  label = as.character(sub_mnn_out$label),
  batch = as.character(sub_mnn_out$batch),
  cell = colnames(sub_mnn_out)
)

pseudo_paths_df <- pseudo_paths %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "cell") %>% 
  rowwise() %>% 
  mutate(average = mean(c(Lineage1, Lineage2, Lineage3, Lineage4), na.rm = T)) %>% 
  full_join(sc_info_df) 

# Lineage 1
label_lineage_1_fig <- pseudo_paths_df %>% 
  dplyr::select(cell, Lineage1, label, batch) %>% 
  drop_na(Lineage1) %>% 
  arrange(Lineage1) %>% 
  mutate(order = 1:nrow(.)) %>% 
  mutate(y_axis = rep("y_axis", nrow(.))) %>% 
  ggplot(aes(x = order, y = y_axis, fill = batch)) +
  geom_tile() +
  theme_void() +
  # scale_y_discrete(labels = c("Embryonic stage"), position = "right") +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank()
        # axis.text.y = element_blank(),
        # axis.ticks.y = element_blank()
        ) +
  labs(x = "", y = "", fill = "Embryonic stage") 

batch_lineage_1_fig <- pseudo_paths_df %>% 
  dplyr::select(cell, Lineage1, label, batch) %>% 
  drop_na(Lineage1) %>% 
  arrange(Lineage1) %>% 
  mutate(order = 1:nrow(.)) %>% 
  mutate(y_axis = rep("y_axis", nrow(.))) %>% 
  ggplot(aes(x = order, y = y_axis, fill = label)) +
  geom_tile() +
  theme_void() +
    scale_fill_manual(
    values = c(
      # "#209c41", # cluster_1.1
      # "#6fa8dc", # cluster_1.2.1.1
      # "#f67618", # cluster_1.2.1.2
      # "#000000", # cluster_1.2.1.3
      # "#0b5394", # cluster_1.2.2
      "#999999", # cluster_1.3
      "#e71c0b", # cluster_1.4
      "#f5c530", # cluster_1.5
      # "#c6d844", # cluster_1.6
      "#42af9d", # cluster_2.1
      "#971958", # cluster_2.2
      "#f156a4" # cluster_2.3
      ),
    labels = c(
      # "cluster_1.1",
      # "cluster_1.2.1.1",
      # "cluster_1.2.1.2",
      # "cluster_1.2.1.3",
      # "cluster_1.2.2",
      "cluster_1.3",
      "cluster_1.4",
      "cluster_1.5",
      # "cluster_1.6",
      "cluster_2.1",
      "cluster_2.2",
      "cluster_2.3"
      )) +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  labs(x = "", y = "", fill = "Cluster")

sling_lineage_1_fig <- pseudo_paths_df %>% 
  dplyr::select(cell, Lineage1, label, batch) %>% 
  drop_na(Lineage1) %>% 
  arrange(Lineage1) %>% 
  mutate(order = 1:nrow(.)) %>% 
  mutate(y_axis = rep("y_axis", nrow(.))) %>% 
  ggplot(aes(x = order, y = y_axis, fill = Lineage1)) +
  geom_tile() +
  scale_fill_viridis(option = "viridis") +
  theme_void() +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  theme(legend.position = "bottom") +
  labs(x = "", y = "", fill = "Sling_pseutotime_1")

# Lineage 2
label_lineage_2_fig <- pseudo_paths_df %>% 
  dplyr::select(cell, Lineage2, label, batch) %>% 
  drop_na(Lineage2) %>% 
  arrange(Lineage2) %>% 
  mutate(order = 1:nrow(.)) %>% 
  mutate(y_axis = rep("y_axis", nrow(.))) %>% 
  ggplot(aes(x = order, y = y_axis, fill = batch)) +
  geom_tile() +
  theme_void() +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  labs(x = "", y = "", fill = "Embryonic stage") 

batch_lineage_2_fig <- pseudo_paths_df %>% 
  dplyr::select(cell, Lineage2, label, batch) %>% 
  drop_na(Lineage2) %>% 
  arrange(Lineage2) %>% 
  mutate(order = 1:nrow(.)) %>% 
  mutate(y_axis = rep("y_axis", nrow(.))) %>% 
  ggplot(aes(x = order, y = y_axis, fill = label)) +
  geom_tile() +
  theme_void() +
    scale_fill_manual(
    values = c(
      # "#209c41", # cluster_1.1
      "#6fa8dc", # cluster_1.2.1.1
      "#f67618", # cluster_1.2.1.2
      "#000000", # cluster_1.2.1.3
      "#0b5394", # cluster_1.2.2
      # "#999999", # cluster_1.3
      # "#e71c0b", # cluster_1.4
      "#f5c530" # cluster_1.5
      # "#c6d844", # cluster_1.6
      # "#42af9d", # cluster_2.1
      # "#971958", # cluster_2.2
      # "#f156a4" # cluster_2.3
      ),
    labels = c(
      # "cluster_1.1",
      "cluster_1.2.1.1",
      "cluster_1.2.1.2",
      "cluster_1.2.1.3",
      "cluster_1.2.2",
      # "cluster_1.3",
      # "cluster_1.4",
      "cluster_1.5"
      # "cluster_1.6",
      # "cluster_2.1",
      # "cluster_2.2",
      # "cluster_2.3"
      )) +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  labs(x = "", y = "", fill = "Cluster")

sling_lineage_2_fig <- pseudo_paths_df %>% 
  dplyr::select(cell, Lineage2, label, batch) %>% 
  drop_na(Lineage2) %>% 
  arrange(Lineage2) %>% 
  mutate(order = 1:nrow(.)) %>% 
  mutate(y_axis = rep("y_axis", nrow(.))) %>% 
  ggplot(aes(x = order, y = y_axis, fill = Lineage2)) +
  geom_tile() +
  scale_fill_viridis(option = "viridis") +
  theme_void() +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  theme(legend.position = "bottom") +
  labs(x = "", y = "", fill = "Sling_pseutotime_2")  

# Lineage 3
label_lineage_3_fig <- pseudo_paths_df %>% 
  dplyr::select(cell, Lineage3, label, batch) %>% 
  drop_na(Lineage3) %>% 
  arrange(Lineage3) %>% 
  mutate(order = 1:nrow(.)) %>% 
  mutate(y_axis = rep("y_axis", nrow(.))) %>% 
  ggplot(aes(x = order, y = y_axis, fill = batch)) +
  geom_tile() +
  theme_void() +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  labs(x = "", y = "", fill = "Embryonic stage") 

batch_lineage_3_fig <- pseudo_paths_df %>% 
  dplyr::select(cell, Lineage3, label, batch) %>% 
  drop_na(Lineage3) %>% 
  arrange(Lineage3) %>% 
  mutate(order = 1:nrow(.)) %>% 
  mutate(y_axis = rep("y_axis", nrow(.))) %>% 
  ggplot(aes(x = order, y = y_axis, fill = label)) +
  geom_tile() +
  theme_void() +
    scale_fill_manual(
    values = c(
      # "#209c41", # cluster_1.1
      # "#6fa8dc", # cluster_1.2.1.1
      # "#f67618", # cluster_1.2.1.2
      "#000000", # cluster_1.2.1.3
      # "#0b5394", # cluster_1.2.2
      "#999999", # cluster_1.3
      "#e71c0b", # cluster_1.4
      "#f5c530", # cluster_1.5
      "#c6d844" # cluster_1.6
      # "#42af9d", # cluster_2.1
      # "#971958", # cluster_2.2
      # "#f156a4" # cluster_2.3
      ),
    labels = c(
      # "cluster_1.1",
      # "cluster_1.2.1.1",
      # "cluster_1.2.1.2",
      "cluster_1.2.1.3",
      # "cluster_1.2.2",
      "cluster_1.3",
      "cluster_1.4",
      "cluster_1.5",
      "cluster_1.6"
      # "cluster_2.1",
      # "cluster_2.2",
      # "cluster_2.3"
      )) +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  labs(x = "", y = "", fill = "Cluster")

sling_lineage_3_fig <- pseudo_paths_df %>% 
  dplyr::select(cell, Lineage3, label, batch) %>% 
  drop_na(Lineage3) %>% 
  arrange(Lineage3) %>% 
  mutate(order = 1:nrow(.)) %>% 
  mutate(y_axis = rep("y_axis", nrow(.))) %>% 
  ggplot(aes(x = order, y = y_axis, fill = Lineage3)) +
  geom_tile() +
  scale_fill_viridis(option = "viridis") +
  theme_void() +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  theme(legend.position = "bottom") +
  labs(x = "", y = "", fill = "Sling_pseutotime_3")  

# Lineage 4
label_lineage_4_fig <- pseudo_paths_df %>% 
  dplyr::select(cell, Lineage4, label, batch) %>% 
  drop_na(Lineage4) %>% 
  arrange(Lineage4) %>% 
  mutate(order = 1:nrow(.)) %>% 
  mutate(y_axis = rep("y_axis", nrow(.))) %>% 
  ggplot(aes(x = order, y = y_axis, fill = batch)) +
  geom_tile() +
  theme_void() +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  labs(x = "", y = "", fill = "Embryonic stage") 

batch_lineage_4_fig <- pseudo_paths_df %>% 
  dplyr::select(cell, Lineage4, label, batch) %>% 
  drop_na(Lineage4) %>% 
  arrange(Lineage4) %>% 
  mutate(order = 1:nrow(.)) %>% 
  mutate(y_axis = rep("1", nrow(.))) %>% 
  ggplot(aes(x = order, y = y_axis, fill = label)) +
  geom_tile() +
  theme_void() +
  scale_fill_manual(
    values = c(
      "#209c41", # cluster_1.1
      "#6fa8dc", # cluster_1.2.1.1
      # "#f67618", # cluster_1.2.1.2
      "#000000", # cluster_1.2.1.3
      # "#0b5394", # cluster_1.2.2
      # "#999999", # cluster_1.3
      "#e71c0b", # cluster_1.4
      "#f5c530" # cluster_1.5
      # "#c6d844", # cluster_1.6
      # "#42af9d", # cluster_2.1
      # "#971958", # cluster_2.2
      # "#f156a4" # cluster_2.3
      ),
    labels = c(
      "cluster_1.1",
      "cluster_1.2.1.1",
      # "cluster_1.2.1.2",
      "cluster_1.2.1.3",
      # "cluster_1.2.2",
      # "cluster_1.3",
      "cluster_1.4",
      "cluster_1.5"
      # "cluster_1.6",
      # "cluster_2.1",
      # "cluster_2.2",
      # "cluster_2.3"
      )) +
  # scale_fill_continuous(limits = c(1, 797)) +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  labs(x = "", y = "", fill = "Cluster")

sling_lineage_4_fig <- pseudo_paths_df %>% 
  dplyr::select(cell, Lineage4, label, batch) %>% 
  drop_na(Lineage4) %>% 
  arrange(Lineage4) %>% 
  mutate(order = 1:nrow(.)) %>% 
  mutate(y_axis = rep("y_axis", nrow(.))) %>% 
  ggplot(aes(x = order, y = y_axis, fill = Lineage4)) +
  geom_tile() +
  scale_fill_viridis(option = "viridis") +
  theme_void() +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  theme(legend.position = "bottom") +
  labs(x = "", y = "", fill = "Sling_pseutotime_4")
  # labs(title = "Sling_pseudotime_4", x = "", y = "", fill = "Sling_pseutotime_4") +
  # theme(plot.title = element_text(hjust = 1.2, vjust = -50))
```

# organize heatmap figure
```{r}
dendro_heatmap_lineage_1_fig_legend <- get_legend(dendro_heatmap_lineage_1_fig$plt_hmap)
# ggdraw(dendro_heatmap_lineage_1_fig_legend)

dendro_heatmap_lineage_2_fig_legend <- get_legend(dendro_heatmap_lineage_2_fig$plt_hmap)
# ggdraw(dendro_heatmap_lineage_2_fig_legend)

dendro_heatmap_lineage_3_fig_legend <- get_legend(dendro_heatmap_lineage_3_fig$plt_hmap)
# ggdraw(dendro_heatmap_lineage_3_fig_legend)

dendro_heatmap_lineage_4_fig_legend <- get_legend(dendro_heatmap_lineage_4_fig$plt_hmap)
# ggdraw(dendro_heatmap_lineage_4_fig_legend)

label_lineage_1_fig_legend <- get_legend(label_lineage_1_fig)
# ggdraw(label_lineage_1_fig_legend)

label_lineage_2_fig_legend <- get_legend(label_lineage_2_fig)
# ggdraw(label_lineage_2_fig_legend)

label_lineage_3_fig_legend <- get_legend(label_lineage_3_fig)
# ggdraw(label_lineage_3_fig_legend)

label_lineage_4_fig_legend <- get_legend(label_lineage_4_fig)
# ggdraw(label_lineage_4_fig_legend)

batch_lineage_1_fig_legend <- get_legend(batch_lineage_1_fig)
# ggdraw(batch_lineage_1_fig_legend)

batch_lineage_2_fig_legend <- get_legend(batch_lineage_2_fig)
# ggdraw(batch_lineage_2_fig_legend)

batch_lineage_3_fig_legend <- get_legend(batch_lineage_3_fig)
# ggdraw(batch_lineage_3_fig_legend)

batch_lineage_4_fig_legend <- get_legend(batch_lineage_4_fig)
# ggdraw(batch_lineage_4_fig_legend)

sling_lineage_1_fig_legend <- get_legend(sling_lineage_1_fig)
# ggdraw(sling_lineage_1_fig_legend)

sling_lineage_2_fig_legend <- get_legend(sling_lineage_2_fig)
# ggdraw(sling_lineage_2_fig_legend)

sling_lineage_3_fig_legend <- get_legend(sling_lineage_3_fig)
# ggdraw(sling_lineage_3_fig_legend)

sling_lineage_4_fig_legend <- get_legend(sling_lineage_4_fig)
# ggdraw(sling_lineage_4_fig_legend)

cowplot::plot_grid(
  label_lineage_1_fig + theme(legend.position = "none"), 
  batch_lineage_1_fig + theme(legend.position = "none"), 
  sling_lineage_1_fig + theme(legend.position = "none"), 
  dendro_heatmap_lineage_1_fig$plt_hmap + theme(legend.position = "none"), 
  ncol = 1, 
  align = "v",
  rel_widths = c(1, 1, 1, 1),
  rel_heights = c(0.05, 0.05, 0.05, 1)
  )

label_lineage_1_fig_no_legend <- label_lineage_1_fig + theme(legend.position = "none")
label_lineage_2_fig_no_legend <- label_lineage_2_fig + theme(legend.position = "none")
label_lineage_3_fig_no_legend <- label_lineage_3_fig + theme(legend.position = "none")
label_lineage_4_fig_no_legend <- label_lineage_4_fig + theme(legend.position = "none")

batch_lineage_1_fig_no_legend <- batch_lineage_1_fig + theme(legend.position = "none")
batch_lineage_2_fig_no_legend <- batch_lineage_2_fig + theme(legend.position = "none")
batch_lineage_3_fig_no_legend <- batch_lineage_3_fig + theme(legend.position = "none")
batch_lineage_4_fig_no_legend <- batch_lineage_4_fig + theme(legend.position = "none")

sling_lineage_1_fig_no_legend <- sling_lineage_1_fig + theme(legend.position = "none")
sling_lineage_2_fig_no_legend <- sling_lineage_2_fig + theme(legend.position = "none")
sling_lineage_3_fig_no_legend <- sling_lineage_3_fig + theme(legend.position = "none")
sling_lineage_4_fig_no_legend <- sling_lineage_4_fig + theme(legend.position = "none")

dendro_heatmap_lineage_1_fig_no_legend <- dendro_heatmap_lineage_1_fig$plt_hmap + theme(legend.position = "none")
dendro_heatmap_lineage_2_fig_no_legend <- dendro_heatmap_lineage_2_fig$plt_hmap + theme(legend.position = "none")
dendro_heatmap_lineage_3_fig_no_legend <- dendro_heatmap_lineage_3_fig$plt_hmap + theme(legend.position = "none")
dendro_heatmap_lineage_4_fig_no_legend <- dendro_heatmap_lineage_4_fig$plt_hmap + theme(legend.position = "none")

### Lineage 1
right_legend_1 <- cowplot::plot_grid(
 ggdraw(label_lineage_1_fig_legend),
 ggdraw(batch_lineage_1_fig_legend),
 ggdraw(sling_lineage_1_fig_legend),
 ncol = 1, 
 align = "hv"
)

# from aplot
figure_1 <- dendro_heatmap_lineage_1_fig_no_legend %>% 
  insert_top(plot = sling_lineage_1_fig_no_legend, height = 1/20) %>% 
  insert_top(plot = batch_lineage_1_fig_no_legend, height = 1/20) %>% 
  insert_top(plot = label_lineage_1_fig_no_legend, height = 1/20) %>% 
  insert_left(plot = dendro_heatmap_lineage_1_fig$plt_dendr, width = 1/2) %>% 
  insert_bottom(plot = ggdraw(dendro_heatmap_lineage_1_fig_legend), height = 1/20) %>% 
  insert_right(plot = right_legend_1, width = 2/3)

figure_1 %>% aplot::ggsave(filename = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/heatmap_dendro_lineage_1.png", width = 12, height = 7, dpi = 900)

### Lineage 2
right_legend_2 <- cowplot::plot_grid(
 ggdraw(label_lineage_2_fig_legend),
 ggdraw(batch_lineage_2_fig_legend),
 ggdraw(sling_lineage_2_fig_legend),
 ncol = 1, 
 align = "hv"
)

# from aplot
figure_2 <- dendro_heatmap_lineage_2_fig_no_legend %>% 
  insert_top(plot = sling_lineage_2_fig_no_legend, height = 1/20) %>% 
  insert_top(plot = batch_lineage_2_fig_no_legend, height = 1/20) %>% 
  insert_top(plot = label_lineage_2_fig_no_legend, height = 1/20) %>% 
  insert_left(plot = dendro_heatmap_lineage_2_fig$plt_dendr, width = 1/2) %>% 
  insert_bottom(plot = ggdraw(dendro_heatmap_lineage_2_fig_legend), height = 1/20) %>% 
  insert_right(plot = right_legend_2, width = 2/3)

figure_2 %>% aplot::ggsave(filename = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/heatmap_dendro_lineage_2.png", width = 12, height = 7, dpi = 900)

### Lineage 3
right_legend_3 <- cowplot::plot_grid(
 ggdraw(label_lineage_3_fig_legend),
 ggdraw(batch_lineage_3_fig_legend),
 ggdraw(sling_lineage_3_fig_legend),
 ncol = 1, 
 align = "hv"
)

# from aplot
figure_3 <- dendro_heatmap_lineage_3_fig_no_legend %>% 
  insert_top(plot = sling_lineage_3_fig_no_legend, height = 1/20) %>% 
  insert_top(plot = batch_lineage_3_fig_no_legend, height = 1/20) %>% 
  insert_top(plot = label_lineage_3_fig_no_legend, height = 1/20) %>% 
  insert_left(plot = dendro_heatmap_lineage_3_fig$plt_dendr, width = 1/2) %>% 
  insert_bottom(plot = ggdraw(dendro_heatmap_lineage_3_fig_legend), height = 1/20) %>% 
  insert_right(plot = right_legend_3, width = 2/3)

figure_3 %>% aplot::ggsave(filename = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/heatmap_dendro_lineage_3.png", width = 12, height = 7, dpi = 900)

### Lineage 4
right_legend_4 <- cowplot::plot_grid(
 ggdraw(label_lineage_4_fig_legend),
 ggdraw(batch_lineage_4_fig_legend),
 ggdraw(sling_lineage_4_fig_legend),
 ncol = 1, 
 align = "hv"
)

# from aplot
figure_4 <- dendro_heatmap_lineage_4_fig_no_legend %>% 
  insert_top(plot = sling_lineage_4_fig_no_legend, height = 1/20) %>% 
  insert_top(plot = batch_lineage_4_fig_no_legend, height = 1/20) %>% 
  insert_top(plot = label_lineage_4_fig_no_legend, height = 1/20) %>% 
  insert_left(plot = dendro_heatmap_lineage_4_fig$plt_dendr, width = 1/2) %>% 
  insert_bottom(plot = ggdraw(dendro_heatmap_lineage_4_fig_legend), height = 1/20) %>% 
  insert_right(plot = right_legend_4, width = 2/3)

figure_4 %>% aplot::ggsave(filename = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/heatmap_dendro_lineage_4.png", width = 12, height = 7, dpi = 900)
```
# Expression plots for submnn
```{r}
sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_trajectory.qs", nthreads = 50)

for(i in 1:length(rownames(sub_mnn_out))) {
    violin_plot <- sub_mnn_out %>% 
      plotExpression(x = "label", features = rownames(sub_mnn_out)[i],  colour_by = "label") +
      theme_bw() +
      theme(axis.text.x = element_blank()) +
      ggeasy::easy_all_text_size(size = 20) +
      guides(color = guide_legend("Cluster")) +
      labs(x = "", y = "Normalized gene expression") +
      ggeasy::easy_legend_at(to = "bottom") 
    
    violin_plot %>% ggsave(filename = paste0( "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/violin_plots/", "merged_dataset_", rownames(sub_mnn_out)[i], "_ violin_plot.png"), width = 10, height = 5)  
}
```

# tsne plot of every hvg form E7.5, E8.5, and E9.5
```{r}
sce_E7.5 <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_7.5_reboot_no_histone_no_doublet_clustered_annotated_2022_10_20.qs", nthreads = 50)
sce_E8.5 <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_8.5_reboot_no_histone_no_doublet_clustered_annotated_2022_10_20.qs", nthreads = 50)
sce_E9.5 <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_9.5_reboot_no_histone_no_doublet_clustered_annotated_2022_10_20.qs", nthreads = 50)

sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_trajectory.qs", nthreads = 50)

gene_tsne_umap_plots <- function(gene, sce){
  
  tsne_df <- sce@int_colData@listData$reducedDims$TSNE %>% 
    as.data.frame() %>% 
    mutate(gene_expression = sce@assays@data$logcounts[gene,] %>% as.numeric()) %>% 
    arrange(gene_expression)
  
  umap_df <- sce@int_colData@listData$reducedDims$UMAP %>% 
    as.data.frame() %>% 
    mutate(gene_expression = sce@assays@data$logcounts[gene,] %>% as.numeric()) %>% 
    arrange(gene_expression)

gene_plots <- list(
  # tsne_plot = tsne_df %>% ggplot(aes(x = V1, y = V2, color = gene_expression, shape = batch)) +
  tsne_plot = tsne_df %>% ggplot(aes(x = V1, y = V2, color = gene_expression)) +
    geom_point() +
    scale_color_continuous(low = "grey", high = "red", guide = "colourbar", aesthetics = "colour") +
    theme_bw() + 
    labs(x = "TSNE 1", y = "TSNE 2", title = gene, color = "Gene expression") +
    ggeasy::easy_all_text_size(size = 20) +
    ggeasy::easy_legend_at(to = "bottom"), 
  
  umap_plot = umap_df %>% ggplot(aes(x = V1, y = V2, color = gene_expression)) +
    geom_point() +
    scale_color_continuous(low = "grey", high = "red", guide = "colourbar", aesthetics = "colour") +
    theme_bw() + 
    labs(x = "UMAP 1", y = "UMAP 2", title = gene, color = "Gene expression") +
    easy_plot_title_size(size = 20) +
    ggeasy::easy_legend_at(to = "bottom")
) %>% 
  purrr::set_names("tsne", "umap")
return(gene_plots)
}

sce_E7.5$batch <- "7.5"
sce_E8.5$batch <- "8.5"
sce_E9.5$batch <- "9.5"

colData(sce_E8.5) <- colData(sce_E8.5) %>% 
  as.data.frame() %>% 
  dplyr::rename(annotation = sce_gas) %>% 
  # dplyr::select(-c(sum.1, detected.1, total.1)) %>% 
  dplyr::select(colnames(colData(sce_E7.5))) %>% 
  as(Class = "DFrame")

colData(sce_E9.5) <- colData(sce_E9.5) %>% 
  as.data.frame() %>% 
  dplyr::rename(annotation = sce_gas) %>% 
  dplyr::select(colnames(colData(sce_E7.5))) %>% 
  as(Class = "DFrame")

dec_E7.5_default <- sce_E7.5 %>% modelGeneVar()
dec_E8.5_default <- sce_E8.5 %>% modelGeneVar()
dec_E9.5_default <- sce_E9.5 %>% modelGeneVar()

fit_E7.5_default <- metadata(dec_E7.5_default)
fit_E8.5_default <- metadata(dec_E8.5_default)
fit_E9.5_default <- metadata(dec_E9.5_default)

# Based on the largest metrics
hvg_var_E7.5 <- dec_E7.5_default %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames()
hvg_var_E8.5 <- dec_E8.5_default %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames()
hvg_var_E9.5 <- dec_E9.5_default %>% as.data.frame() %>% dplyr::filter(bio >= 0 & FDR <= 0.05) %>% rownames()

hv_genes_E7.5 <- hvg_var_E7.5 %>% unique() %>% sort()
hv_genes_E8.5 <- hvg_var_E8.5 %>% unique() %>% sort()
hv_genes_E9.5 <- hvg_var_E9.5 %>% unique() %>% sort()

sce_E7.5 <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_7.5_reboot_no_histone_no_doublet_clustered_annotated_2022_10_20.qs", nthreads = 50)
sce_E8.5 <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_8.5_reboot_no_histone_no_doublet_clustered_annotated_2022_10_20.qs", nthreads = 50)
sce_E9.5 <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_9.5_reboot_no_histone_no_doublet_clustered_annotated_2022_10_20.qs", nthreads = 50)

for(i in 1:length(hv_genes_E7.5)) {
  print(i)
    tsne_umap_plot <- sce_E7.5 %>%
      gene_tsne_umap_plots(gene = hv_genes_E7.5[i])

    tsne_umap_plot$tsne %>% ggsave(filename = paste0( "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/tsne_plots/", "sce_E7.5_hvg_", hv_genes_E7.5[i], "_tsne.png"), width = 7.5, height = 7.5, dpi = 900)  
}

for(i in 1:length(hv_genes_E8.5)) {
  print(i)
    tsne_umap_plot <- sce_E8.5 %>%
      gene_tsne_umap_plots(gene = hv_genes_E8.5[i])

    tsne_umap_plot$tsne %>% ggsave(filename = paste0( "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/tsne_plots/", "sce_E8.5_hvg_", hv_genes_E8.5[i], "_tsne.png"), width = 7.5, height = 7.5, dpi = 900)  
}

for(i in 1:length(hv_genes_E9.5)) {
  print(i)
    tsne_umap_plot <- sce_E9.5 %>%
      gene_tsne_umap_plots(gene = hv_genes_E9.5[i])

    tsne_umap_plot$tsne %>% ggsave(filename = paste0( "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/tsne_plots/", "sce_E9.5_hvg_", hv_genes_E9.5[i], "_tsne.png"), width = 7.5, height = 7.5, dpi = 900)  
}

for(i in 1:length(rownames(sub_mnn_out))) {
  print(i)
    tsne_umap_plot <- sub_mnn_out %>%
      gene_tsne_umap_plots(gene = rownames(sub_mnn_out)[i])

    tsne_umap_plot$tsne %>% ggsave(filename = paste0( "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/tsne_plots/", "merged_dataset_hvg_", rownames(sub_mnn_out)[i], "_tsne.png"), width = 7.5, height = 7.5, dpi = 900)  
}


```


```{r}
plot_heatmap_and_dendro <- function(df) {
  library(ggdendro)
  library(tidyverse)

  sample_names <- names(df)
  # Calculate the dendrogram
  dend <- df %>%
    # column_to_rownames(var = "Gene") %>%
    as.matrix() %>%
    dist() %>%
    hclust() %>%
    as.dendrogram()
  
  dend_data <- dend %>% dendro_data()

  # Setup the data, so that the layout is inverted (this is more
  # "clear" than simply using coord_flip())
  segment_data <- with(
    segment(dend_data),
    data.frame(x = y, y = x, xend = yend, yend = xend)
  )

  # Use the dendrogram label data to position the gene labels
  gene_pos_table <- with(
    dend_data$labels,
    data.frame(y_center = x, Gene = as.character(label), height = 1)
  )

  # Table to position the samples
  sample_pos_table <- data.frame(Sample = sample_names) %>%
    mutate(x_center = (1:nrow(.)), width = 1)

  # Neglecting the gap parameters
  heatmap_data <- df %>%
    rownames_to_column("Gene") %>%
    gather(value = "expr", key = "Sample", -Gene) %>%
    # reshape2::melt(value.name = "expr", varnames = c("Gene", "Sample")) %>%
    left_join(gene_pos_table) %>%
    left_join(sample_pos_table)

  # Limits for the vertical axes
  gene_axis_limits <- with(
    gene_pos_table,
    c(min(y_center - 0.5 * height), max(y_center + 0.5 * height))
  ) +
    0.1 * c(-1, 1) # extra spacing: 0.1

  
  max_expr <- max(abs(floor(min(heatmap_data$expr))), abs(ceiling(max(heatmap_data$expr))))
  
  # Heatmap plot
  plt_hmap <- ggplot(
    heatmap_data,
    aes(
      x = x_center, y = y_center, fill = expr,
      height = height, width = width
    )
  ) +
    geom_tile() +
    # scale_fill_gradient2("Expr", high = "red", low = "blue") +
    scale_fill_gradientn("log2FC", 
                         limits = c(-1*max_expr, max_expr),
                         colours = rev(colorRampPalette(colors = c("red", "white", "#094FED"))(50))) +
    scale_x_continuous(
      breaks = sample_pos_table$x_center,
      labels = sample_pos_table$Sample,
      expand = c(0, 0)
    ) +
    # For the y axis, alternatively set the labels as: gene_position_table$gene
    scale_y_continuous(
      breaks = gene_pos_table[, "y_center"],
      # labels = rep("", nrow(gene_pos_table)),
      labels = heatmap_data %>% arrange(y_center) %>% pull(Gene) %>% unique(),
      limits = gene_axis_limits,
      expand = c(0, 0),
      position = "right"
    ) +
    labs(x = "Treatment", y = "") +
    theme_bw() +
    theme(
      axis.text.x = element_text(size = rel(1), hjust = 0.5, angle = 0),
      # margin: top, right, bottom, and left
      plot.margin = unit(c(1, 0.2, 0.2, -0.7), "cm"),
      panel.grid.minor = element_blank()
    ) +
    ggeasy::easy_all_text_size(size = 20) +
    ggeasy::easy_y_axis_labels_size(size = 0) +
    ggeasy::easy_rotate_x_labels(angle = 90)

  # Dendrogram plot
  plt_dendr <- ggplot(segment_data) +
    geom_segment(aes(x = x, y = y, xend = xend, yend = yend)) +
    scale_x_reverse(expand = c(0, 0.5)) +
    scale_y_continuous(
      breaks = gene_pos_table$y_center,
      # labels = gene_pos_table$Gene,
      labels = NULL,
      limits = gene_axis_limits,
      expand = c(0, 0)
    ) +
    labs(x = "Distance", y = "", colour = "", size = "") +
    theme(panel.grid.minor = element_blank()) +
    theme_bw() +
    ggeasy::easy_all_text_size(size = 20)
  
  # results <- list(
    # heatmap = plot_grid(plt_dendr, plt_hmap, align = "h", rel_widths = c(1, 1)),
    # heatmap_data = 
  # )
  
  plot_grid(plt_dendr, plt_hmap, align = "h", rel_widths = c(0.4, 1))
}
```


# Heatmap clusters in lineage 2, cells vs genes
```{r}
lineage2_clusters <- pseudo_paths %>% 
  as.data.frame() %>% 
  mutate(cluster = as.character(colLabels(sub_mnn_out))) %>% 
  drop_na(Lineage2) %>% 
  group_by(cluster) %>% 
  tally() %>% 
  pull(cluster)

sub_mnn_out_lineage2 <- sub_mnn_out[, sub_mnn_out$label %>% str_which(pattern = paste(lineage2_clusters, collapse = "|"))]

cell_label <- data.frame(cell = sub_mnn_out_lineage2 %>% colnames(),
                         label = sub_mnn_out_lineage2@colData$label)

df_to_heatmap <- sub_mnn_out_lineage2@assays@data$logcounts %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "gene") %>% 
  pivot_longer(cols = 2:ncol(.), names_to = "cell", values_to = "logcounts") %>% 
  full_join(cell_label)

df_to_heatmap %>% 
  ggplot(aes(x = cell, y = gene, fill = logcounts)) +
  geom_tile() +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_blank()) +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red")  
  # facet_wrap(vars(label))


```

# Plot gene expression by lineage and cluster
```{r}
sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_trajectory.qs", nthreads = 50)

lineage_clusters_df <- pseudo_paths %>% 
  as.data.frame() %>% 
  mutate(cluster = as.character(colLabels(sub_mnn_out))) %>% 
  rownames_to_column(var = "cell")

lineage_clusters_df <- lineage_clusters_df %>% 
  pivot_longer(cols = Lineage1:Lineage4, names_to = "lineage", values_to = "path") %>% 
  drop_na(path)

lineage_clusters_list <- lineage_clusters_df %>% split(f = .$lineage)

# Figures - lineage 1
sub_mnn_out_lineage_1 <- sub_mnn_out[, which(colnames(sub_mnn_out) %in% lineage_clusters_list$Lineage1$cell)]
colLabels(sub_mnn_out_lineage_1) <- colLabels(sub_mnn_out_lineage_1) %>% 
  forcats::fct_drop() %>% 
  forcats::fct_relevel(c("cluster_1.5", "cluster_1.3", "cluster_1.4", "cluster_2.1", "cluster_2.3", "cluster_2.2"))

gene_expression_lineage_1_cluster_list <- purrr::map(seq_along(rownames(sub_mnn_out_lineage_1)), function(i) {
  print(i)
  plotExpression(
    object = sub_mnn_out_lineage_1, 
    features = rownames(sub_mnn_out_lineage_1)[i], 
    x = "label",
    colour_by = "label", 
    exprs_values = "logcounts"
    ) 
})

purrr::map(seq_along(gene_expression_lineage_1_cluster_list), function(i) {
  print(i)
  gene_expression_lineage_1_cluster_list[[i]] %>% 
    ggsave(filename = paste0("~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/lineage_cluster_expression_plots/lineage_1_", rownames(sub_mnn_out_lineage_1)[i], ".png"), width = 12, height = 7, dpi = 900)
})

expression_lineage_1_df <- sub_mnn_out_lineage_1@assays@data$logcounts %>%
  t() %>%
  as.data.frame() %>%
  mutate(cluster = sub_mnn_out_lineage_1$label) %>%
  rownames_to_column(var = "cell") %>%
  pivot_longer(cols = 2:1421, names_to = "gene", values_to = "logcounts")

median_by_cluster_lineage_1_df <- sub_mnn_out_lineage_1@assays@data$logcounts %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate(cluster = sub_mnn_out_lineage_1$label) %>% 
  rownames_to_column(var = "cell") %>% 
  pivot_longer(cols = 2:1421, names_to = "gene", values_to = "logcounts") %>% 
  group_by(gene, cluster) %>% 
  summarise(median = median(logcounts, na.rm = TRUE)) %>% 
  pivot_wider(names_from = "cluster", values_from = "median")

median_by_cluster_lineage_1_df %>% 
  dplyr::filter(cluster_1.3 > cluster_1.5) %>% 
  dplyr::filter(cluster_1.4 > cluster_1.3) %>% 
  dplyr::filter(cluster_2.1 > cluster_1.4) %>% 
  dplyr::filter(cluster_2.3 > cluster_2.1) %>% 
  dplyr::filter(cluster_2.2 > cluster_2.3) %>% 
  vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/lineage_cluster_expression_plots/lineage_1_median_by_cluster_df.tsv", delim = "\t")

# Figures - lineage 2
sub_mnn_out_lineage_2 <- sub_mnn_out[, which(colnames(sub_mnn_out) %in% lineage_clusters_list$Lineage2$cell)]
colLabels(sub_mnn_out_lineage_2) <- colLabels(sub_mnn_out_lineage_2) %>% 
  forcats::fct_drop() %>% 
  forcats::fct_relevel(c("cluster_1.5", "cluster_1.2.1.3", "cluster_1.2.1.1", "cluster_1.2.1.2", "cluster_1.2.2"))

gene_expression_lineage_2_cluster_list <- purrr::map(seq_along(rownames(sub_mnn_out_lineage_2)), function(i) {
  print(i)
  plotExpression(
    object = sub_mnn_out_lineage_2, 
    features = rownames(sub_mnn_out_lineage_2)[i], 
    x = "label",
    colour_by = "label", 
    exprs_values = "logcounts"
    ) 
})

purrr::map(seq_along(gene_expression_lineage_2_cluster_list), function(i) {
  print(i)
  gene_expression_lineage_2_cluster_list[[i]] %>% 
    ggsave(filename = paste0("~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/lineage_cluster_expression_plots/lineage_2_", rownames(sub_mnn_out_lineage_2)[i], ".png"), width = 12, height = 7, dpi = 900)
})

expression_lineage_2_df <- sub_mnn_out_lineage_2@assays@data$logcounts %>%
  t() %>%
  as.data.frame() %>%
  mutate(cluster = sub_mnn_out_lineage_2$label) %>%
  rownames_to_column(var = "cell") %>%
  pivot_longer(cols = 2:1421, names_to = "gene", values_to = "logcounts")

median_by_cluster_lineage_2_df <- sub_mnn_out_lineage_2@assays@data$logcounts %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate(cluster = sub_mnn_out_lineage_2$label) %>% 
  rownames_to_column(var = "cell") %>% 
  pivot_longer(cols = 2:1421, names_to = "gene", values_to = "logcounts") %>% 
  group_by(gene, cluster) %>% 
  summarise(median = median(logcounts, na.rm = TRUE)) %>% 
  pivot_wider(names_from = "cluster", values_from = "median")

median_by_cluster_lineage_2_df %>% 
  dplyr::filter(cluster_1.2.1.3 > cluster_1.5) %>% 
  dplyr::filter(cluster_1.2.1.1 > cluster_1.2.1.3) %>% 
  dplyr::filter(cluster_1.2.1.2 > cluster_1.2.1.1) %>% 
  dplyr::filter(cluster_1.2.2 > cluster_1.2.1.2) %>% 
  vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/lineage_cluster_expression_plots/lineage_2_median_by_cluster_df.tsv", delim = "\t")

# Figures - lineage 3
sub_mnn_out_lineage_3 <- sub_mnn_out[, which(colnames(sub_mnn_out) %in% lineage_clusters_list$Lineage3$cell)]
colLabels(sub_mnn_out_lineage_3) <- colLabels(sub_mnn_out_lineage_3) %>% 
  forcats::fct_drop() %>% 
  forcats::fct_relevel(c("cluster_1.5", "cluster_1.2.1.3", "cluster_1.3", "cluster_1.4", "cluster_1.6"))

gene_expression_lineage_3_cluster_list <- purrr::map(seq_along(rownames(sub_mnn_out_lineage_3)), function(i) {
# gene_expression_lineage_3_cluster_list <- purrr::map(700:length(rownames(sub_mnn_out_lineage_3)), function(i) {  
  print(i)
  plotExpression(
    object = sub_mnn_out_lineage_3, 
    features = rownames(sub_mnn_out_lineage_3)[i], 
    x = "label",
    colour_by = "label", 
    exprs_values = "logcounts"
    ) 
})

# purrr::map(seq_along(gene_expression_lineage_3_cluster_list), function(i) {
purrr::map(700:length(rownames(sub_mnn_out_lineage_3)), function(i) {  
  print(i)
  gene_expression_lineage_3_cluster_list[[i]] %>% 
    ggsave(filename = paste0("~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/lineage_cluster_expression_plots/lineage_3_", rownames(sub_mnn_out_lineage_3)[i], ".png"), width = 12, height = 7, dpi = 900)
})

expression_lineage_3_df <- sub_mnn_out_lineage_3@assays@data$logcounts %>%
  t() %>%
  as.data.frame() %>%
  mutate(cluster = sub_mnn_out_lineage_3$label) %>%
  rownames_to_column(var = "cell") %>%
  pivot_longer(cols = 2:1421, names_to = "gene", values_to = "logcounts")

median_by_cluster_lineage_3_df <- sub_mnn_out_lineage_3@assays@data$logcounts %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate(cluster = sub_mnn_out_lineage_3$label) %>% 
  rownames_to_column(var = "cell") %>% 
  pivot_longer(cols = 2:1421, names_to = "gene", values_to = "logcounts") %>% 
  group_by(gene, cluster) %>% 
  summarise(median = median(logcounts, na.rm = TRUE)) %>% 
  pivot_wider(names_from = "cluster", values_from = "median")

median_by_cluster_lineage_3_df %>% 
  dplyr::filter(cluster_1.2.1.3 > cluster_1.5) %>% 
  dplyr::filter(cluster_1.3 > cluster_1.2.1.3) %>% 
  dplyr::filter(cluster_1.4 > cluster_1.3) %>% 
  dplyr::filter(cluster_1.6 > cluster_1.4) %>% 
  vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/lineage_cluster_expression_plots/lineage_3_median_by_cluster_df.tsv", delim = "\t")

# Figures - lineage 4
sub_mnn_out_lineage_4 <- sub_mnn_out[, which(colnames(sub_mnn_out) %in% lineage_clusters_list$Lineage4$cell)]
colLabels(sub_mnn_out_lineage_4) <- colLabels(sub_mnn_out_lineage_4) %>% 
  forcats::fct_drop() %>% 
  forcats::fct_relevel(c("cluster_1.5", "cluster_1.3", "cluster_1.2.1.1", "cluster_1.2.1.3", "cluster_1.1"))

gene_expression_lineage_4_cluster_list <- purrr::map(seq_along(rownames(sub_mnn_out_lineage_4)), function(i) {
  print(i)
  plotExpression(
    object = sub_mnn_out_lineage_4, 
    features = rownames(sub_mnn_out_lineage_4)[i], 
    x = "label",
    colour_by = "label", 
    exprs_values = "logcounts"
    ) 
})

purrr::map(seq_along(gene_expression_lineage_4_cluster_list), function(i) {
  print(i)
  gene_expression_lineage_4_cluster_list[[i]] %>% 
    ggsave(filename = paste0("~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/lineage_cluster_expression_plots/lineage_4_", rownames(sub_mnn_out_lineage_4)[i], ".png"), width = 12, height = 7, dpi = 900)
})

expression_lineage_4_df <- sub_mnn_out_lineage_4@assays@data$logcounts %>%
  t() %>%
  as.data.frame() %>%
  mutate(cluster = sub_mnn_out_lineage_4$label) %>%
  rownames_to_column(var = "cell") %>%
  pivot_longer(cols = 2:1421, names_to = "gene", values_to = "logcounts")

median_by_cluster_lineage_4_df <- sub_mnn_out_lineage_4@assays@data$logcounts %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate(cluster = sub_mnn_out_lineage_4$label) %>% 
  rownames_to_column(var = "cell") %>% 
  pivot_longer(cols = 2:1421, names_to = "gene", values_to = "logcounts") %>% 
  group_by(gene, cluster) %>% 
  summarise(median = median(logcounts, na.rm = TRUE)) %>% 
  pivot_wider(names_from = "cluster", values_from = "median")

median_by_cluster_lineage_4_df %>% 
  dplyr::filter(cluster_1.3 > cluster_1.5) %>% 
  dplyr::filter(cluster_1.2.1.1 > cluster_1.3) %>% 
  dplyr::filter(cluster_1.2.1.3 > cluster_1.2.1.1) %>% 
  dplyr::filter(cluster_1.1 > cluster_1.2.1.3) %>% 
  vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/lineage_cluster_expression_plots/lineage_4_median_by_cluster_df.tsv", delim = "\t")
```


# DE analysis
```{r}
uncorrected_of_interest <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/uncorrected_of_interest_2nd_reboot_annotated_2022_10_21.qs", nthreads = 50)
# sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_reboot_annotated_sub_clusters_final.qs", nthreads = 50)
sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/sub_mnn_out_2nd_reboot_annotated_trajectory.qs", nthreads = 50)

# sub_mnn_out <- qs::qread(file = "/run/user/1002/gvfs/smb-share:server=primus.img.cas.cz,share=data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_trajectory.qs", nthreads = 50)

# sub_mnn_out %>% 
#   plotTSNE(colour_by = "batch") +
#   ggeasy::easy_text_size(size = 20) %>%  
#   ggsave(
#     filename = "/run/user/1002/gvfs/smb-share:server=primus.img.cas.cz,share=data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/tsne_plots/sub_mnn_out_coloured_by_batch.png",
#     device = "png", 
#     dpi = 900, 
#     width = 10, 
#     height = 10
#     )

data.frame(mnn_clusters = sub_mnn_out$label, uncorrected_clusters = uncorrected_of_interest$label) %>% 
  dplyr::filter(mnn_clusters == "cluster_1.2.1.3") %>% 
  group_by(uncorrected_clusters) %>% 
  tally()



# Marker gene detection
# pval.type="all" only considers genes that are differentially expressed in all pairwise comparisons involving the cluster of interest
# pval.type="all" uses an intersection-union test (Berger and Hsu 1996) where the combined p-value for each gene is the maximum of the  p-values from all pairwise comparisons.
# pval.type="some" applies the Holm-Bonferroni correction across its p-values and take the middle-most value as the combined  p -value

### Assigning cluster labels from markers
# Perform a gene set enrichment analysis on the marker genes defining each cluster
# markers <- findMarkers(sce_9.5, direction = "up", lfc = 1)

markers <- sub_mnn_out %>% findMarkers(test = "binom", direction = "up", lfc = 1, pval.type = "any")
# markers <- sub_mnn_out %>% findMarkers(test = "binom", direction = "up", lfc = 1, pval.type = "all")

markers_list <- purrr::map(seq_along(markers), function(i) markers[[i]])
markers_list <- markers_list %>% purrr::set_names(sort(unique(sub_mnn_out$label)))
# DE_list <- purrr::map(seq_along(markers), function(i) markers_list[[i]] %>% as.data.frame() %>% dplyr::filter(FDR <= 0.05 & summary.logFC >= 1.0)) %>% purrr::set_names(names(markers_list))
DE_list <- purrr::map(seq_along(markers), function(i) markers_list[[i]] %>% as.data.frame() %>% dplyr::filter(FDR <= 0.05 & summary.logFC >= 1.0 & Top <= 10))

# DE_list$cluster_1.1 %>% rownames_to_column(var = "gene") %>%  vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/endothelial_vs_all_DEGs.tsv", delim = "\t") 

gene_df <- markers_list %>% purrr::map(as.data.frame) %>% bind_rows(.id = "cluster") %>% rownames_to_column(var = "gene") %>% mutate_at(.vars = c("p.value", "FDR"),  .funs = . %>% formatC(digits = 3, format = "e")) %>% mutate_if(.predicate = is.numeric, .funs = . %>% formatC(digits = 3))

# gene_df %>% vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/c", delim = "\t")

DE_df <- DE_list %>% bind_rows(.id = "cluster") %>% rownames_to_column(var = "gene") %>% mutate_at(.vars = c("p.value", "FDR"),  .funs = . %>% formatC(digits = 3, format = "e")) %>% mutate_if(.predicate = is.numeric, .funs = . %>% formatC(digits = 3))

# DE_df %>% vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/DEG_results_df.tsv", delim = "\t")


top_genes <- purrr::map(seq_along(markers), function(i) head(rownames(markers_list[[i]]))) 

purrr::map(seq_along(markers), function(i) sub_mnn_out %>% 
             plotExpression(x = "label", features = top_genes[[i]],  colour_by = "label") +
             theme(axis.text.x = element_blank()) +
             ggeasy::easy_all_text_size(size = 20)
           )

# erythocytes (cl_2.1, cl_2.2, cl_2.3 vs all)
sub_mnn_out_v1 <- sub_mnn_out
sub_mnn_out_v1$label <- sub_mnn_out_v1$label %>% 
  dplyr::recode(cluster_2.1 = "cluster_2", cluster_2.2 = "cluster_2", cluster_2.3 = "cluster_2")
markers_v1 <- sub_mnn_out_v1 %>% findMarkers(test = "binom", direction = "up", lfc = 1, pval.type = "all")
markers_v1_list <- purrr::map(seq_along(markers_v1), function(i) markers_v1[[i]])
markers_v1_list <- markers_v1_list %>% purrr::set_names(sort(unique(sub_mnn_out_v1$label)))
DE_v1_list <- purrr::map(seq_along(markers_v1), function(i) markers_v1_list[[i]] %>% as.data.frame() %>% dplyr::filter(FDR <= 0.05 & summary.logFC >= 1.0)) %>% purrr::set_names(names(markers_v1_list))
DE_v1_list$cluster_2 %>% rownames_to_column(var = "gene") %>% vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/erythrocytes_vs_all_DEGs.tsv", delim = "\t")

# megakaryocytes (cl_1.6) vs 1.4
sub_mnn_out_v2 <- sub_mnn_out[, sub_mnn_out$label %>% str_which(pattern = "cluster_1.4|cluster_1.6")]
sub_mnn_out_v2$label <- sub_mnn_out_v2$label %>% fct_drop()
markers_v2 <- sub_mnn_out_v2 %>% findMarkers(test = "binom", direction = "up", lfc = 1, pval.type = "all")
markers_v2_list <- purrr::map(seq_along(markers_v2), function(i) markers_v2[[i]])
markers_v2_list <- markers_v2_list %>% purrr::set_names(sort(unique(sub_mnn_out_v2$label)))
DE_v2_list <- purrr::map(seq_along(markers_v2), function(i) markers_v2_list[[i]] %>% as.data.frame() %>% dplyr::filter(FDR <= 0.05 & summary.logFC >= 1.0)) %>% purrr::set_names(names(markers_v2_list))
DE_v2_list$cluster_1.4 %>% rownames_to_column(var = "gene") %>% vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/1.6_vs_1.4_DEGs.tsv", delim = "\t")

# myeloid (cl_1.2.2) vs 1.4
sub_mnn_out_v3 <- sub_mnn_out[, sub_mnn_out$label %>% str_which(pattern = "cluster_1.2.2|cluster_1.6")]
sub_mnn_out_v3$label <- sub_mnn_out_v3$label %>% fct_drop()
markers_v3 <- sub_mnn_out_v3 %>% findMarkers(test = "binom", direction = "up", lfc = 1, pval.type = "all")
markers_v3_list <- purrr::map(seq_along(markers_v3), function(i) markers_v3[[i]])
markers_v3_list <- markers_v3_list %>% purrr::set_names(sort(unique(sub_mnn_out_v3$label)))
DE_v3_list <- purrr::map(seq_along(markers_v3), function(i) markers_v3_list[[i]] %>% as.data.frame() %>% dplyr::filter(FDR <= 0.05 & summary.logFC >= 1.0)) %>% purrr::set_names(names(markers_v3_list))
DE_v3_list$cluster_1.2.2 %>% rownames_to_column(var = "gene") %>% vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/1.2.2_vs_1.4_DEGs.tsv", delim = "\t")
```

# hierarchical clustering on genes
```{r}
lineage_clusters_list

sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_trajectory.qs", nthreads = 50)

markers <- sub_mnn_out %>% findMarkers(test = "binom", direction = "up", lfc = 1, pval.type = "any")
markers_list <- purrr::map(seq_along(markers), function(i) markers[[i]])
markers_list <- markers_list %>% purrr::set_names(sort(unique(sub_mnn_out$label)))
DE_list <- purrr::map(seq_along(markers), function(i) markers_list[[i]] %>% as.data.frame() %>% dplyr::filter(FDR <= 0.05 & summary.logFC >= 1.0 & Top <= 10))

gene_df <- markers_list %>% purrr::map(as.data.frame) %>% bind_rows(.id = "cluster") %>% rownames_to_column(var = "gene") %>% mutate_at(.vars = c("p.value", "FDR"),  .funs = . %>% formatC(digits = 3, format = "e")) %>% mutate_if(.predicate = is.numeric, .funs = . %>% formatC(digits = 3))


gene_df %>% 
  ggplot()



gene_df %>% 
  unite(gene:cluster, col = gene_cluster, sep = "_") %>% 
  dplyr::select(-c(Top, p.value, FDR)) %>% 
  column_to_rownames(var = "gene_cluster") %>% 
  mutate_all(as.numeric)

library(parallelDist)
library(ggdendro)

gene_dist <- gene_df %>% 
  dist()

gene_hc <- gene_dist %>% hclust()

ggdendrogram(gene_hc, rotate = FALSE, size = 2)
```




# Heatmap clusters in lineage 2 vs genes
```{r}
lineage2_clusters <- pseudo_paths %>% 
  as.data.frame() %>% 
  mutate(cluster = as.character(colLabels(sub_mnn_out))) %>% 
  drop_na(Lineage2) %>% 
  group_by(cluster) %>% 
  tally() %>% 
  pull(cluster)

gene_df %>% 
  dplyr::filter(cluster %in% lineage2_clusters) %>% 
  mutate(gene = gene %>% str_remove(pattern = "\\.\\.\\..*")) %>% 
  dplyr::select()
  ggplot(aes(x = cluster, y = gene, fill = summary.logFC)) +
  geom_tile()
```



# ORA
```{r}
sce_E9.5 <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sce_9.5_reboot_no_histone_no_doublet_clustered_annotated_2022_10_20.qs", nthreads = 50)
# sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_trajectory.qs", nthreads = 50)

markers <- sce_E9.5 %>% findMarkers(test = "binom", direction = "up", lfc = 1, pval.type = "any")
# markers <- sub_mnn_out %>% findMarkers(test = "binom", direction = "up", lfc = 1, pval.type = "any")

markers_list <- purrr::map(seq_along(markers), function(i) markers[[i]])
markers_list <- markers_list %>% purrr::set_names(sort(unique(sce_E9.5$label)))
# markers_list <- markers_list %>% purrr::set_names(sort(unique(sub_mnn_out$label)))
DE_list <- purrr::map(seq_along(markers), function(i) markers_list[[i]] %>% as.data.frame() %>% dplyr::filter(FDR <= 0.05 & summary.logFC >= 1.0 & Top <= 10)) %>% 
  purrr::set_names(sce_E9.5$label %>% unique() %>% sort())

DE_list <- DE_list %>%
  purrr::map(mutate_at, .vars = c("p.value", "FDR"), .funs = formatC, digits = 2, format = "e") %>% 
  purrr::map(mutate_if, .predicate = is.numeric, .funs = round, digits = 2) 

DE_list <- DE_list %>%
  purrr::map(rownames_to_column, var = "gene")
  
DE_df <- DE_list %>%
  bind_rows(.id = "reference_cluster") %>% 
  vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/DE_results_sce_E9.5", delim = "\t")

DEG_list <- DE_list %>% 
  # purrr::map(rownames_to_column, var = "gene") %>% 
  purrr::map(pull, gene) %>% 
  purrr::map(str_to_upper) %>% 
  purrr::map(sort) %>% 
  # purrr::set_names(sub_mnn_out$label %>% unique() %>% sort())
  purrr::set_names(sce_E9.5$label %>% unique() %>% sort())

KEGG_immune_system_pathways <- c("Hematopoietic cell lineage", "Complement and coagulation cascades", "Platelet activation", "Neutrophil extracellular trap formation", "Toll-like receptor signaling pathway", "Toll and Imd signaling pathway", "NOD-like receptor signaling pathway", "RIG-I-like receptor signaling pathway", "Cytosolic DNA-sensing pathway", "C-type lectin receptor signaling pathway", "Natural killer cell mediated cytotoxicity", "Antigen processing and presentation", "T cell receptor signaling pathway", "Th1 and Th2 cell differentiation", "Th17 cell differentiation", "IL-17 signaling pathway", "B cell receptor signaling pathway", "Fc epsilon RI signaling pathway", "Fc gamma R-mediated phagocytosis", "Leukocyte transendothelial migration", "Intestinal immune network for IgA production", "Chemokine signaling pathway")

ora_KEGG <- DEG_list %>% 
  # ora_analysis(database = KEGG_2019_Mouse, subset = KEGG_immune_system_pathways) 
  ora_analysis(database = KEGG_2019_Mouse) 

# ora_KEGG@compareClusterResult <- ora_KEGG@compareClusterResult %>% 
#   dplyr::filter(ID %in% KEGG_immune_system_pathways)
  
ora_KEGG %>% dotplot()

ora_KEGG@compareClusterResult %>% 
  mutate_at(.vars = c("pvalue", "p.adjust", "qvalue"), .funs = formatC, digits = 2, format = "e") %>% 
  vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/ora_KEGG_sce_E9.5_results_table.tsv", delim = "\t")

gmt_df <- GO_Biological_Process_2021 %>% clusterProfiler::read.gmt()
terms_list <- gmt_df %>% split(f = .$term) 
GO_terms_df <- data.frame(GO_terms = names(terms_list))
GO_terms_df %>% dplyr::filter(GO_terms  %>% str_detect(pattern = "antibody|antigen|attack|B cell|basophil|clearance|complement|crystal|cytokine|defense|dendritic|eosinophil|erythrocyte|Fc-gamma |granuloma|hemocyte|hemopoiesis|hypersensitivity|immune|immunoglobulin|immunological|interleukin|isotype|killer|lamellocyte|leukocyte|lymph|lymphoid|lymphocyte|M cells|macrophage|mast|MDA-5|MHC|microglial|monocyte|mononuclear|myeloid|neutrophil|oligomerization|opsonization|pattern|plasma|plasmacytoid|recognition|resistance|RIG-I|STING|T cell|T-helper|tolerance|toll"))

ora_GO_BP <- DEG_list %>% 
  ora_analysis(database = GO_Biological_Process_2021) 

ora_GO_BP@compareClusterResult <- ora_GO_BP@compareClusterResult %>% 
  dplyr::filter(ID %>% str_detect(pattern = "antibody|antigen|attack|B cell|basophil|clearance|complement|crystal|cytokine|defense|dendritic|eosinophil|erythrocyte|Fc-gamma |granuloma|hemocyte|hemopoiesis|hypersensitivity|immune|immunoglobulin|immunological|interleukin|isotype|killer|lamellocyte|leukocyte|lymph|lymphoid|lymphocyte|M cells|macrophage|mast|MDA-5|MHC|microglial|monocyte|mononuclear|myeloid|neutrophil|oligomerization|opsonization|pattern|plasma|plasmacytoid|recognition|resistance|RIG-I|STING|T cell|T-helper|tolerance|toll"))
  
ora_GO_BP %>% dotplot(showCategory = 2)

ora_GO_MF <- DEG_list %>% 
  ora_analysis(database = GO_Molecular_Function_2021) 
  
ora_GO_MF %>% dotplot(showCategory = 2)

ora_WikiPathways <- DEG_list %>% 
  ora_analysis(database = WikiPathways_2019_Mouse) 
  
ora_WikiPathways %>% dotplot()

ora_GO_CC <- DEG_list %>% 
  ora_analysis(database = GO_Cellular_Component_2021) 
  
ora_GO_CC %>% dotplot(showCategory = 2)

ora_BioPlanet <- DEG_list %>% 
  ora_analysis(database = BioPlanet_2019) 
  
ora_BioPlanet %>% dotplot(showCategory = 2)
```

# GSEA
```{r}
library(clusterProfiler)
library(enrichplot)

# Import Mouse Gene Ids dictionary
gene_id <- rowData(sub_mnn_out) %>% as.data.frame() %>% rownames_to_column(var = "gene") %>% pull(gene) %>% sort()
gene_id_dict = AnnotationDbi::select(org.Mm.eg.db, keys = gene_id, columns = c("SYMBOL","GENENAME", "ENTREZID"),  keytype = "SYMBOL")

ranked_genes_list <- markers_list %>% purrr::map(function(x) x %>% as.data.frame() %>% rownames_to_column(var = "gene") %>% dplyr::select(gene, summary.logFC) %>% arrange(desc(summary.logFC)))

ranked_genes_list <- ranked_genes_list %>% purrr::map(full_join, gene_id_dict, by = c("gene" = "SYMBOL"))

gene_list <- purrr::map(seq_along(ranked_genes_list), function(i) ranked_genes_list[[i]] %>% pull(summary.logFC) %>% purrr::set_names(ranked_genes_list[[i]]$ENTREZID)) %>% purrr::set_names(names(ranked_genes_list))

gse_KEGG_results_list <- purrr::map(seq_along(gene_list), function(i) {
  print(i)
  gseKEGG(geneList = gene_list[[i]], exponent = 1, organism = "mmu", eps = 0)
} ) %>% purrr::set_names(source)

gseKEGG(geneList = gene_list[[11]], exponent = 1, organism = "mmu", eps = 0)
```



# Gene expression plots
```{r}
uncorrected_of_interest <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/uncorrected_of_interest_2nd_reboot_annotated_2022_10_21.qs", nthreads = 50)
sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_trajectory.qs", nthreads = 50)

genes <- c("Dapp1", "Ctla2a", "Angpt1", "Samsn1", "Gucy1a1", "Ptpn7", "Plek", "Alox5ap", "Meis1", "Mrvi1", "Myb", "Klhl6", "Laptm5", "Fcer1g", "Treml2", "Coro1a", "Was", "Celf2", "Cdc42ep5", "Ptgs1", "Sla2", "Treml1", "Spi1", "Hapln1", "Lcp2", "Adgrg1", "Gm4779", "Pf4", "Sla", "Fli1", "Rab27b", "Rhoj")

genes <- c("Gp9", "Gp1bb", "Sarch", "F2rl2", "Tgfb1i1")

genes <- rownames(sub_mnn_out)[rownames(sub_mnn_out) %>% str_which(pattern = "mt-")]

genes <- rownames(sub_mnn_out)

genes <- genes %>% sort()
genes <- genes %>% intersect(rownames(uncorrected_of_interest))

# gene_tsne_umap_plots <- function(gene, uncorrected_sce, batch_corrected_sce){
#   
#   tsne_df <- batch_corrected_sce@int_colData@listData$reducedDims$TSNE %>% as.data.frame()
#   umap_df <- batch_corrected_sce@int_colData@listData$reducedDims$UMAP %>% as.data.frame()
#   gene_expression <- uncorrected_sce@assays@data$logcounts[gene,] %>% as.numeric() %>% 
#   batch <- batch_corrected_sce$batch
# 
# gene_plots <- list(
#   # tsne_plot = tsne_df %>% ggplot(aes(x = V1, y = V2, color = gene_expression, shape = batch)) +
#   tsne_plot = tsne_df %>% ggplot(aes(x = V1, y = V2, color = gene_expression)) +
#     geom_point() +
#     scale_color_continuous(low = "grey", high = "red", guide = "colourbar", aesthetics = "colour") +
#     theme_bw() + 
#     labs(x = "TSNE 1", y = "TSNE 2", title = gene),
#   
#   umap_plot = umap_df %>% ggplot(aes(x = V1, y = V2, color = gene_expression, shape = batch)) +
#     geom_point() +
#     scale_color_continuous(low = "grey", high = "red", guide = "colourbar", aesthetics = "colour") +
#     theme_bw() + 
#     labs(x = "TSNE 1", y = "TSNE 2", title = gene) +
#     easy_plot_title_size(size = 20)
# ) %>% 
#   purrr::set_names("tsne", "umap")
# return(gene_plots)
# }

gene_tsne_umap_plots <- function(gene, uncorrected_sce, batch_corrected_sce){
  
  tsne_df <- batch_corrected_sce@int_colData@listData$reducedDims$TSNE %>% 
    as.data.frame() %>% 
    mutate(gene_expression = sce@assays@data$logcounts[gene,] %>% as.numeric()) %>% 
    mutate(batch = batch_corrected_sce$batch) %>% 
    arrange(gene_expression)
  
  umap_df <- batch_corrected_sce@int_colData@listData$reducedDims$UMAP %>% 
    as.data.frame() %>% 
    mutate(gene_expression = sce@assays@data$logcounts[gene,] %>% as.numeric()) %>% 
    mutate(batch = batch_corrected_sce$batch) %>% 
    arrange(gene_expression)

gene_plots <- list(
  # tsne_plot = tsne_df %>% ggplot(aes(x = V1, y = V2, color = gene_expression, shape = batch)) +
  tsne_plot = tsne_df %>% ggplot(aes(x = V1, y = V2, color = gene_expression)) +
    geom_point() +
    scale_color_continuous(low = "grey", high = "red", guide = "colourbar", aesthetics = "colour") +
    theme_bw() + 
    labs(x = "TSNE 1", y = "TSNE 2", title = gene),
  
  umap_plot = umap_df %>% ggplot(aes(x = V1, y = V2, color = gene_expression, shape = batch)) +
    geom_point() +
    scale_color_continuous(low = "grey", high = "red", guide = "colourbar", aesthetics = "colour") +
    theme_bw() + 
    labs(x = "TSNE 1", y = "TSNE 2", title = gene) +
    easy_plot_title_size(size = 20)
) %>% 
  purrr::set_names("tsne", "umap")
return(gene_plots)
}

plots_list <- purrr::map(seq_along(genes), function(i){
  print(i)
  gene_tsne_umap_plots(gene = genes[i], 
                       uncorrected_sce = uncorrected_of_interest, 
                       batch_corrected_sce = sub_mnn_out
                       )
  }) %>% purrr::set_names(genes)

# plots_list %>% qs::qsave(file = paste0("~/PRIMUS/home/trufenc/merged_clusters_all_genes_plots_list.qs"), nthreads = 50)
# plots_list <- qs::qread(file = paste0("~/PRIMUS/home/trufenc/merged_clusters_all_genes_plots_list.qs"), nthreads = 50)

purrr::map(seq_along(plots_list), function(i){
  print(i)
 plots_list[[i]]$tsne %>% ggsave(filename = paste0("~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/merged_datasets_genes_expression_tsne_plots/merged_dataset_", genes[i] , "_tsne_plot.png"))
 })

purrr::map(seq_along(plots_list), function(i) {
  print(i)
  plots_list[[i]]$umap %>% ggsave(filename = paste0("~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/merged_datasets_genes_expression_umap_plots/merged_dataset_", genes[i] , "_umap_plot.png"))
  })


plot_grid(plots_list[[1]]$tsne,
          plots_list[[2]]$tsne,
          plots_list[[3]]$tsne,
          plots_list[[4]]$tsne,
          ncol = 2
          )

plot_grid(plots_list[[5]]$tsne,
          plots_list[[6]]$tsne,
          plots_list[[7]]$tsne,
          plots_list[[8]]$tsne,
          ncol = 2
          )

plot_grid(plots_list[[9]]$tsne,
          plots_list[[10]]$tsne,
          plots_list[[11]]$tsne,
          plots_list[[12]]$tsne,
          ncol = 2
          )

plot_grid(plots_list[[13]]$tsne,
          plots_list[[14]]$tsne,
          plots_list[[15]]$tsne,
          plots_list[[16]]$tsne,
          ncol = 2
          )

plot_grid(plots_list[[17]]$tsne,
          plots_list[[18]]$tsne,
          plots_list[[19]]$tsne,
          plots_list[[20]]$tsne,
          ncol = 2
          )

plot_grid(plots_list[[21]]$tsne,
          plots_list[[22]]$tsne,
          plots_list[[23]]$tsne,
          plots_list[[24]]$tsne,
          ncol = 2
          )

plot_grid(plots_list[[25]]$tsne,
          plots_list[[26]]$tsne,
          plots_list[[27]]$tsne,
          plots_list[[28]]$tsne,
          ncol = 2
          )

plot_grid(plots_list[[29]]$tsne,
          plots_list[[30]]$tsne,
          plots_list[[31]]$tsne,
          plots_list[[32]]$tsne,
          ncol = 2
          )

cowplot::plot_grid(
  sub_mnn_out %>%
    plotTSNE(colour_by = I(dplyr::if_else(condition = sub_mnn_out$label %in% c("E7.5_2"), true = "E7.5_2", false = NA_character_))),
   sub_mnn_out %>%
    plotTSNE(colour_by = I(dplyr::if_else(condition = sub_mnn_out$label %in% c("E7.5_3"), true = "E7.5_3", false = NA_character_))),
   sub_mnn_out %>%
    plotTSNE(colour_by = I(dplyr::if_else(condition = sub_mnn_out$label %in% c("E7.5_4"), true = "E7.5_4", false = NA_character_))),
   sub_mnn_out %>%
    plotTSNE(colour_by = I(dplyr::if_else(condition = sub_mnn_out$label %in% c("E7.5_5"), true = "E7.5_5", false = NA_character_))),
   sub_mnn_out %>%
    plotTSNE(colour_by = I(dplyr::if_else(condition = sub_mnn_out$label %in% c("E7.5_6"), true = "E7.5_6", false = NA_character_))),
   sub_mnn_out %>%
    plotTSNE(colour_by = I(dplyr::if_else(condition = sub_mnn_out$label %in% c("E7.5_7"), true = "E7.5_7", false = NA_character_))),
   sub_mnn_out %>%
    plotTSNE(colour_by = I(dplyr::if_else(condition = sub_mnn_out$label %in% c("E7.5_8"), true = "E7.5_8", false = NA_character_))),
  ncol = 4
)

cowplot::plot_grid(
  sub_mnn_out %>%
    plotTSNE(colour_by = I(dplyr::if_else(condition = sub_mnn_out$label %in% c("E8.5_1"), true = "E8.5_1", false = NA_character_))),
  sub_mnn_out %>%
    plotTSNE(colour_by = I(dplyr::if_else(condition = sub_mnn_out$label %in% c("E8.5_2"), true = "E8.5_2", false = NA_character_))),
   sub_mnn_out %>%
    plotTSNE(colour_by = I(dplyr::if_else(condition = sub_mnn_out$label %in% c("E8.5_3"), true = "E8.5_3", false = NA_character_)))
)

    
```

# expressed gene pair tsne plot
```{r}
uncorrected_of_interest <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/uncorrected_of_interest_2nd_reboot_annotated_2022_10_21.qs", nthreads = 50)
sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_trajectory.qs", nthreads = 50)

uncorrected_sce <- uncorrected_of_interest
batch_corrected_sce <- sub_mnn_out

markers <- sub_mnn_out %>% findMarkers(test = "binom", direction = "up", lfc = 1, pval.type = "any")
markers_list <- purrr::map(seq_along(markers), function(i) markers[[i]])
markers_list <- markers_list %>% purrr::set_names(sort(unique(sub_mnn_out$label)))
DE_list <- purrr::map(seq_along(markers), function(i) markers_list[[i]] %>% as.data.frame() %>% dplyr::filter(FDR <= 0.05 & summary.logFC >= 1.0 & Top <= 10)) %>% purrr::set_names(names(markers_list))

gene_v <- DE_list$cluster_1.2.1.3 %>% rownames()
gene_pair_df <- expand.grid(gene_1 = gene_v, gene_2 = gene_v) %>% 
  dplyr::filter(gene_1 != gene_2)

for (i in 1:nrow(gene_pair_df)){
    gene_pair_df[i, ] = sort(gene_pair_df[i, ])
}

gene_pair_df <- gene_pair_df[!duplicated(gene_pair_df),]

gene_pair_df <- gene_pair_df %>% mutate_all(as.character)

gene_pair <- c("Lyve1", "Alox5ap")

# https://stackoverflow.com/questions/48101617/dplyr-conditional-column-ifelse-with-vector-input
expressed_gene_pair_tsne_umap_plots <- function(gene_pair, uncorrected_sce, batch_corrected_sce){
  
  tsne_df <- batch_corrected_sce@int_colData@listData$reducedDims$TSNE %>% as.data.frame()
  umap_df <- batch_corrected_sce@int_colData@listData$reducedDims$UMAP %>% as.data.frame()
  gene_expression <- uncorrected_sce@assays@data$logcounts[gene_pair,] %>% 
    t() %>% 
    as.data.frame() %>% 
    mutate(gene_1_expressed = dplyr::if_else(condition = !!sym(gene_pair[1]) > 0, true = TRUE, false = FALSE)) %>% 
    mutate(gene_2_expressed = dplyr::if_else(condition = !!sym(gene_pair[2]) > 0, true = TRUE, false = FALSE)) %>% 
    mutate(both_genes_expressed = gene_1_expressed & gene_2_expressed)

gene_plots <- list(
  tsne_plot = tsne_df %>% ggplot(aes(x = V1, y = V2, color = gene_expression$both_genes_expressed)) +
    geom_point() +
    scale_color_discrete(type = c("grey", "blue")) +
    theme_bw() + 
    labs(x = "TSNE 1", y = "TSNE 2", title = paste0(gene_pair, collapse = "_&_"), color = "both_genes_expressed"),
  
  umap_plot = umap_df %>% ggplot(aes(x = V1, y = V2, color = gene_expression$both_genes_expressed)) +
    geom_point() +
    scale_color_discrete(type = c("grey", "blue")) +
    theme_bw() + 
    labs(x = "UMAP 1", y = "UMAP 2", title = paste0(gene_pair, collapse = "_&_"), color = "both_genes_expressed") +
    easy_plot_title_size(size = 20)
) %>% 
  purrr::set_names("tsne", "umap")
return(gene_plots)
}

expressed_gene_pair_tsne_umap_plots(gene_pair = c("Lyve1", "Alox5ap"), uncorrected_sce = uncorrected_of_interest, batch_corrected_sce = sub_mnn_out)
expressed_gene_pair_tsne_umap_plots(gene_pair = c("Lyve1", "Fcer1g"), uncorrected_sce = uncorrected_of_interest, batch_corrected_sce = sub_mnn_out)

plot_list <- purrr::map(seq_len(nrow(gene_pair_df)), function(i) {
  print(i) 
  expressed_gene_pair_tsne_umap_plots(gene_pair = c(gene_pair_df$gene_1[i], gene_pair_df$gene_2[i]), uncorrected_sce = uncorrected_of_interest, batch_corrected_sce = sub_mnn_out)
  })%>% purrr::set_names(paste0(c(gene_pair_df$gene_1, gene_pair_df$gene_2), collapse = "_&_"))

purrr::map(seq_along(plot_list), function(i) {
  print(i)
  plot_list[[i]]$tsne %>% 
             ggsave(filename = paste0("~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/gene_pair_tsne_plots/gene_pair_", paste0(c(gene_pair_df$gene_1[i], gene_pair_df$gene_2[i]), collapse = "_&_") , "_tsne_plot.png"))
  })
```


# Pairwise gene correlation
```{r}
uncorrected_of_interest <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/uncorrected_of_interest_2nd_reboot_annotated_2022_10_21.qs", nthreads = 50)
sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_trajectory.qs", nthreads = 50)

markers <- sub_mnn_out %>% findMarkers(test = "binom", direction = "up", lfc = 1, pval.type = "any")
markers_list <- purrr::map(seq_along(markers), function(i) markers[[i]])
markers_list <- markers_list %>% purrr::set_names(sort(unique(sub_mnn_out$label)))
DE_list <- purrr::map(seq_along(markers), function(i) markers_list[[i]] %>% as.data.frame() %>% dplyr::filter(FDR <= 0.05 & summary.logFC >= 1.0 & Top <= 10)) %>% purrr::set_names(names(markers_list))

gene_v <- DE_list$cluster_1.2.1.3 %>% rownames()
gene_pair_df <- expand.grid(gene_1 = gene_v, gene_2 = gene_v) %>% 
  dplyr::filter(gene_1 != gene_2)

for (i in 1:nrow(gene_pair_df)){
    gene_pair_df[i, ] = sort(gene_pair_df[i, ])
}

gene_pair_df <- gene_pair_df[!duplicated(gene_pair_df),]

gene_pair_df <- gene_pair_df %>% mutate_all(as.character)

gene_pair <- c("Lyve1", "Alox5ap")

uncorrected_sce_subset <- uncorrected_sce[, sub_mnn_out$label %>% str_which(pattern = "cluster_1.2.1.3")]

# gene_expression <- uncorrected_sce_subset@assays@data$logcounts[gene_pair,] %>%
#   t() %>%
#   as.data.frame() %>%
#   mutate(gene_1_expressed = dplyr::if_else(condition = Lyve1 > 0, true = TRUE, false = FALSE)) %>%
#   mutate(gene_2_expressed = dplyr::if_else(condition = Alox5ap > 0, true = TRUE, false = FALSE)) %>%
#   mutate(both_genes_expressed = gene_1_expressed & gene_2_expressed)
# 
# 
# cor.test(x = gene_expression$Lyve1, y = gene_expression$Alox5ap)  
# 
# gene_expression %>% 
#   ggplot(aes(x = Lyve1, y = Alox5ap)) +
#   geom_point() +
#   geom_smooth() +
#   theme_bw() +
#   ggeasy::easy_all_text_size(size = 20)


calculate_pairwise_gene_cor <- function(gene_pair, sce){
  gene_expression <- uncorrected_sce_subset@assays@data$logcounts[gene_pair,] %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate(gene_1_expressed = dplyr::if_else(condition = !!sym(gene_pair[1]) > 0, true = TRUE, false = FALSE)) %>% 
  mutate(gene_2_expressed = dplyr::if_else(condition = !!sym(gene_pair[2]) > 0, true = TRUE, false = FALSE)) %>% 
  mutate(both_genes_expressed = gene_1_expressed & gene_2_expressed)
  
  cor_result <- cor.test(x = gene_expression[[gene_pair[1]]], y = gene_expression[[gene_pair[2]]])
  return(cor_result)
}

uncorrected_sce_subset <- uncorrected_sce[, sub_mnn_out$label %>% str_which(pattern = "cluster_1.2.1.3")]

calculate_pairwise_gene_cor(gene_pair = c("Lyve1", "Alox5ap"), sce = uncorrected_sce_subset)

cor_list <- purrr::map(seq_len(nrow(gene_pair_df)), function(i) {
  print(i) 
  calculate_pairwise_gene_cor(gene_pair = c(gene_pair_df$gene_1[i], gene_pair_df$gene_2[i]), sce = uncorrected_sce_subset)
  }) 

cor_df <- cor_list %>% purrr::map(broom::tidy) %>% bind_rows() 

cor_df <- cor_df %>% 
  mutate_at(.vars = c("estimate", "statistic", "conf.low", "conf.high"), .funs = . %>% round(digits = 2)) %>% 
  mutate_at(.vars = c("p.value"), .funs = . %>% formatC(digits = 2, format = "e")) %>% 
  bind_cols(gene_pair_df) %>% 
  arrange(desc(estimate))

calculate_gene_cor_per_cluster <- function(gene_pair_df, sce, cluster){
  
  sce <- sce[, sub_mnn_out$label %>% str_which(pattern = cluster)]

  cor_list <- purrr::map(seq_len(nrow(gene_pair_df)), function(i) {
    print(i) 
    calculate_pairwise_gene_cor(gene_pair = c(gene_pair_df$gene_1[i], gene_pair_df$gene_2[i]), sce = sce)
    }) 
  
  cor_df <- cor_list %>% purrr::map(broom::tidy) %>% bind_rows() 
  
  cor_df <- cor_df %>% 
    mutate_at(.vars = c("estimate", "statistic", "conf.low", "conf.high"), .funs = . %>% round(digits = 2)) %>% 
    mutate_at(.vars = c("p.value"), .funs = . %>% formatC(digits = 2, format = "e")) %>% 
    bind_cols(gene_pair_df) %>% 
    arrange(desc(estimate))
  
  return(cor_df)
}

cor_df <- calculate_gene_cor_per_cluster(gene_pair_df = gene_pair_df, sce = uncorrected_sce, cluster = "cluster_1.2.1.3")

cor_df %>% vroom::vroom_write(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/correlation_results/paiwise_gene_cor_cluster_1.2.1.3.tsv", delim = "\t")


# cor plot
pairwise_gene_cor_plot <- function(gene_pair, sce){
  gene_expression <- uncorrected_sce_subset@assays@data$logcounts[gene_pair,] %>% 
  t() %>% 
  as.data.frame() 
  
  cor_plot <- gene_expression %>% 
    ggplot(aes(x = !!sym(gene_pair[1]), y = !!sym(gene_pair[2]))) +
    geom_point() +
    geom_smooth() +
    theme_bw() +
    ggeasy::easy_all_text_size(size = 20)
  
  return(cor_plot)
}

uncorrected_sce_subset <- uncorrected_sce[, sub_mnn_out$label %>% str_which(pattern = "cluster_1.2.1.3")]

pairwise_gene_cor_plot(gene_pair = c("Lyve1", "Alox5ap"), sce = uncorrected_sce_subset)

gene_pair_df <- gene_pair_df %>% 
  unite(gene_1:gene_2, col = "gene_pair", sep = "_x_", remove = FALSE) %>% 
  dplyr::select(gene_1, gene_2, gene_pair)

cor_plot_list <- purrr::map(seq_len(nrow(gene_pair_df)), function(i) {
    print(i) 
    pairwise_gene_cor_plot(gene_pair = c(gene_pair_df$gene_1[i], gene_pair_df$gene_2[i]), sce = uncorrected_sce_subset)
    }) %>% purrr::set_names(gene_pair_df$gene_pair)

purrr::map(seq_along(cor_plot_list), function(i){
  print(i)
  cor_plot_list[[i]] %>% ggsave(filename = paste0("~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/correlation_results/", names(cor_plot_list)[i],"_cluster_1.2.1.3.png"), device = "png", dpi = 900, width = 9, height = 9)
  })
```
    
# Number of double positive cells per cluster - Misa
```{r}
# uncorrected_of_interest <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/uncorrected_of_interest_2nd_reboot_annotated_2022_10_21.qs", nthreads = 50)
# sub_mnn_out <- qs::qread(file = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_trajectory.qs", nthreads = 50)

# uncorrected_of_interest <- qs::qread(file = "/run/user/1002/gvfs/smb-share:server=primus.img.cas.cz,share=data/83_BIOINFORMATICS/Kadu/uncorrected_of_interest_2nd_reboot_annotated_2022_10_21.qs", nthreads = 50)
sub_mnn_out <- qs::qread(file = "/run/user/1002/gvfs/smb-share:server=primus.img.cas.cz,share=data/83_BIOINFORMATICS/Kadu/sub_mnn_out_2nd_reboot_annotated_trajectory.qs", nthreads = 50)


markers <- sub_mnn_out %>% findMarkers(test = "binom", direction = "up", lfc = 1, pval.type = "any")
markers_list <- purrr::map(seq_along(markers), function(i) markers[[i]])
markers_list <- markers_list %>% purrr::set_names(sort(unique(sub_mnn_out$label)))
DE_list <- purrr::map(seq_along(markers), function(i) markers_list[[i]] %>% as.data.frame() %>% dplyr::filter(FDR <= 0.05 & summary.logFC >= 1.0 & Top <= 10)) %>% purrr::set_names(names(markers_list))

gene_v <- DE_list$cluster_1.2.1.3 %>% rownames()
gene_pair_df <- expand.grid(gene_1 = gene_v, gene_2 = gene_v) %>% 
  dplyr::filter(gene_1 != gene_2)

for (i in 1:nrow(gene_pair_df)){
    gene_pair_df[i, ] = sort(gene_pair_df[i, ])
}

gene_pair_df <- gene_pair_df[!duplicated(gene_pair_df),]

gene_pair_df <- gene_pair_df %>% mutate_all(as.character)

gene_pair_df <- gene_pair_df %>% 
  unite(gene_1:gene_2, col = "gene_pair", sep = "-", remove = FALSE) %>% 
  dplyr::select(gene_1, gene_2, gene_pair)

gene_pair <- c("Lyve1", "Alox5ap")

gene_df_sub_mnn_out <- sub_mnn_out@assays@data$logcounts %>%
  t() %>%
  as.data.frame() %>%
  dplyr::select(Fli1, Hapln1) %>%
  mutate(cluster = as.character(colLabels(sub_mnn_out))) %>%
  mutate(Fli1_pos = dplyr::if_else(condition = Fli1 > 0, true = "pos", false = "neg")) %>%
  mutate(Hapln1_pos = dplyr::if_else(condition = Hapln1 > 0, true = "pos", false = "neg"))

gene_df_sub_mnn_out %>%
  group_by(cluster, Fli1_pos, Hapln1_pos) %>%
  tally() %>%
  ungroup() %>% 
  group_by(cluster) %>% 
  mutate(percentage = 100 * n/sum(n)) %>% 
  ungroup() %>% 
  mutate_at(.vars = c("percentage"), .funs = round, digits = 2) %>% 
  print(n = Inf)

# https://stackoverflow.com/questions/46740807/dynamic-group-by-argument-based-on-a-list-r-for-loop-function
# https://stackoverflow.com/questions/26003574/use-dynamic-variable-names-in-dplyr
# https://stackoverflow.com/questions/60895225/dplyr-mutate-with-an-ifelse-using-dynamic-variable-names
find_n_double_positives <- function(sce, genes){
  library(tidyverse)
  library(rlang)
  gene_df <- sce@assays@data$counts %>% 
    t() %>% 
    as.data.frame() %>% 
    dplyr::select(genes[1], genes[2]) %>% 
    mutate(cluster = as.character(colLabels(sce))) %>% 
    mutate(!!paste0(genes[1], "_expression") := dplyr::if_else(condition = !!sym(genes[1]) > 0, true = "pos", false = "neg")) %>% 
    mutate(!!paste0(genes[2], "_expression") := dplyr::if_else(condition = !!sym(genes[2]) > 0, true = "pos", false = "neg"))
  
  gene_df %>% 
    group_by(cluster, !!!syms(paste0(genes[1], "_expression")), !!!syms(paste0(genes[2], "_expression"))) %>% 
    tally() %>%
    ungroup() %>% 
    group_by(cluster) %>% 
    mutate(percentage = 100 * n/sum(n)) %>% 
    ungroup() %>% 
    mutate_at(.vars = c("percentage"), .funs = round, digits = 2) %>%
    print(n = Inf)
  }

genes <- c("Fli1", "Hapln1")

sub_mnn_out %>% find_n_double_positives(genes = c("Fli1", "Hapln1")) %>% dplyr::filter(cluster == "cluster_1.2.1.3")


double_positive_list <- purrr::map(seq_len(nrow(gene_pair_df)), function(i) {
  print(i)
  sub_mnn_out %>% find_n_double_positives(genes = c(gene_pair_df$gene_1[i], gene_pair_df$gene_2[i])) 
  } ) %>% purrr::set_names(gene_pair_df$gene_pair)

double_positive_df <- double_positive_list %>% 
  purrr::map(function(x) x %>% purrr::set_names("cluster", "gene_1", "gene_2", "n", "percentage") %>% 
               dplyr::filter(gene_1 == "pos" & gene_2 == "pos") %>% 
               dplyr::select(cluster, percentage)
  ) %>% bind_rows(.id = "gene_pair")

# double_positive_df %>% vroom::vroom_write(file = "/run/user/1002/gvfs/smb-share:server=primus.img.cas.cz,share=data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/double_positive_results/double_positive_longer_df.tsv", delim = "\t")

double_positive_df <- read_delim("/run/user/1002/gvfs/smb-share:server=primus.img.cas.cz,share=data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/double_positive_results/double_positive_longer_df.tsv", delim = "\t", escape_double = FALSE, trim_ws = TRUE)

double_positive_df %>% 
  pivot_wider(names_from = "cluster", values_from = "percentage") %>% 
  mutate_at(.vars = 2:ncol(.), .funs = . %>% replace_na(replace = 0)) %>% 
  vroom::vroom_write(file = "/run/user/1002/gvfs/smb-share:server=primus.img.cas.cz,share=data/83_BIOINFORMATICS/Kadu/scRNAseq_Immunology/Results/double_positive_results/double_positive_wider_df.tsv", delim = "\t")

double_positive_df %>% 
  dplyr::filter(cluster == "cluster_1.2.1.3") %>% 
  arrange(desc(percentage))

double_positive_df %>% 
  split(f = .$cluster) %>% 
  purrr::map(arrange, percentage)

double_positive_df %>% 
  split(f = .$cluster) %>% 
  purrr::map(arrange, desc(percentage))
```



