---
title: "scRNAseq_Shweta"
author: "Carlos Eduardo Madureira Trufen"
date: "2023-08-10"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

```{r}
# library(DropletUtils)

library(AUCell)
library(aplot)
library(batchelor)
# library(beachmat)
# library(biomaRt)
library(bluster)
library(cluster)
library(celldex)
library(corral)
library(cowplot)
# library(clusterProfiler)
library(clustree)
library(DESeq2)
library(edgeR) 
library(factoextra)
library(ggeasy)
library(ggpubr)
library(GSEABase)
library(HDF5Array)
library(kBET)
library(limma)
library(Matrix)
library(mmtable2)
library(org.Mm.eg.db)
library(msigdbr)
library(patchwork)
library(pathview)
library(PCAtools)
library(pheatmap)
library(qs)
library(Rhdf5lib)
library(rhdf5)
library(Seurat)
library(slingshot)
library(scran)
library(scater)
library(scDblFinder)
library(scRNAseq)
library(SingleCellExperiment)
library(tictoc)
library(tidyverse)
library(TSCAN)
library(uwot)
library(zellkonverter)

path_to_data <- "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/"
```

```{r}
integrated_data_subset <- readRDS(paste0(path_to_data, "integrated.data.subset.rds"))

rna_counts <- integrated_data_subset@assays$RNA@counts %>% as.data.frame()

# sce <- SingleCellExperiment(assays = list(counts = rna_counts, logcounts = log_counts))
sce <- SingleCellExperiment(assays = list(counts = rna_counts))

sce <- sce %>% logNormCounts()

sce$cluster <- integrated_data_subset$seurat_clusters %>% as.character()

sce$cell_type <- integrated_data_subset$seurat_clusters %>% as.character()

sce$cell_type[sce$cell_type == "0"] <- "CD8"
sce$cell_type[sce$cell_type == "1"] <- "NKT"
sce$cell_type[sce$cell_type == "2"] <- "T_cells"
sce$cell_type[sce$cell_type == "3"] <- "Tregs"
sce$cell_type[sce$cell_type == "4"] <- "CD4"

colLabels(sce) <- sce$cell_type

sce$sample <- integrated_data_subset$sample

sce@int_colData@listData$reducedDims$UMAP <- integrated_data_subset@reductions$umap@cell.embeddings %>% as.data.frame()

sce %>% plotUMAP(colour_by = "label")

markers <- sce %>% findMarkers(test = "binom", direction = "up", lfc = 1, pval.type = "all")
markers <- sce %>% findMarkers(test = "binom", direction = "up", lfc = 1, pval.type = "any")

markers_list <- purrr::map(seq_along(markers), function(i) markers[[i]])

markers_list <- markers_list %>% purrr::set_names("CD8", "NKT", "T_cells", "Tregs", "CD4")

# DE_list <- purrr::map(seq_along(markers), function(i) markers_list[[i]] %>% as.data.frame() %>% dplyr::filter(FDR <= 0.05 & summary.logFC >= 1.0)) %>% purrr::set_names(names(markers_list))

DE_list <- purrr::map(seq_along(markers), function(i) {
  markers_list[[i]] %>%
    as.data.frame() %>%
    dplyr::filter(FDR <= 0.05 & summary.logFC >= 1.0 & Top <= 10)
}) %>% purrr::set_names(names(markers_list))

for (i in 1:length(DE_list)) {
  for (j in 1:length(rownames(DE_list[[i]]))) {
    print(i)
    print(j)
    violin_plot <- sce %>%
      plotExpression(x = "label", features = rownames(DE_list[[i]])[j], colour_by = "label") +
      theme_bw() +
      theme(axis.text.x = element_blank()) +
      ggeasy::easy_all_text_size(size = 20) +
      guides(color = guide_legend("Cell type")) +
      labs(x = "", y = "Normalized gene expression") +
      ggeasy::easy_legend_at(to = "bottom")

    violin_plot %>% ggsave(filename = paste0(path_to_data, "violin_plots/", names(DE_list)[i], "_", rownames(DE_list[[i]])[j], "_ violin_plot.png"), width = 10, height = 5)
  }
}


```



# custom enrichment analysis
```{r}
ora_analysis <- function(genes_list, database = KEGG_2019_Human, pvalue_cutoff = 1, qvalue_cutoff = 1, p.adjust_cutoff = 0.1, count_at_least = 3){
  library(clusterProfiler)
  library(tidyverse)
  gmt_df <- database %>% clusterProfiler::read.gmt()
  terms_list <- gmt_df %>% split(f = .$term) %>% purrr::map(pull, gene)
  genes_set <- gmt_df[, "gene"] %>% unique()
  enrichment_result <- genes_list %>% 
    purrr::map(clusterProfiler::enricher, pvalueCutoff = pvalue_cutoff, qvalueCutoff = qvalue_cutoff, universe = genes_set, TERM2GENE = gmt_df)
  clProf.df <- enrichment_result %>% purrr::map(as.data.frame) %>% bind_rows(.id = "Cluster")
  clProf.df <- clProf.df %>% dplyr::filter(p.adjust < p.adjust_cutoff & Count >= count_at_least)
  geneClusters <- genes_list
  compareCluster_output <- new("compareClusterResult", 
                               compareClusterResult = clProf.df, 
                               geneClusters = list(geneClusters), 
                               fun = "enricher",
                               .call = match.call(expand.dots = TRUE))
  return(compareCluster_output)
}

shape_data <- function(ora_results_output) {
  ora_results_output@compareClusterResult <- ora_results_output@compareClusterResult %>% 
    mutate_at(.vars = c("pvalue", "p.adjust", "qvalue"), .funs = formatC, digits = 2, format = "e") 
  return(ora_results_output)
}
```

# GSEA
```{r}
gsea_analysis <- function(database = KEGG_2019_Mouse, sorted_genes, p_value_cutoff = 0.05, p_adjust_cutoff = 0.05, min_size = 10, max_size = 500, p_adjust_method = "BH") {
  
  # Load required packages
  library(clusterProfiler)
  library(tidyverse)

  # Read the GMT file into a data frame
  gmt_df <- database %>% clusterProfiler::read.gmt()

  # Extract the gene sets (terms) from the GMT data frame
  gmt_list <- gmt_df %>%
    split(f = .$term) %>%
    purrr::map(pull, gene)

   # Perform Gene Set Enrichment Analysis (GSEA) using the sorted_genes and the database
  gsea_result <- clusterProfiler::GSEA(
    geneList = sorted_genes, 
    TERM2GENE = gmt_df, 
    verbose = FALSE, 
    minGSSize = min_size, 
    maxGSSize = max_size, 
    pvalueCutoff = p_value_cutoff, 
    pAdjustMethod = p_adjust_method
    ) %>% suppressWarnings()

  # Generate enrichment plots for each gene set and store them in a list
  gseaplot_results <- purrr::map(gsea_result@result$Description, function(i) {
    enrichplot::gseaplot2(
      x = gsea_result,
      geneSetID = i,
      title = gsea_result@result$Description[i]
    )
  }) %>% purrr::set_names(gsea_result@result$Description)

  # Store the GSEA results and the enrichment plot results in a list
  results <- list(
    gsea_results = gsea_result,
    gseaplot_results = gseaplot_results
  )

  # Return the results
  return(results)
}

```

# Databases
```{r}
# path_to_pathways <- "~/Dropbox/GMT/"
path_to_pathways <- "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/GMT/"
# path_to_pathways <- "C:\\Users\\Carlos\\Dropbox\\GMT\\"
WikiPathways_2019_Human <- path_to_pathways %>% paste0("WikiPathways_2019_Human.txt") # WikiPathways_2019_Human gene sets
WikiPathways_2019_Mouse <- path_to_pathways %>% paste0("WikiPathways_2019_Mouse.txt") # WikiPathways_2019_Mouse gene sets
KEGG_2019_Human <- path_to_pathways %>% paste0("KEGG_2019_Human.txt") # KEGG_2019_Human gene sets
KEGG_2019_Mouse <- path_to_pathways %>% paste0("KEGG_2019_Mouse.txt") # KEGG_2019_Mouse gene sets
LINCS_L1000_Chem_Pert_down <- path_to_pathways %>% paste0("LINCS_L1000_Chem_Pert_down.txt") # LINCS_L1000_Chem_Pert_down gene sets
LINCS_L1000_Chem_Pert_up <- path_to_pathways %>% paste0("LINCS_L1000_Chem_Pert_up.txt") # LINCS_L1000_Chem_Pert_up gene sets
Reactome_2016 <- path_to_pathways %>% paste0("Reactome_2016.txt") # Reactome_2016 gene sets
GO_Molecular_Function_2018 <- path_to_pathways %>% paste0("GO_Molecular_Function_2018.txt") # GO_Molecular_Function_2018 gene sets
GO_Cellular_Component_2018 <- path_to_pathways %>% paste0("GO_Cellular_Component_2018.txt") # GO_Cellular_Component_2018 gene sets
GO_Biological_Process_2018 <- path_to_pathways %>% paste0("GO_Biological_Process_2018.txt") # GO_Biological_Process_2018 gene sets
GO_Molecular_Function_2021 <- path_to_pathways %>% paste0("GO_Molecular_Function_2021.txt") # GO_Molecular_Function_2021 gene sets
GO_Cellular_Component_2021 <- path_to_pathways %>% paste0("GO_Cellular_Component_2021.txt") # GO_Cellular_Component_2021 gene sets
GO_Biological_Process_2021 <- path_to_pathways %>% paste0("GO_Biological_Process_2021.txt") # GO_Biological_Process_2021 gene sets
GO_Molecular_Function_2023 <- path_to_pathways %>% paste0("GO_Molecular_Function_2023.txt") # GO_Molecular_Function_2023 gene sets
GO_Cellular_Component_2023 <- path_to_pathways %>% paste0("GO_Cellular_Component_2023.txt") # GO_Cellular_Component_2023 gene sets
GO_Biological_Process_2023 <- path_to_pathways %>% paste0("GO_Biological_Process_2023.txt") # GO_Biological_Process_2023 gene sets
BioPlanet_2019 <- path_to_pathways %>% paste0("BioPlanet_2019.txt") # BioPlanet_2019 gene sets
# BioPlanet_pathways<- "~/Dropbox/GMT/BioPlanet_pathways.csv" # BioPlanet_pathways gene sets
C6_Oncogenic_Signatures <- path_to_pathways %>% paste0("c6.all.v2023.1.Hs.symbols.gmt")

C8_cell_type_signatures <- path_to_pathways %>% paste0("c8.all.v2023.1.Hs.symbols.gmt")
```

# GSEA analysis
```{r}
dge_cluster_7_T1_T2 <- read_delim(paste0(path_to_data, "02_Tumor_data/condition_specific_dge/dge_7_T1_T2.tsv"), 
     delim = "\t", escape_double = FALSE, 
     trim_ws = TRUE)

dge_cluster_7_T2_T4 <- read_delim(paste0(path_to_data, "/02_Tumor_data/condition_specific_dge/dge_7_T2_T4.tsv"), 
     delim = "\t", escape_double = FALSE, 
     trim_ws = TRUE)

dge_cluster_7_T3_T4 <- read_delim("~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/02_Tumor_data/condition_specific_dge/dge_7_T3_T4.tsv", 
     delim = "\t", escape_double = FALSE, 
     trim_ws = TRUE)

dge_cluster_9_T1_T2 <- read_delim("~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/02_Tumor_data/condition_specific_dge/dge_9_T1_T2.tsv", 
     delim = "\t", escape_double = FALSE, 
     trim_ws = TRUE)

dge_cluster_9_T2_T4 <- read_delim("~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/02_Tumor_data/condition_specific_dge/dge_9_T2_T4.tsv", 
     delim = "\t", escape_double = FALSE, 
     trim_ws = TRUE)

dge_cluster_9_T3_T4 <- read_delim("~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/02_Tumor_data/condition_specific_dge/dge_9_T3_T4.tsv", 
     delim = "\t", escape_double = FALSE, 
     trim_ws = TRUE)

dge_cluster_11_T1_T2 <- read_delim("~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/02_Tumor_data/condition_specific_dge/dge_11_T1_T2.tsv", 
     delim = "\t", escape_double = FALSE, 
     trim_ws = TRUE)

dge_cluster_11_T2_T4 <- read_delim("~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/02_Tumor_data/condition_specific_dge/dge_11_T2_T4.tsv", 
     delim = "\t", escape_double = FALSE, 
     trim_ws = TRUE)

dge_cluster_11_T3_T4 <- read_delim("~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/02_Tumor_data/condition_specific_dge/dge_11_T3_T4.tsv", 
     delim = "\t", escape_double = FALSE, 
     trim_ws = TRUE)

DGE_results_list <- list(
  cluster_7_T1_T2 = dge_cluster_7_T1_T2,
  cluster_7_T2_T4 = dge_cluster_7_T2_T4,
  cluster_7_T3_T4 = dge_cluster_7_T3_T4,
  cluster_9_T1_T2 = dge_cluster_9_T1_T2,
  cluster_9_T2_T4 = dge_cluster_9_T2_T4,
  cluster_9_T3_T4 = dge_cluster_9_T3_T4,
  cluster_11_T1_T2 = dge_cluster_11_T1_T2,
  cluster_11_T2_T4 = dge_cluster_11_T2_T4,
  cluster_11_T3_T4 = dge_cluster_11_T3_T4
     )

gene_list <- DGE_results_list %>% 
  purrr::map(function(x) x %>% 
               arrange(desc(as.numeric(avg_log2FC))) %>% 
               distinct(gene, .keep_all = TRUE) %>% 
               mutate(rank = rank(avg_log2FC, ties.method = "random"), 
                      gene = gene %>% str_to_upper) %>% 
               arrange(desc(rank)) %>%
               dplyr::select(gene, avg_log2FC)
  )

ranked_gene_list <- purrr::map(seq_along(gene_list), function(i) gene_list[[i]] %>% 
  pull(as.numeric(avg_log2FC)) %>%
  purrr::set_names(gene_list[[i]]$gene)) %>%
  purrr::set_names(names(DGE_results_list))

gsea_KEGG_results_list <- purrr::map(seq_along(ranked_gene_list), function(i) gsea_analysis(database = KEGG_2019_Mouse, sorted_genes = ranked_gene_list[[i]], p_value_cutoff = 0.05, p_adjust_cutoff = 0.05, min_size = 10, max_size = 500, p_adjust_method = "BH")) %>% purrr::set_names(names(DGE_results_list))

# gsea_KEGG_results_list %>% qs::qsave(paste0(path_to_data, "gsea_KEGG_results_list.qs"), nthreads = 50)
gsea_KEGG_results_list <- qs::qread(paste0(path_to_data, "gsea_KEGG_results_list.qs"), nthreads = 50)

for (i in 1:length(gsea_KEGG_results_list)) {
  for (j in 1:length(gsea_KEGG_results_list[[i]]$gsea_results)) {
    print(i)
    print(j)
    gsea_KEGG_results_list[[i]]$gsea_results@result %>%  
    writexl::write_xlsx(paste0(path_to_data, "/GSEA_results/", names(gsea_KEGG_results_list)[[i]], "_result_table.xlsx"))
  }
}

for (i in 1:length(gsea_KEGG_results_list)) {
  for (j in 1:length(gsea_KEGG_results_list[[i]]$gseaplot_results)) {
    print(i)
    print(j)
    gsea_KEGG_results_list[[i]]$gseaplot_results[[j]] %>% ggdraw() %>% 
    ggsave(filename = paste0(path_to_data, "/GSEA_results/", names(gsea_KEGG_results_list)[[i]], "_", names(gsea_KEGG_results_list[[i]]$gseaplot_results)[[j]], ".png"), device = "png", width = 7, height = 5, dpi = 900)
  }
}

gsea_KEGG_results_list$cluster_7_T1_T2$gseaplot_results$`Antigen processing and presentation`

for (i in seq_along(gsea_KEGG_results_list)) {
  for (j in seq_along(gsea_KEGG_results_list[[i]]$gsea_results@result$Description)) {
    print(i)
    print(j)
    gsea_plot <- enrichplot::gseaplot2(
      x = gsea_KEGG_results_list[[i]]$gsea_results,
      geneSetID = j,
      title = gsea_KEGG_results_list[[i]]$gsea_results@result$Description[j]
    )
    aplot::plot_list(gglist = gsea_plot, ncol = 1, heights = c(1.5, 0.5, 1)) %>%
      aplot::ggsave(filename = paste0(
        path_to_data, "GSEA_results/",
        names(gsea_KEGG_results_list)[[i]], "_",
        gsea_KEGG_results_list[[i]]$gsea_results@result$Description[j], ".png"
        ), 
        device = "png", width = 7, height = 5, dpi = 900)
  }
}

```

# GSEA analysis
```{r}
dge_cluster_0_T3_T4 <- read_delim(paste0(path_to_data, "DE_results/dge_0_T3_T4.tsv"), 
     delim = "\t", escape_double = FALSE, 
     col_names = FALSE, trim_ws = TRUE, skip = 1) %>% 
  purrr::set_names("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj")

dge_cluster_1_T3_T4 <- read_delim(paste0(path_to_data, "DE_results/dge_1_T3_T4.tsv"), 
     delim = "\t", escape_double = FALSE, 
     col_names = FALSE, trim_ws = TRUE, skip = 1) %>% 
  purrr::set_names("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj")

# dge_cluster_2_T3_T4 <- read_delim(paste0(path_to_data, "DE_results/dge_2_T3_T4.tsv"), 
     # delim = "\t", escape_double = FALSE, 
     # col_names = FALSE, trim_ws = TRUE, skip = 1) %>% 
     # purrr::set_names("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj")

dge_cluster_3_T3_T4 <- read_delim(paste0(path_to_data, "DE_results/dge_3_T3_T4.tsv"), 
     delim = "\t", escape_double = FALSE, 
     col_names = FALSE, trim_ws = TRUE, skip = 1) %>% 
  purrr::set_names("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj")

dge_cluster_4_T3_T4 <- read_delim(paste0(path_to_data, "DE_results/dge_4_T3_T4.tsv"), 
     delim = "\t", escape_double = FALSE, 
     col_names = FALSE, trim_ws = TRUE, skip = 1) %>% 
  purrr::set_names("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj")

DGE_results_list <- list(
  cluster_0_T3_T4 = dge_cluster_0_T3_T4,
  cluster_1_T3_T4 = dge_cluster_1_T3_T4,
  # cluster_2_T3_T4 = dge_cluster_2_T3_T4,
  cluster_3_T3_T4 = dge_cluster_3_T3_T4,
  cluster_4_T3_T4 = dge_cluster_4_T3_T4
     )

gene_list <- DGE_results_list %>% 
  purrr::map(function(x) x %>% 
               arrange(desc(as.numeric(avg_log2FC))) %>% 
               distinct(gene, .keep_all = TRUE) %>% 
               mutate(rank = rank(avg_log2FC, ties.method = "random"), 
                      gene = gene %>% str_to_upper) %>% 
               arrange(desc(rank)) %>%
               dplyr::select(gene, avg_log2FC)
  )

ranked_gene_list <- purrr::map(seq_along(gene_list), function(i) gene_list[[i]] %>% 
  pull(as.numeric(avg_log2FC)) %>%
  purrr::set_names(gene_list[[i]]$gene)) %>%
  purrr::set_names(names(DGE_results_list))

gsea_KEGG_results_list <- purrr::map(seq_along(ranked_gene_list), function(i) gsea_analysis(database = KEGG_2019_Mouse, sorted_genes = ranked_gene_list[[i]], p_value_cutoff = 0.05, p_adjust_cutoff = 0.05, min_size = 10, max_size = 500, p_adjust_method = "BH")) %>% purrr::set_names(names(DGE_results_list))

# gsea_KEGG_results_list %>% qs::qsave(paste0(path_to_data, "gsea_KEGG_results_list.qs"), nthreads = 50)
gsea_KEGG_results_list <- qs::qread(paste0(path_to_data, "gsea_KEGG_results_list.qs"), nthreads = 50)

# for (i in 1:length(gsea_KEGG_results_list)) {
for (i in 1:3) {  
  for (j in 1:length(gsea_KEGG_results_list[[i]]$gsea_results)) {
    print(i)
    print(j)
    gsea_KEGG_results_list[[i]]$gsea_results@result %>%  
    writexl::write_xlsx(paste0(path_to_data, "/GSEA_results/", names(gsea_KEGG_results_list)[[i]], "_result_table.xlsx"))
  }
}

# for (i in 1:length(gsea_KEGG_results_list)) {
for (i in 1:3) {    
  for (j in 1:length(gsea_KEGG_results_list[[i]]$gseaplot_results)) {
    print(i)
    print(j)
    gsea_KEGG_results_list[[i]]$gseaplot_results[[j]] %>% ggdraw() %>% 
    ggsave(filename = paste0(path_to_data, "/GSEA_results/", names(gsea_KEGG_results_list)[[i]], "_", names(gsea_KEGG_results_list[[i]]$gseaplot_results)[[j]], ".png"), device = "png", width = 7, height = 5, dpi = 900)
  }
}

gsea_KEGG_results_list$cluster_7_T1_T2$gseaplot_results$`Antigen processing and presentation`

for (i in seq_along(gsea_KEGG_results_list)) {
  for (j in seq_along(gsea_KEGG_results_list[[i]]$gsea_results@result$Description)) {
    print(i)
    print(j)
    gsea_plot <- enrichplot::gseaplot2(
      x = gsea_KEGG_results_list[[i]]$gsea_results,
      geneSetID = j,
      title = gsea_KEGG_results_list[[i]]$gsea_results@result$Description[j]
    )
    aplot::plot_list(gglist = gsea_plot, ncol = 1, heights = c(1.5, 0.5, 1)) %>%
      aplot::ggsave(filename = paste0(
        path_to_data, "GSEA_results/",
        names(gsea_KEGG_results_list)[[i]], "_",
        gsea_KEGG_results_list[[i]]$gsea_results@result$Description[j], ".png"
        ), 
        device = "png", width = 7, height = 5, dpi = 900)
  }
}

```

# Annnotate cluster 2 - Is it gdTCR?
```{r}
# T cell subset
# dge_cluster_2 <- read_delim(
#   file = paste0(path_to_data,
#                 "02_Tumor_data/condition_specific_dge/dge_2_T3_T4.tsv"),
#   delim = "\t", escape_double = FALSE,
#   col_names = FALSE, 
#   skip = 1,
#   trim_ws = TRUE) %>%
# purrr::set_names("p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "cluster", "gene") %>% 
#   dplyr::select("gene", "avg_log2FC", "p_val", "p_val_adj", everything())

dge_cluster_2 <- read_delim(
  file = paste0(path_to_data,
                "DE_results_2/dge_2_S3_S4.tsv"),
  delim = "\t", 
  escape_double = FALSE,
  col_names = FALSE, 
  skip = 1,
  trim_ws = TRUE) %>%
purrr::set_names("p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "cluster", "gene") %>% 
  dplyr::select("gene", "avg_log2FC", "p_val", "p_val_adj", everything())

# supercluster
# dge_cluster_2 <- read_delim(paste0(path_to_data, "DE_results/dge_2_T3_T4.tsv"),
# delim = "\t", escape_double = FALSE,
# col_names = FALSE, trim_ws = TRUE, skip = 1) %>%
# purrr::set_names("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj")

gene_df <- dge_cluster_2 %>% 
  arrange(desc(as.numeric(avg_log2FC))) %>% 
  distinct(gene, .keep_all = TRUE) %>% 
  mutate(rank = rank(avg_log2FC, ties.method = "random"), 
         gene = gene %>% str_to_upper) %>% 
  arrange(desc(rank)) %>%  
  dplyr::select(gene, avg_log2FC)

ranked_gene_v <- gene_df %>% 
  pull(as.numeric(avg_log2FC)) %>%
  purrr::set_names(gene_df$gene) 

gsea_KEGG_results <- gsea_analysis(database = C8_cell_type_signatures, sorted_genes = ranked_gene_v, p_value_cutoff = 0.05, p_adjust_cutoff = 0.05, min_size = 10, max_size = 500, p_adjust_method = "BH") 

gsea_KEGG_results$gsea_results@result %>% vroom::vroom_write(file = paste0(path_to_data, "GSEA_C8_cell_type_signature_result_", "cluster_2", ".tsv"), delim = "\t")

dge_cluster_2 %>% dplyr::filter(abs(avg_log2FC) > 1)
```

# For supercluster, cluster 0-8, 10 run GSEA using KEGG
```{r}
dge_cluster_files <- list.files(paste0(path_to_data, "02_Tumor_data/condition_specific_dge/"), full.names = TRUE)
file_names <- list.files(paste0(path_to_data, "02_Tumor_data/condition_specific_dge/"), full.names = FALSE) %>% str_remove(pattern = ".tsv")

dge_cluster_list <- dge_cluster_files %>% purrr::map(function(x) x %>% read_delim(delim = "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE, skip = 1) %>% 
  purrr::set_names("p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "cluster", "gene") %>%
  dplyr::select("gene", "avg_log2FC", "p_val", "p_val_adj", everything())
)

dge_cluster_list <- dge_cluster_list %>% purrr::set_names(file_names)

gene_list <- dge_cluster_list %>% 
  purrr::map(function(x) x %>% 
               arrange(desc(as.numeric(avg_log2FC))) %>% 
               distinct(gene, .keep_all = TRUE) %>% 
               mutate(rank = rank(avg_log2FC, ties.method = "random"), 
                      gene = gene %>% str_to_upper) %>% 
               arrange(desc(rank)) %>%
               dplyr::select(gene, avg_log2FC)
  )

ranked_gene_list <- purrr::map(seq_along(gene_list), function(i) gene_list[[i]] %>% 
  pull(as.numeric(avg_log2FC)) %>%
  purrr::set_names(gene_list[[i]]$gene)) %>%
  purrr::set_names(names(dge_cluster_list))

gsea_KEGG_results_list <- purrr::map(seq_along(ranked_gene_list), function(i) gsea_analysis(database = KEGG_2019_Mouse, sorted_genes = ranked_gene_list[[i]], p_value_cutoff = 0.05, p_adjust_cutoff = 0.05, min_size = 10, max_size = 500, p_adjust_method = "BH")) %>% purrr::set_names(names(dge_cluster_list))

# purrr::map(seq_along(gsea_KEGG_results_list), function(i) {
#   if (nrow(gsea_KEGG_results_list[[i]]$gsea_results@result > 0)) {
#     gsea_KEGG_results_list[[i]]$gsea_results@result %>% vroom::vroom_write(file = paste0(path_to_data, "GSEA_results_merged_cluster/GSEA_KEGG_results_", names(gsea_KEGG_results_list)[i], ".tsv"), delim = "\t")
#   }
# })


```





# merged dataset (Shweta's "supercluster")
```{r}
integrated_data <- readRDS(paste0(path_to_data, "02_Tumor_data/integrated_data.rds"))

rna_counts <- integrated_data@assays$RNA@counts %>% as.data.frame()

# sce <- SingleCellExperiment(assays = list(counts = rna_counts, logcounts = log_counts))
sce <- SingleCellExperiment(assays = list(counts = rna_counts))

sce <- sce %>% logNormCounts()

sce$cluster <- integrated_data$seurat_clusters %>% as.character()

sce$cell_type <- integrated_data$seurat_clusters %>% as.character()

sce$cell_type[sce$cell_type == "0"] <- "Macrophages_0"
sce$cell_type[sce$cell_type == "1"] <- "Macrophages_1"
sce$cell_type[sce$cell_type == "2"] <- "CCR2+_ILB+_infiltrating_macrophages"
sce$cell_type[sce$cell_type == "3"] <- "M2_macrophages"
sce$cell_type[sce$cell_type == "4"] <- "Mononuclear_phagocytes"
sce$cell_type[sce$cell_type == "5"] <- "Inflammatory_Macrophages"

sce$cell_type[sce$cell_type == "7"] <- "T_NK_NKT_cells"
sce$cell_type[sce$cell_type == "9"] <- "T_lymphocytes"
sce$cell_type[sce$cell_type == "11"] <- "T_cells"

sce$cell_type[sce$cell_type == "8"] <- "Conventional_dendritic_cells"

sce$cell_type[sce$cell_type == "6"] <- "Mixed_proliferating_cells"

sce$cell_type[sce$cell_type == "10"] <- "Epithelial_like_cancer_cells"

sce$cell_type[sce$cell_type == "12"] <- "Neutrophils"

colLabels(sce) <- sce$cell_type

sce$sample <- integrated_data$sample

sce$cell_type_sample <- paste0(sce$cell_type, "_", sce$sample)

sce@int_colData@listData$reducedDims$UMAP <- integrated_data@reductions$umap@cell.embeddings %>% as.data.frame()

sce %>% plotUMAP(colour_by = "label")

sce %>% plotUMAP(colour_by = "cell_type_sample")


### DEGs
dge_cluster_files <- list.files(paste0(path_to_data, "02_Tumor_data/condition_specific_dge_20230316/"), full.names = TRUE, pattern = "T1_T2|T1_T3|T1_T4")
file_names <- list.files(paste0(path_to_data, "02_Tumor_data/condition_specific_dge_20230316/"), full.names = FALSE, pattern = "T1_T2|T1_T3|T1_T4") %>% str_remove(pattern = ".tsv")

dge_cluster_list <- dge_cluster_files %>% purrr::map(function(x) x %>% read_delim(delim = "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE, skip = 1) %>% 
  purrr::set_names("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj") %>%
  dplyr::select("gene", "avg_log2FC", "p_val", "p_val_adj", everything())
)

dge_cluster_list <- dge_cluster_list %>% purrr::set_names(file_names)
 
# dge_cluster_list <- dge_cluster_list %>% purrr::map(dplyr::filter, abs(avg_log2FC) > 1 & p_val_adj < 0.05)

gene_umap_plots <- function(gene, sce) {
  umap_df <- sce@int_colData@listData$reducedDims$UMAP %>%
    as.data.frame() %>%
    mutate(gene_expression = sce@assays@data$logcounts[gene, ] %>% as.numeric()) %>%
    arrange(gene_expression)

  gene_plots <- umap_plot <- umap_df %>% ggplot(aes(x = UMAP_1, y = UMAP_2, color = gene_expression)) +
    geom_point() +
    scale_color_continuous(low = "grey", high = "red", guide = "colourbar", aesthetics = "colour") +
    theme_bw() +
    labs(x = "UMAP 1", y = "UMAP 2", title = gene) +
    easy_plot_title_size(size = 20)

  return(gene_plots)
}

gene_umap_plots(gene = "Irf1", sce = sce)

sub_sce_1 <- sce[, sce$cluster %in% c("0", "1", "2", "3", "4", "5")]
sub_sce_2 <- sce[, sce$cluster %in% c("7", "9", "11")]
sub_sce_3 <- sce[, sce$cluster %in% c("8")]
sub_sce_4 <- sce[, sce$cluster %in% c("12")]

# Violin plots
for (i in 1:length(DE_list)) {
  for (j in 1:length(rownames(DE_list[[i]]))) {
    print(i)
    print(j)
    violin_plot <- sub_sce_1 %>%
      plotExpression(x = "cell_type_sample", features = dge_cluster_list[[i]]$gene[j], colour_by = "cell_type_sample") +
      theme_bw() +
      theme(axis.text.x = element_blank()) +
      ggeasy::easy_all_text_size(size = 20) +
      guides(color = guide_legend("Cell type")) +
      labs(x = "", y = "Normalized gene expression") +
      ggeasy::easy_legend_at(to = "bottom")

    violin_plot %>% ggsave(filename = paste0(path_to_data, "violin_plots/", names(DE_list)[i], "_", rownames(DE_list[[i]])[j], "_ violin_plot.png"), width = 10, height = 5)
  }
}

# dot plot
# metadata_df <- sce@colData %>% 
#   as.data.frame() %>% 
#   rownames_to_column(var = "cell")
# 
# sub_sce_3_df <- sub_sce_3@assays@data$counts %>% 
#   as.data.frame() %>% 
#   rownames_to_column(var = "gene") %>% 
#   pivot_longer(cols = 2:ncol(.), names_to = "cell", values_to = "expression") %>% 
#   left_join(metadata_df, by = "cell") 
# 
# sub_sce_3_df <- sub_sce_3_df %>% 
#   dplyr::filter(gene %in% degs_cluster_8_T1_T3)
# 
# sub_sce_3_summarized_df <- sub_sce_3_df %>% 
#   group_by(gene, sample) %>% 
#   summarize(expression_sum = sum(expression), max_expr = max(expression_sum), detection = expression_sum/max_expr) 
# 
# max_expr <- sub_sce_3_summarized_df %>% 
#   pull(expression_sum) %>%
#   max()
# 
# sub_sce_3_summarized_df <- sub_sce_3_summarized_df %>% 
#   mutate(percentage = expression_sum/max_expr)
# 
# degs_cluster_8_T1_T3 <- dge_cluster_list[[27]]$gene



# sub_sce_3_df %>% ggplot(aes_string(x = x, y = "sample", size=size, color=colorBy)) +
#         geom_point() +
#         scale_color_continuous(low="red", high="blue", name = color,
#             guide=guide_colorbar(reverse=TRUE)) +
#         scale_y_discrete(labels = label_func) +
#         ylab(NULL) + ggtitle(title) + theme_dose(font.size) +
#         scale_size(range=c(3, 8))


# Barplot for GSEA results
gene_list <- dge_cluster_list %>% 
  purrr::map(function(x) x %>% 
               arrange(desc(as.numeric(avg_log2FC))) %>% 
               distinct(gene, .keep_all = TRUE) %>% 
               mutate(rank = rank(avg_log2FC, ties.method = "random"), 
                      gene = gene %>% str_to_upper) %>% 
               arrange(desc(rank)) %>%
               dplyr::select(gene, avg_log2FC)
  )

ranked_gene_list <- purrr::map(seq_along(gene_list), function(i) gene_list[[i]] %>% 
  pull(as.numeric(avg_log2FC)) %>%
  purrr::set_names(gene_list[[i]]$gene)) %>%
  purrr::set_names(names(dge_cluster_list))

gsea_KEGG_results_list <- purrr::map(seq_along(ranked_gene_list), function(i) gsea_analysis(database = KEGG_2019_Mouse, sorted_genes = ranked_gene_list[[i]], p_value_cutoff = 0.05, p_adjust_cutoff = 0.05, min_size = 10, max_size = 500, p_adjust_method = "BH")) %>% purrr::set_names(names(dge_cluster_list))

# gsea_KEGG_results_list %>% qs::qsave(file = paste0(path_to_data, "gsea_KEGG_results_list_supercluster.qs"), nthreads = 50)
gsea_KEGG_results_list <- qs::qread(file = paste0(path_to_data, "gsea_KEGG_results_list_supercluster.qs"), nthreads = 50)
# purrr::map(seq_along(gsea_KEGG_results_list), function(i) {
#   if (nrow(gsea_KEGG_results_list[[i]]$gsea_results@result > 0)) {
#     gsea_KEGG_results_list[[i]]$gsea_results@result %>% vroom::vroom_write(file = paste0(path_to_data, "GSEA_KEGG_results_", names(gsea_KEGG_results_list)[i], ".tsv"), delim = "\t")
#   }
# })

# gsea_KEGG_results_list$dge_0_T1_T2$gsea_results@result %>% 
# gsea_KEGG_results_list$dge_5_T1_T3$gsea_results@result %>% 
gsea_KEGG_results_list$dge_5_T1_T4$gsea_results@result %>%   
  mutate(colog_FDR = -log10(p.adjust)) %>% 
  arrange(desc(colog_FDR)) %>% 
  ggplot(aes(x = colog_FDR, y = reorder(Description, colog_FDR), fill = colog_FDR)) +
  geom_bar(stat = "identity") +
  geom_vline(xintercept = -log10(0.05), linetype = 'dotted', col = 'black') +
  labs(y = "KEGG terms", x = "Statistical significancy")


pathways_of_interest <- c("Antigen processing and presentation", "NOD-like receptor signaling pathway", "NF-kappa B signaling pathway", "Phagosome", "Hematopoietic cell lineage", "Toll-like receptor signaling pathway", "Oxidative phosphorylation", "Hepatitis C", "JAK-STAT signaling pathway", "Pathways in cancer", "HIF-1 signaling pathway", "Oxidative phosphorylation", "Parkinson disease", "Cell adhesion molecules (CAMs)", "Phagosome", "NOD-like receptor signaling pathway", "Cytosolic DNA-sensing pathway", "NF-kappa B signaling pathway", "Antigen processing and presentation", "Cell adhesion molecules (CAMs)", "NOD-like receptor signaling pathway", "RIG-I-like receptor signaling pathway", "Inflammatory bowel disease (IBD)", "Hematopoietic cell lineage", "Pathways in cancer", "Phagosome", "Antigen processing and presentation", "Cell adhesion molecules (CAMs)", "Phagosome", "Hematopoietic cell lineage", "RIG-I-like receptor signaling pathway", "NOD-like receptor signaling pathway", "C-type lectin receptor signaling pathway", "Proteasome", "Cytosolic DNA-sensing pathway", "Cellular senescence", "Cell adhesion molecules (CAMs)", "Antigen processing and presentation", "Phagosome", "Rap1 signaling pathway", "Hematopoietic cell lineage", "Cell adhesion molecules (CAMs)", "Th17 cell differentiation", "Antigen processing and presentation", "Phagosome", "NF-kappa B signaling pathway", "Oxidative phosphorylation", "HIF-1 signaling pathway", "NOD-like receptor signaling pathway", "RIG-I-like receptor signaling pathway", "Rap1 signaling pathway", "Lysosome", "MAPK signaling pathway")

pathways_of_interest <- c("Chemokine signaling pathway")

pathways_of_interest <- pathways_of_interest %>% unique() %>% sort()

gsea_KEGG_result_table_sub_list <- purrr::map(seq_along(gsea_KEGG_results_list), function(i) gsea_KEGG_results_list[[i]]$gsea_results@result %>% dplyr::filter(ID %in% pathways_of_interest))

genes_of_interest <- purrr::map(seq_along(gsea_KEGG_result_table_sub_list), function(i) gsea_KEGG_result_table_sub_list[[i]]$core_enrichment %>% str_split(pattern = "/") %>% unlist() %>% unique() %>% sort()) %>% unlist() %>% unique() %>% sort() %>% str_to_title()

# Heatmap
T_cell_receptor_genes <- KEGG_2019_Mouse %>% 
  clusterProfiler::read.gmt() %>% 
  dplyr::filter(term == "T cell receptor signaling pathway") %>% 
  pull(gene) %>% 
  str_to_title() %>% 
  sort()

heatmap_df <- dge_cluster_list$dge_0_T1_T2 %>% dplyr::select(gene, avg_log2FC) %>% 
  full_join(dge_cluster_list$dge_0_T1_T3 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_0_T1_T4 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  purrr::set_names("Gene", "T1_vs_T2", "T1_vs_T3", "T1_vs_T4") %>% 
  dplyr::filter(Gene %in% T_cell_receptor_genes) %>% 
  pivot_longer(cols = T1_vs_T2:T1_vs_T4, names_to = "Comparison", values_to = "LFC")

max_expr <- max(abs(min(heatmap_df$LFC, na.rm = TRUE)), max(heatmap_df$LFC, na.rm = TRUE))

heatmap_df %>% 
  ggplot(aes(x = Comparison, y = Gene)) + 
  geom_tile(aes(fill = LFC)) +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", limits = c(-1*max_expr, max_expr))

Chemokine_signaling_pathway_genes <- KEGG_2019_Mouse %>% 
  clusterProfiler::read.gmt() %>% 
  dplyr::filter(term == "Chemokine signaling pathway") %>% 
  pull(gene) %>% 
  str_to_title() %>% 
  sort()

heatmap_df_2 <- dge_cluster_list$dge_0_T1_T2 %>% dplyr::select(gene, avg_log2FC) %>% 
  full_join(dge_cluster_list$dge_0_T1_T3 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_0_T1_T4 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_1_T1_T2 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_1_T1_T3 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_1_T1_T4 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_2_T1_T2 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_2_T1_T3 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_2_T1_T4 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_3_T1_T2 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_3_T1_T3 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_3_T1_T4 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_4_T1_T2 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_4_T1_T3 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_4_T1_T4 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_5_T1_T2 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_5_T1_T3 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_5_T1_T4 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  # full_join(dge_cluster_list$dge_8_T1_T2 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  # full_join(dge_cluster_list$dge_8_T1_T3 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  # full_join(dge_cluster_list$dge_8_T1_T4 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  # full_join(dge_cluster_list$dge_12_T1_T2 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  # full_join(dge_cluster_list$dge_12_T1_T3 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  # full_join(dge_cluster_list$dge_12_T1_T4 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  purrr::set_names("Gene", 
                   "cl0_T1_vs_T2", "cl0_T1_vs_T3", "cl0_T1_vs_T4", 
                   "cl1_T1_vs_T2", "cl1_T1_vs_T3", "cl1_T1_vs_T4",
                   "cl2_T1_vs_T2", "cl2_T1_vs_T3", "cl2_T1_vs_T4",
                   "cl3_T1_vs_T2", "cl3_T1_vs_T3", "cl3_T1_vs_T4",
                   "cl4_T1_vs_T2", "cl4_T1_vs_T3", "cl4_T1_vs_T4",
                   "cl5_T1_vs_T2", "cl5_T1_vs_T3", "cl5_T1_vs_T4"
                   # "cl8_T1_vs_T2", "cl8_T1_vs_T3", "cl8_T1_vs_T4",
                   # "cl12_T1_vs_T2", "cl12_T1_vs_T3", "cl12_T1_vs_T4"
                   ) %>% 
  dplyr::filter(Gene %in% genes_of_interest) %>%
  pivot_longer(cols = cl0_T1_vs_T2:cl5_T1_vs_T4, names_to = "Comparison", values_to = "LFC") %>% 
  mutate(LFC = LFC %>% replace(is.na(.), 0))

max_expr <- max(abs(min(heatmap_df_2$LFC, na.rm = TRUE)), max(heatmap_df_2$LFC, na.rm = TRUE))

heatmap_df_2 %>% 
  ggplot(aes(x = Comparison, y = Gene)) + 
  geom_tile(aes(fill = LFC)) +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", limits = c(-1*max_expr, max_expr))
```


```{r}
pathways_of_interest <- c("Antigen processing and presentation", "NOD-like receptor signaling pathway", "NF-kappa B signaling pathway", "Phagosome", "Hematopoietic cell lineage", "Toll-like receptor signaling pathway", "Oxidative phosphorylation", "Hepatitis C", "JAK-STAT signaling pathway", "Pathways in cancer", "HIF-1 signaling pathway", "Oxidative phosphorylation", "Parkinson disease", "Cell adhesion molecules (CAMs)", "Phagosome", "NOD-like receptor signaling pathway", "Cytosolic DNA-sensing pathway", "NF-kappa B signaling pathway", "Antigen processing and presentation", "Cell adhesion molecules (CAMs)", "NOD-like receptor signaling pathway", "RIG-I-like receptor signaling pathway", "Inflammatory bowel disease (IBD)", "Hematopoietic cell lineage", "Pathways in cancer", "Phagosome", "Antigen processing and presentation", "Cell adhesion molecules (CAMs)", "Phagosome", "Hematopoietic cell lineage", "RIG-I-like receptor signaling pathway", "NOD-like receptor signaling pathway", "C-type lectin receptor signaling pathway", "Proteasome", "Cytosolic DNA-sensing pathway", "Cellular senescence", "Cell adhesion molecules (CAMs)", "Antigen processing and presentation", "Phagosome", "Rap1 signaling pathway", "Hematopoietic cell lineage", "Cell adhesion molecules (CAMs)", "Th17 cell differentiation", "Antigen processing and presentation", "Phagosome", "NF-kappa B signaling pathway", "Oxidative phosphorylation", "HIF-1 signaling pathway", "NOD-like receptor signaling pathway", "RIG-I-like receptor signaling pathway", "Rap1 signaling pathway", "Lysosome", "MAPK signaling pathway")

# pathways_of_interest <- c("Chemokine signaling pathway")

pathways_of_interest <- pathways_of_interest %>% unique() %>% sort()

purrr::map(seq_along(pathways_of_interest), function(pathway) {
  gsea_KEGG_result_table_sub_list <- purrr::map(seq_along(gsea_KEGG_results_list), function(i) gsea_KEGG_results_list[[i]]$gsea_results@result %>% dplyr::filter(ID %in% pathways_of_interest[pathway]))
  
  genes_of_interest <- purrr::map(seq_along(gsea_KEGG_result_table_sub_list), function(i) gsea_KEGG_result_table_sub_list[[i]]$core_enrichment %>% str_split(pattern = "/") %>% unlist() %>% unique() %>% sort()) %>% unlist() %>% unique() %>% sort() %>% str_to_title()
  
  heatmap_df_2 <- dge_cluster_list$dge_0_T1_T2 %>% dplyr::select(gene, avg_log2FC) %>% 
  full_join(dge_cluster_list$dge_0_T1_T3 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_0_T1_T4 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_1_T1_T2 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_1_T1_T3 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_1_T1_T4 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_2_T1_T2 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_2_T1_T3 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_2_T1_T4 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_3_T1_T2 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_3_T1_T3 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_3_T1_T4 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_4_T1_T2 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_4_T1_T3 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_4_T1_T4 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_5_T1_T2 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_5_T1_T3 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(dge_cluster_list$dge_5_T1_T4 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  # full_join(dge_cluster_list$dge_8_T1_T2 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  # full_join(dge_cluster_list$dge_8_T1_T3 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  # full_join(dge_cluster_list$dge_8_T1_T4 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  # full_join(dge_cluster_list$dge_12_T1_T2 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  # full_join(dge_cluster_list$dge_12_T1_T3 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  # full_join(dge_cluster_list$dge_12_T1_T4 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  purrr::set_names("Gene", 
                   "cl0_T1_vs_T2", "cl0_T1_vs_T3", "cl0_T1_vs_T4", 
                   "cl1_T1_vs_T2", "cl1_T1_vs_T3", "cl1_T1_vs_T4",
                   "cl2_T1_vs_T2", "cl2_T1_vs_T3", "cl2_T1_vs_T4",
                   "cl3_T1_vs_T2", "cl3_T1_vs_T3", "cl3_T1_vs_T4",
                   "cl4_T1_vs_T2", "cl4_T1_vs_T3", "cl4_T1_vs_T4",
                   "cl5_T1_vs_T2", "cl5_T1_vs_T3", "cl5_T1_vs_T4"
                   # "cl8_T1_vs_T2", "cl8_T1_vs_T3", "cl8_T1_vs_T4",
                   # "cl12_T1_vs_T2", "cl12_T1_vs_T3", "cl12_T1_vs_T4"
                   ) %>% 
  dplyr::filter(Gene %in% genes_of_interest) %>%
  pivot_longer(cols = cl0_T1_vs_T2:cl5_T1_vs_T4, names_to = "Comparison", values_to = "LFC") %>% 
  mutate(LFC = LFC %>% replace(is.na(.), 0))

max_expr <- max(abs(min(heatmap_df_2$LFC, na.rm = TRUE)), max(heatmap_df_2$LFC, na.rm = TRUE))

heatmap_figure <- heatmap_df_2 %>% 
  ggplot(aes(x = Comparison, y = Gene)) + 
  geom_tile(aes(fill = LFC)) +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", limits = c(-1*max_expr, max_expr)) +
  labs(title = pathways_of_interest[pathway])

heatmap_figure %>% ggsave(filename = paste0(path_to_data, "Pathway_heatmaps/", pathways_of_interest[pathway], ".svg"), width = 15, height = 8, dpi = 900)

})


```



# Wordcloud from GSEA results
```{r}
gsea_KEGG_results_list$cluster_7_T3_T4$gsea_results@result$p.adjust
gsea_KEGG_results_list$cluster_7_T3_T4$gsea_results@result$Description

plot_wordcloud <- function(gsea_results_df){
  set.seed(123)
  gsea_results_df %>% 
    dplyr::filter(p.adjust <= 0.001) %>% 
    ggplot(aes(label = Description, 
               x = Description, 
               # color = Cluster, 
               size = -log10(p.adjust))) +
    geom_text_wordcloud(shape = "cardioid") +
    theme_minimal()
}

word_cloud_plot_KEGG_cluster_7_T3_T4 <- gsea_KEGG_results_list$cluster_7_T3_T4$gsea_results@result %>% 
  plot_wordcloud() +
  labs(x = "KEGG, cluster_7 T3_vs_T4")

word_cloud_plot_KEGG_cluster_9_T3_T4 <- gsea_KEGG_results_list$cluster_9_T3_T4$gsea_results@result %>% 
  plot_wordcloud() +
  labs(x = "KEGG, cluster_9 T3_vs_T4")

word_cloud_plot_KEGG_cluster_11_T3_T4 <- gsea_KEGG_results_list$cluster_11_T3_T4$gsea_results@result %>% 
  plot_wordcloud() +
  labs(x = "KEGG, cluster_11 3_vs_T4")
```

# bar plot
```{r}
cluster_color_df <- data.frame(
  cluster = 0:12,
  color = c("#FF0000", "#705C83", "#3E8E93", "#2CA02C", "#7e6e85", "#ba5e6c", "#FF7F00", "#FFD421", "#e1c62f", "#A0582C", "#db728c", "#d789b2", "#999999")
)


# gsea_KEGG_results_list$cluster_7_T3_T4$gsea_results@result %>% 
  # set.seed(123)
plot_flipped_barplot <- function(df, fill_color = "red"){
  df %>% 
    dplyr::filter(p.adjust <= 0.01) %>%
    ggplot(aes(x = reorder(Description, -log10(p.adjust)), 
               y = -log10(p.adjust))) +
  geom_bar(stat="identity", fill = fill_color) +
    coord_flip() +
    theme_minimal() +
    ggeasy::easy_x_axis_title_size(size = 20) +
    ggeasy::easy_y_axis_title_size(size = 20)
    
}    

barplot_1 <- gsea_KEGG_results_list$cluster_7_T3_T4$gsea_results@result %>% 
  plot_flipped_barplot(fill_color = cluster_color_df$color[8]) +
  labs(x = "GSEA enriched terms", title = "Cluster 7, T3 vs T4") 
  
barplot_1 %>% 
  ggsave(filename = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/barplot_GSEA_cluster_7_T3_T4.png", width = 10, height = 10, dpi = 900)


barplot_2 <- gsea_KEGG_results_list$cluster_9_T3_T4$gsea_results@result %>% 
  plot_flipped_barplot(fill_color = cluster_color_df$color[10]) +
  labs(x = "GSEA enriched terms", title = "Cluster 9, T3 vs T4")

barplot_2 %>% 
  ggsave(filename = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/barplot_GSEA_cluster_9_T3_T4.png", width = 10, height = 10, dpi = 900)
  
barplot_3 <- gsea_KEGG_results_list$cluster_11_T3_T4$gsea_results@result %>% 
  plot_flipped_barplot(fill_color = cluster_color_df$color[12]) +
  labs(x = "GSEA enriched terms", title = "Cluster 11, T3 vs T4")

barplot_3 %>% 
  ggsave(filename = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/barplot_GSEA_cluster_11_T3_T4.png", width = 10, height = 10, dpi = 900)

```

```{r}
# Use Hallmarks C6 dataset, for all clusters
gsea_C6_Hallmarks_results_list <- purrr::map(seq_along(ranked_gene_list), function(i) gsea_analysis(database = C6_Oncogenic_Signatures, sorted_genes = ranked_gene_list[[i]], p_value_cutoff = 0.05, p_adjust_cutoff = 0.05, min_size = 10, max_size = 500, p_adjust_method = "BH")) %>% purrr::set_names(names(DGE_results_list))
# Use KEGG on cl t-cell subsets 1, 2, 3, 4, 5

barplot_C6_cluster_7_T1_T2 <- gsea_C6_Hallmarks_results_list$cluster_7_T1_T2$gsea_results@result %>% 
  plot_flipped_barplot(fill_color = cluster_color_df$color[8]) +
  labs(x = "GSEA enriched terms", title = "Cluster 7, T1 vs T2") 

barplot_C6_cluster_9_T1_T2 <- gsea_C6_Hallmarks_results_list$cluster_9_T1_T2$gsea_results@result %>% 
  plot_flipped_barplot(fill_color = cluster_color_df$color[10]) +
  labs(x = "GSEA enriched terms", title = "Cluster 9, T1 vs T2")

barplot_C6_cluster_11_T1_T2 <- gsea_C6_Hallmarks_results_list$cluster_11_T1_T2$gsea_results@result %>% 
  plot_flipped_barplot(fill_color = cluster_color_df$color[12]) +
  labs(x = "GSEA enriched terms", title = "Cluster 11, T1 vs T2") 


barplot_C6_cluster_7_T2_T4 <- gsea_C6_Hallmarks_results_list$cluster_7_T2_T4$gsea_results@result %>% 
  plot_flipped_barplot(fill_color = cluster_color_df$color[8]) +
  labs(x = "GSEA enriched terms", title = "Cluster 7, T2 vs T4") 

barplot_C6_cluster_9_T2_T4 <- gsea_C6_Hallmarks_results_list$cluster_9_T2_T4$gsea_results@result %>% 
  plot_flipped_barplot(fill_color = cluster_color_df$color[10]) +
  labs(x = "GSEA enriched terms", title = "Cluster 9, T2 vs T4")

barplot_C6_cluster_11_T2_T4 <- gsea_C6_Hallmarks_results_list$cluster_11_T2_T4$gsea_results@result %>% 
  plot_flipped_barplot(fill_color = cluster_color_df$color[12]) +
  labs(x = "GSEA enriched terms", title = "Cluster 11, T2 vs T4") 

barplot_C6_cluster_7_T3_T4 <- gsea_C6_Hallmarks_results_list$cluster_7_T3_T4$gsea_results@result %>% 
  plot_flipped_barplot(fill_color = cluster_color_df$color[8]) +
  labs(x = "GSEA enriched terms", title = "Cluster 7, T3 vs T4") 

barplot_C6_cluster_9_T3_T4 <- gsea_C6_Hallmarks_results_list$cluster_9_T3_T4$gsea_results@result %>% 
  plot_flipped_barplot(fill_color = cluster_color_df$color[10]) +
  labs(x = "GSEA enriched terms", title = "Cluster 9, T3 vs T4")

barplot_C6_cluster_11_T3_T4 <- gsea_C6_Hallmarks_results_list$cluster_11_T3_T4$gsea_results@result %>% 
  plot_flipped_barplot(fill_color = cluster_color_df$color[12]) +
  labs(x = "GSEA enriched terms", title = "Cluster 11, T3 vs T4") 


barplot_C6_cluster_7_T1_T2 %>% 
  ggsave(filename = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/barplot_GSEA_C6_cluster_7_T1_T2.png", width = 10, height = 10, dpi = 900)

barplot_C6_cluster_9_T1_T2 %>% 
  ggsave(filename = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/barplot_GSEA_C6_cluster_9_T1_T2.png", width = 10, height = 10, dpi = 900)

barplot_C6_cluster_11_T1_T2 %>% 
  ggsave(filename = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/barplot_GSEA_C6_cluster_11_T1_T2.png", width = 10, height = 10, dpi = 900)

barplot_C6_cluster_7_T2_T4 %>% 
  ggsave(filename = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/barplot_GSEA_C6_cluster_7_T2_T4.png", width = 10, height = 10, dpi = 900)

barplot_C6_cluster_9_T2_T4 %>% 
  ggsave(filename = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/barplot_GSEA_C6_cluster_9_T2_T4.png", width = 10, height = 10, dpi = 900)

barplot_C6_cluster_11_T2_T4 %>% 
  ggsave(filename = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/barplot_GSEA_C6_cluster_11_T2_T4.png", width = 10, height = 10, dpi = 900)

barplot_C6_cluster_7_T3_T4 %>% 
  ggsave(filename = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/barplot_GSEA_C6_cluster_7_T3_T4.png", width = 10, height = 10, dpi = 900)

barplot_C6_cluster_9_T3_T4 %>% 
  ggsave(filename = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/barplot_GSEA_C6_cluster_9_T3_T4.png", width = 10, height = 10, dpi = 900)

barplot_C6_cluster_11_T3_T4 %>% 
  ggsave(filename = "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/scRNAseq_Shweta/barplot_GSEA_C6_cluster_11_T3_T4.png", width = 10, height = 10, dpi = 900)
```

# Load DE results
```{r}
DE_result_names <- DE_result_path_list <- list.files(path = paste0(path_to_data, "DE_results"), pattern = "T1_T2|T1_T3|T1_T4", full.names = FALSE) %>% str_remove(pattern = ".tsv")
DE_result_path_list <- list.files(path = paste0(path_to_data, "DE_results"), pattern = "T1_T2|T1_T3|T1_T4", full.names = TRUE)

DGE_results_list <- DE_result_path_list %>% purrr::map(function(x) x %>% read_delim(delim = "\t", escape_double = FALSE, 
     col_names = FALSE, trim_ws = TRUE, skip = 1) %>% 
  purrr::set_names("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj"))

DGE_results_list <- DGE_results_list %>% purrr::set_names(DE_result_names)
```

# Load DE results 2
```{r}
DE_result_names <- list.files(path = paste0(path_to_data, "DE_results_2"), full.names = FALSE) %>% str_remove(pattern = ".tsv")
DE_result_path_list <- list.files(path = paste0(path_to_data, "DE_results_2"), full.names = TRUE)

DGE_results_list <- DE_result_path_list %>% purrr::map(function(x) x %>% read_delim(delim = "\t", escape_double = FALSE, 
     col_names = FALSE, trim_ws = TRUE, skip = 1) %>% 
  purrr::set_names("p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "cluster", "gene") %>% 
  dplyr::select("gene", "p_val", "p_val_adj", "avg_log2FC", "pct.1", "pct.2", "cluster"))

DGE_results_list <- DGE_results_list %>% purrr::set_names(DE_result_names)
```

```{r}
DGE_results_list <- DGE_results_list %>% 
  purrr::map(dplyr::filter, cluster == "S3") %>% 
  purrr::map(arrange, p_val_adj)
```

# Load DE results 3
```{r}
DE_result_names <- list.files(path = paste0(path_to_data, "DE_results_3"), pattern = "T1_T2|T1_T3|T1_T4", full.names = FALSE) %>% str_remove(pattern = ".tsv")

DE_result_path_list <- list.files(path = paste0(path_to_data, "DE_results_3"), pattern = "T1_T2|T1_T3|T1_T4", full.names = TRUE)

DGE_results_list <- DE_result_path_list %>% purrr::map(function(x) x %>% 
                                                         read_delim(delim = "\t", 
                                                                    escape_double = FALSE, 
                                                                    col_names = FALSE, 
                                                                    trim_ws = TRUE, 
                                                                    skip = 1) %>% 
  purrr::set_names("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj") %>% 
  dplyr::select("gene", "p_val", "p_val_adj", "avg_log2FC", "pct.1", "pct.2"))

DGE_results_list <- DGE_results_list %>% purrr::set_names(DE_result_names)
```

```{r}
# Heatmap
T_cell_receptor_genes <- KEGG_2019_Mouse %>% 
  clusterProfiler::read.gmt() %>% 
  dplyr::filter(term == "T cell receptor signaling pathway") %>% 
  pull(gene) %>% 
  str_to_title() %>% 
  sort()

heatmap_df <- DGE_results_list$dge_0_T1_T2 %>% dplyr::select(gene, avg_log2FC) %>% 
  full_join(DGE_results_list$dge_0_T1_T3 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  full_join(DGE_results_list$dge_0_T1_T4 %>% dplyr::select(gene, avg_log2FC), by = "gene") %>% 
  purrr::set_names("Gene", "T1_vs_T2", "T1_vs_T3", "T1_vs_T4") %>% 
  dplyr::filter(Gene %in% T_cell_receptor_genes) %>% 
  pivot_longer(cols = T1_vs_T2:T1_vs_T4, names_to = "Comparison", values_to = "LFC")

max_expr <- max(abs(min(heatmap_df$LFC, na.rm = TRUE)), max(heatmap_df$LFC, na.rm = TRUE))

heatmap_df %>% 
  ggplot(aes(x = Comparison, y = Gene)) + 
  geom_tile(aes(fill = LFC)) +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", limits = c(-1*max_expr, max_expr))
```



# Mouse gene dictionary
```{r}
mouse_gene_id_dict <- vroom::vroom(paste0(path_to_data, "mouse_gene_id_dict.tsv"), delim = "\t", col_names = TRUE)

mouse_gene_id_dict <- mouse_gene_id_dict %>% 
  mutate(strand = strand %>% str_replace_all(pattern = "-1", replacement = "-")) %>% 
  mutate(strand = strand %>% str_replace_all(pattern = "1", replacement = "+"))

DGE_results_list <- DGE_results_list %>% purrr::map(left_join, mouse_gene_id_dict, c("gene" = "mgi_symbol"))
```



# Volcano plot
```{r}
DGE_results_list <- DGE_results_list %>% 
  purrr::map(mutate, DEG_status = dplyr::if_else(condition = avg_log2FC > 1 & p_val_adj < 0.05, true = "Up", 
                                                 false = dplyr::if_else(condition = avg_log2FC < -1 & p_val_adj < 0.05, true = "Down", 
                                                                        false = "Not_DEG")))

x_range <- DGE_results_list %>% purrr::map(function(x) x %>% pull(avg_log2FC)) %>% unlist() %>% range() %>% abs() %>% max()
y_range <- DGE_results_list %>% purrr::map(function(x) x %>% pull(p_val_adj) %>% log10() %>% '*'(-1)) %>% unlist() %>% range()

volcano_plot_list <- DGE_results_list %>% purrr::map(function(x) x %>% 
             ggplot(aes(x = avg_log2FC, y = -log10(as.numeric(p_val_adj)), colour = DEG_status)) +
             geom_point(alpha = 0.4, size = 1 / 2) +
             theme_bw() +
             scale_color_manual(values = c("Down" = "#094FED", "Not_DEG" = "black", "Up" = "red"),
                                labels = c("Downregulated DEG", "Not significant", "Upregulated DEG")) +
             xlim(c(-x_range, x_range)) + ylim(c(y_range[1], y_range[2])) +
             labs(x = "log2 Fold Change", y = "-log10 adjusted p-value") +
             theme(legend.title = element_blank()) +
             guides(color = guide_legend(override.aes = list(size = 5))) + # change size of dots in legend
             theme(legend.position = "bottom") +
             easy_all_text_size(size = 20)
           )
```

# GSEA
```{r}
gene_list <- DGE_results_list %>% 
  purrr::map(function(x) x %>% 
               arrange(desc(as.numeric(avg_log2FC))) %>% 
               distinct(gene, .keep_all = TRUE) %>% 
               mutate(rank = rank(avg_log2FC, ties.method = "random"), 
                      gene = gene %>% str_to_upper) %>% 
               arrange(desc(rank)) %>%
               dplyr::select(gene, avg_log2FC)
  )

ranked_gene_list <- purrr::map(seq_along(gene_list), function(i) gene_list[[i]] %>% 
  pull(as.numeric(avg_log2FC)) %>%
  purrr::set_names(gene_list[[i]]$gene)) %>%
  purrr::set_names(names(DGE_results_list))

# ranked_gene_list <- list(dge_0_T1_T3 = ranked_gene_list$dge_0_T1_T3,
#                          dge_0_T1_T4 = ranked_gene_list$dge_0_T1_T4)

gsea_KEGG_results_list <- purrr::map(seq_along(ranked_gene_list), function(i) gsea_analysis(database = KEGG_2019_Mouse, sorted_genes = ranked_gene_list[[i]], p_value_cutoff = 0.05, p_adjust_cutoff = 0.05, min_size = 10, max_size = 500, p_adjust_method = "BH")) %>% purrr::set_names(names(ranked_gene_list))
```

```{r}
gsea_KEGG_results_list$dge_0_T1_T3$gsea_results@result %>% vroom::vroom_write(file = paste0(path_to_data, "GSEA_results_", "dge_0_T1_T3", ".tsv"), delim = "\t")

gsea_KEGG_results_list$dge_0_T1_T4$gsea_results@result %>% vroom::vroom_write(file = paste0(path_to_data, "GSEA_results_", "dge_0_T1_T4", ".tsv"), delim = "\t")
```

```{r}
purrr::map(seq_along(gsea_KEGG_results_list), function(i) {
  gsea_KEGG_results_list[[i]]$gsea_results@result %>% 
    vroom::vroom_write(file = paste0(path_to_data, "GSEA_results_", names(gsea_KEGG_results_list)[i], ".tsv"), delim = "\t")
  })
```

# Pathview
```{r}
data(paths.hsa)
KEGG_pathway_dict_df <- data.frame(pathway = paths.hsa,
                             kegg_id = names(paths.hsa)) %>% 
  mutate(kegg_id = kegg_id %>% str_remove(pattern = "hsa")) %>% 
  rownames_to_column(var = "species_id") 

KEGG_pathway_v <- KEGG_pathway_dict_df %>% 
  pull(as.numeric(kegg_id)) %>%
  purrr::set_names(KEGG_pathway_dict_df$pathway)

KEGG_gene_list <- purrr::map(seq_along(DGE_results_list), function(i) {
  DGE_results_list[[i]] %>% 
  pull(as.numeric(avg_log2FC)) %>%
  purrr::set_names(DGE_results_list[[i]]$entrezgene_id)}
  ) %>%
  purrr::set_names(names(DGE_results_list))

setwd(paste0(path_to_data, "KEGG_pathways/"))
purrr::map(seq_along(gsea_KEGG_results_list$dge_0_T1_T3$gsea_results@result$ID), function(pathway_id) {
  print(pathway_id)
  pathview(gene.data = KEGG_gene_list$dge_0_T1_T3,
           species = "mmu",
           pathway.id = KEGG_pathway_v[gsea_KEGG_results_list$dge_0_T1_T3$gsea_results@result$ID[pathway_id]],
           kegg.dir = paste0(path_to_data, "KEGG_pathways/"),
           out.suffix = paste0(str_replace_all(str_replace_all(gsea_KEGG_results_list$dge_0_T1_T3$gsea_results@result$ID[pathway_id], pattern = " ", replacement = "_"), pattern = "\\/", replacement = ""), "_KEGG_dge_0_T1_T3"),
           kegg.native = TRUE, same.layer = TRUE)
  })

setwd(paste0(path_to_data, "KEGG_pathways/"))
purrr::map(seq_along(gsea_KEGG_results_list$dge_0_T1_T4$gsea_results@result$ID), function(pathway_id) {
  print(pathway_id)
  pathview(gene.data = KEGG_gene_list$dge_0_T1_T4,
           species = "mmu",
           pathway.id = KEGG_pathway_v[gsea_KEGG_results_list$dge_0_T1_T4$gsea_results@result$ID[pathway_id]],
           kegg.dir = paste0(path_to_data, "KEGG_pathways/"),
           out.suffix = paste0(str_replace_all(str_replace_all(gsea_KEGG_results_list$dge_0_T1_T4$gsea_results@result$ID[pathway_id], pattern = " ", replacement = "_"), pattern = "\\/", replacement = ""), "_KEGG_dge_0_T1_T4"),
           kegg.native = TRUE, same.layer = TRUE)
  })

```

# CHeck KEGG pathways for T1_T2
```{r}
interesting_pathways <- c("T cell receptor signaling pathway", "Cytokine-cytokine receptor interaction")

purrr::map(seq_along(interesting_pathways), function(pathway_id) {
  setwd(paste0(path_to_data, "KEGG_pathways/"))
  pathview(
    gene.data = KEGG_gene_list$dge_0_T1_T2,
    species = "mmu",
    pathway.id = KEGG_pathway_v[interesting_pathways[pathway_id]],
    kegg.dir = paste0(path_to_data, "KEGG_pathways/"),
    out.suffix = paste0(str_replace_all(str_replace_all(interesting_pathways[pathway_id], pattern = " ", replacement = "_"), pattern = "\\/", replacement = ""), "_KEGG_dge_0_T1_T2"),
    kegg.native = TRUE, same.layer = TRUE
  )
})

purrr::map(seq_along(interesting_pathways), function(pathway_id) {
  setwd(paste0(path_to_data, "KEGG_pathways/"))
  pathview(
    gene.data = KEGG_gene_list$dge_1_S3_S4,
    species = "mmu",
    pathway.id = KEGG_pathway_v[interesting_pathways[pathway_id]],
    kegg.dir = paste0(path_to_data, "KEGG_pathways/"),
    out.suffix = paste0(str_replace_all(str_replace_all(interesting_pathways[pathway_id], pattern = " ", replacement = "_"), pattern = "\\/", replacement = ""), "_KEGG_dge_1_S3_S4"),
    kegg.native = TRUE, same.layer = TRUE
  )
})
```

# Pathview combined results in a single figure
```{r}
# Load DE results
DE_result_names <- list.files(path = paste0(path_to_data, "DE_results"), full.names = FALSE) %>% str_remove(pattern = ".tsv")
DE_result_path_list <- list.files(path = paste0(path_to_data, "DE_results"), full.names = TRUE)

DGE_results_list <- DE_result_path_list %>% purrr::map(function(x) x %>% read_delim(delim = "\t", escape_double = FALSE, 
     col_names = FALSE, trim_ws = TRUE, skip = 1) %>% 
  purrr::set_names("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj"))

DGE_results_list <- DGE_results_list %>% purrr::set_names(DE_result_names)

# Mouse gene dictionary
mouse_gene_id_dict <- vroom::vroom(paste0(path_to_data, "mouse_gene_id_dict.tsv"), delim = "\t", col_names = TRUE)

mouse_gene_id_dict <- mouse_gene_id_dict %>% 
  mutate(strand = strand %>% str_replace_all(pattern = "-1", replacement = "-")) %>% 
  mutate(strand = strand %>% str_replace_all(pattern = "1", replacement = "+"))

DGE_results_list <- DGE_results_list %>% purrr::map(left_join, mouse_gene_id_dict, c("gene" = "mgi_symbol"))

# Pathview
data(paths.hsa)
KEGG_pathway_dict_df <- data.frame(pathway = paths.hsa,
                             kegg_id = names(paths.hsa)) %>% 
  mutate(kegg_id = kegg_id %>% str_remove(pattern = "hsa")) %>% 
  rownames_to_column(var = "species_id") 

KEGG_pathway_v <- KEGG_pathway_dict_df %>% 
  pull(as.numeric(kegg_id)) %>%
  purrr::set_names(KEGG_pathway_dict_df$pathway)


dge_0_list <- list(
  T1_T2 = DGE_results_list$dge_0_T1_T2,
  T1_T3 = DGE_results_list$dge_0_T1_T3,
  T1_T4 = DGE_results_list$dge_0_T1_T4
)

dge_1_list <- list(
  T1_T2 = DGE_results_list$dge_1_T1_T2,
  T1_T3 = DGE_results_list$dge_1_T1_T3,
  T1_T4 = DGE_results_list$dge_1_T1_T4
)

dge_2_list <- list(
  T1_T2 = DGE_results_list$dge_2_T1_T2,
  T1_T3 = DGE_results_list$dge_2_T1_T3,
  T1_T4 = DGE_results_list$dge_2_T1_T4
)

dge_3_list <- list(
  T1_T2 = DGE_results_list$dge_3_T1_T2,
  T1_T3 = DGE_results_list$dge_3_T1_T3,
  T1_T4 = DGE_results_list$dge_3_T1_T4
)


process_dge_for_pathview <- function(list){
  list %>% 
    purrr::map(dplyr::select, c("gene", "entrezgene_id", "avg_log2FC")) %>% 
    purrr::reduce(full_join, by = c("gene", "entrezgene_id")) %>% 
    purrr::set_names("gene", "entrezgene_id", "T1_T2", "T1_T3", "T1_T4") %>% 
    drop_na(entrezgene_id) %>% 
    group_by(entrezgene_id) %>% 
    dplyr::slice(1) %>% 
    # mutate(gene = gene %>% make.unique()) %>% 
    column_to_rownames(var = "entrezgene_id") %>% 
    dplyr::select(-gene)
}


dge_0_df <- dge_0_list %>% process_dge_for_pathview()
dge_1_df <- dge_1_list %>% process_dge_for_pathview()
dge_2_df <- dge_2_list %>% process_dge_for_pathview()
dge_3_df <- dge_3_list %>% process_dge_for_pathview()

dge_list = list(
  dge_0 = dge_0_df, 
  dge_1 = dge_1_df, 
  dge_2 = dge_2_df, 
  dge_3 = dge_3_df
)

# Cytokine-cytokine receptor interaction
# 04060

# purrr::map(seq_along(dge_list), function(dge){
#   pathview(
#   gene.data = dge,
#   pathway.id = "04060",
#   species = "mmu",
#   out.suffix = paste0("Cytokine_cytokine_receptor_interaction", names(dge)),
#   keys.align = "y",
#   kegg.native = TRUE,
#   match.data = FALSE,
#   multi.state = TRUE,
#   same.layer = TRUE
# )})

pathview(
  gene.data = dge_0_df,
  pathway.id = "04060",
  species = "mmu",
  out.suffix = "Cytokine_cytokine_receptor_interaction_dge_0",
  keys.align = "y",
  kegg.native = TRUE,
  match.data = FALSE,
  multi.state = TRUE,
  same.layer = TRUE
)

pathview(
  gene.data = dge_1_df,
  pathway.id = "04060",
  species = "mmu",
  out.suffix = "Cytokine_cytokine_receptor_interaction_dge_1",
  keys.align = "y",
  kegg.native = TRUE,
  match.data = FALSE,
  multi.state = TRUE,
  same.layer = TRUE
)

pathview(
  gene.data = dge_2_df,
  pathway.id = "04060",
  species = "mmu",
  out.suffix = "Cytokine_cytokine_receptor_interaction_dge_2",
  keys.align = "y",
  kegg.native = TRUE,
  match.data = FALSE,
  multi.state = TRUE,
  same.layer = TRUE
)

pathview(
  gene.data = dge_3_df,
  pathway.id = "04060",
  species = "mmu",
  out.suffix = "Cytokine_cytokine_receptor_interaction_dge_3",
  keys.align = "y",
  kegg.native = TRUE,
  match.data = FALSE,
  multi.state = TRUE,
  same.layer = TRUE
)


# T cell receptor signaling pathway
# 04660

pathview(
  gene.data = dge_0_df,
  pathway.id = "04660",
  species = "mmu",
  out.suffix = "T_cell_receptor_signaling_pathway_dge_0",
  keys.align = "y",
  kegg.native = TRUE,
  match.data = FALSE,
  multi.state = TRUE,
  same.layer = TRUE
)

pathview(
  gene.data = dge_1_df,
  pathway.id = "04660",
  species = "mmu",
  out.suffix = "T_cell_receptor_signaling_pathway_dge_1",
  keys.align = "y",
  kegg.native = TRUE,
  match.data = FALSE,
  multi.state = TRUE,
  same.layer = TRUE
)

pathview(
  gene.data = dge_2_df,
  pathway.id = "04660",
  species = "mmu",
  out.suffix = "T_cell_receptor_signaling_pathway_dge_2",
  keys.align = "y",
  kegg.native = TRUE,
  match.data = FALSE,
  multi.state = TRUE,
  same.layer = TRUE
)

pathview(
  gene.data = dge_3_df,
  pathway.id = "04660",
  species = "mmu",
  out.suffix = "T_cell_receptor_signaling_pathway_dge_3",
  keys.align = "y",
  kegg.native = TRUE,
  match.data = FALSE,
  multi.state = TRUE,
  same.layer = TRUE
)

pathview(
  gene.data = dge_0_df,
  pathway.id = "04010",
  species = "mmu",
  out.suffix = "MAPK_signaling_pathway_dge_0",
  keys.align = "y",
  kegg.native = TRUE,
  match.data = FALSE,
  multi.state = TRUE,
  same.layer = TRUE
)

pathview(
  gene.data = dge_0_df,
  pathway.id = "04062",
  species = "mmu",
  out.suffix = "Chemokine_signaling_pathway_dge_0",
  keys.align = "y",
  kegg.native = TRUE,
  match.data = FALSE,
  multi.state = TRUE,
  same.layer = TRUE
)

```


# functions
```{r}
plot_heatmap_and_dendro <- function(df) {
  
  # Load required packages
  library(ggdendro)
  library(tidyverse)

  sample_names <- names(df)
  
  # Calculate the dendrogram
  dend <- df %>%
    # column_to_rownames(var = "Gene") %>%
    as.matrix() %>%
    dist() %>%
    hclust() %>%
    as.dendrogram()
  
  dend_data <- dend %>% dendro_data()

  # Setup the data for dendrogram segments to enable an inverted layout
  segment_data <- with(
    segment(dend_data),
    data.frame(x = y, y = x, xend = yend, yend = xend)
  )

  # Use the dendrogram label data to position the gene labels
  gene_pos_table <- with(
    dend_data$labels,
    data.frame(y_center = x, Gene = as.character(label), height = 1)
  )

  # Table to position the samples
  sample_pos_table <- data.frame(Sample = sample_names) %>%
    mutate(x_center = (1:nrow(.)), width = 1)

  # Reshape the data for the heatmap
  heatmap_data <- df %>%
    rownames_to_column("Gene") %>%
    gather(value = "expr", key = "Sample", -Gene) %>%
    # reshape2::melt(value.name = "expr", varnames = c("Gene", "Sample")) %>%
    left_join(gene_pos_table) %>%
    left_join(sample_pos_table)

  # Determine the limits for the vertical axes
  gene_axis_limits <- with(
    gene_pos_table,
    c(min(y_center - 0.5 * height), max(y_center + 0.5 * height))
  ) +
    0.1 * c(-1, 1) # extra spacing: 0.1

  
  max_expr <- max(abs(floor(min(heatmap_data$expr))), abs(ceiling(max(heatmap_data$expr))))
  
  # Heatmap plot
  plt_hmap <- ggplot(
    heatmap_data,
    aes(
      x = x_center, y = y_center, fill = expr,
      height = height, width = width
    )
  ) +
    geom_tile() +
    # scale_fill_gradient2("Expr", high = "red", low = "blue") +
    scale_fill_gradientn("log2FC", 
                         limits = c(-1*max_expr, max_expr),
                         colours = rev(colorRampPalette(colors = c("red", "white", "#094FED"))(50))) +
    scale_x_continuous(
      breaks = sample_pos_table$x_center,
      labels = sample_pos_table$Sample,
      expand = c(0, 0)
    ) +
    # For the y axis, alternatively set the labels as: gene_position_table$gene
    scale_y_continuous(
      breaks = gene_pos_table[, "y_center"],
      # labels = rep("", nrow(gene_pos_table)),
      labels = heatmap_data %>% arrange(y_center) %>% pull(Gene) %>% unique(),
      limits = gene_axis_limits,
      expand = c(0, 0),
      position = "right"
    ) +
    labs(x = "Treatment", y = "") +
    theme_bw() +
    theme(
      axis.text.x = element_text(size = rel(1), hjust = 0.5, angle = 0),
      # margin: top, right, bottom, and left
      plot.margin = unit(c(1, 0.2, 0.2, -0.7), "cm"),
      panel.grid.minor = element_blank()
    ) +
    ggeasy::easy_all_text_size(size = 20) +
    ggeasy::easy_y_axis_labels_size(size = 0) +
    ggeasy::easy_rotate_x_labels(angle = 90)

  # Dendrogram plot
  plt_dendr <- ggplot(segment_data) +
    geom_segment(aes(x = x, y = y, xend = xend, yend = yend)) +
    scale_x_reverse(expand = c(0, 0.5)) +
    scale_y_continuous(
      breaks = gene_pos_table$y_center,
      # labels = gene_pos_table$Gene,
      labels = NULL,
      limits = gene_axis_limits,
      expand = c(0, 0)
    ) +
    labs(x = "Distance", y = "", colour = "", size = "") +
    theme(panel.grid.minor = element_blank()) +
    theme_bw() +
    ggeasy::easy_all_text_size(size = 20)
  
  # Combine both plots horizontally and return the result
  plot_grid(plt_dendr, plt_hmap, align = "h", rel_widths = c(0.4, 1))
}

```



```{r}
DGE_results_list$dge_0_S3_S4[, c("gene", "avg_log2FC")] %>% full_join(DGE_results_list$dge_1_S3_S4[, c("gene", "avg_log2FC")], by = "gene")

DGE_results_list$dge_0_S3_S4 %>% arrange(desc("gene"))

DGE_results_df <- DGE_results_list %>% 
  purrr::map(dplyr::select, c("gene", "avg_log2FC")) %>% 
  purrr::map(arrange, gene) %>% 
  purrr::reduce(full_join, by = c("gene")) %>% 
  purrr::set_names("gene", paste0(names(DGE_results_list), "_LFC"))

heatmap_fig <- DGE_results_df %>% plot_heatmap_and_dendro()
```



# ORA - dotplot
```{r}
path_to_gmt <- "~/PRIMUS/data/83_BIOINFORMATICS/Kadu/GMT/"

DEG_list <- DGE_results_list %>% 
  purrr::map(dplyr::filter, abs(avg_log2FC) > 1 & p_val_adj < 0.05) %>% 
  purrr::map(function(x) x %>% pull(gene) %>% str_to_upper)
  
# GO_Biological_Process_2018 <- paste0(path_to_gmt, "GO_Biological_Process_2018.txt") # GO_Biological_Process_2018 gene sets
# GO_Biological_Process_2021 <- paste0(path_to_gmt, "GO_Biological_Process_2021.txt") # GO_Biological_Process_2021 gene sets
GO_Biological_Process_2023 <- paste0(path_to_gmt, "GO_Biological_Process_2023.txt") # GO_Biological_Process_2018 gene sets

ora_results_GO_BP_2023 <- DEG_list %>% ora_analysis(database = GO_Biological_Process_2023)
GO_BP_2023_dotplot <- ora_results_GO_BP_2023 %>% clusterProfiler::dotplot()

ora_results_GO_MF_2023 <- DEG_list %>% ora_analysis(database = GO_Molecular_Function_2023)
GO_MF_2023_dotplot <- ora_results_GO_MF_2023 %>% clusterProfiler::dotplot()

ora_results_GO_CC_2023 <- DEG_list %>% ora_analysis(database = GO_Cellular_Component_2023)
GO_CC_2023_dotplot <- ora_results_GO_CC_2023 %>% clusterProfiler::dotplot()

ora_results_KEGG <- DEG_list %>% ora_analysis(database = KEGG_2019_Mouse)
KEGG_dotplot <- ora_results_KEGG %>% clusterProfiler::dotplot()

ora_results_GO_BP_2023@compareClusterResult %>% 
  vroom::vroom_write(file = paste0(path_to_data, "ora_results_GO_BP_2023.tsv"), delim = "\t")

ora_results_GO_CC_2023@compareClusterResult %>% 
  vroom::vroom_write(file = paste0(path_to_data, "ora_results_GO_CC_2023.tsv"), delim = "\t")

ora_results_GO_MF_2023@compareClusterResult %>% 
  vroom::vroom_write(file = paste0(path_to_data, "ora_results_GO_MF_2023.tsv"), delim = "\t")

ora_results_KEGG@compareClusterResult %>% 
  vroom::vroom_write(file = paste0(path_to_data, "ora_results_KEGG.tsv"), delim = "\t")


```


